From 71c81f5773b399026f75885108a634c0e1d66ab4 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 21 Feb 2013 19:21:29 -0600
Subject: [PATCH] Be less strict with illegal stances, fixes BUKKIT-2578


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 217c521..e1c91dd 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -343,7 +343,7 @@ public class PlayerConnection extends Connection {
                     d2 = packet10flying.y;
                     d3 = packet10flying.z;
                     d4 = packet10flying.stance - packet10flying.y;
-                    if (!this.player.isSleeping() && (d4 > 1.65D || d4 < 0.1D)) {
+                    if (!this.player.isSleeping() && (d4 > 1.65D /** || d4 < 0.1D **/)) { // CraftBukkit - be less strict
                         this.disconnect("Illegal stance");
                         this.minecraftServer.getLogger().warning(this.player.getName() + " had an illegal stance: " + d4);
                         return;
-- 
1.7.9.6 (Apple Git-31.1)

