From 4d60759a26e49b3838dc2f9d67b0b049849f5bfc Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 21 Feb 2013 19:21:29 -0600
Subject: [PATCH] Be less strict with illegal stances, fixes BUKKIT-2578


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 3a98439..6e7d74e 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -324,7 +324,7 @@ public class PlayerConnection implements PacketPlayInListener {
                     d2 = packetplayinflying.d();
                     d3 = packetplayinflying.e();
                     d4 = packetplayinflying.f() - packetplayinflying.d();
-                    if (!this.player.isSleeping() && (d4 > 1.65D || d4 < 0.1D)) {
+                    if (!this.player.isSleeping() && (d4 > 1.65D /** || d4 < 0.1D **/)) { // CraftBukkit - be less strict
                         this.disconnect("Illegal stance");
                         c.warn(this.player.getName() + " had an illegal stance: " + d4);
                         return;
-- 
1.8.5.2 (Apple Git-48)

