From f9ecd41a3058df1a0fb35e80992251fe952c4b44 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 25 Nov 2012 13:43:39 -0600
Subject: [PATCH] Add methods for working with arrows stuck in living entities


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 85e8d3d..6424ec5 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1293,10 +1293,12 @@ public abstract class EntityLiving extends Entity {
         return (float) this.getAttributeInstance(GenericAttributes.maxHealth).getValue();
     }
 
+    public int getArrowsStuck() { return cb(); } // SportBukkit - alias for next method
     public final int cb() {
         return ((Integer) this.datawatcher.get(EntityLiving.h)).intValue();
     }
 
+    public void setArrowsStuck(int arrows) { f(arrows); } // SportBukkit - alias for next method
     public final void f(int i) {
         this.datawatcher.set(EntityLiving.h, Integer.valueOf(i));
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index e662e4a..aad4ce6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -550,4 +550,12 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     public boolean isCollidable() {
         return getHandle().collides;
     }
+
+    public int getArrowsStuck() {
+        return getHandle().getArrowsStuck();
+    }
+
+    public void setArrowsStuck(int arrows) {
+        getHandle().setArrowsStuck(arrows);
+    }
 }
-- 
1.9.0

