From bad888598f1393915d2d6c3d4842688abd294636 Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sat, 29 Jun 2013 22:03:21 -0700
Subject: [PATCH] Add BlockFallEvent


diff --git a/src/main/java/net/minecraft/server/BlockSand.java b/src/main/java/net/minecraft/server/BlockSand.java
index 3ca6e41..0f65eb3 100644
--- a/src/main/java/net/minecraft/server/BlockSand.java
+++ b/src/main/java/net/minecraft/server/BlockSand.java
@@ -2,6 +2,8 @@ package net.minecraft.server;
 
 import java.util.Random;
 
+import org.bukkit.event.block.BlockFallEvent;  //Craftbukkit
+
 public class BlockSand extends Block {
 
     public static boolean instaFall;
@@ -38,6 +40,18 @@ public class BlockSand extends Block {
                     EntityFallingBlock entityfallingblock = new EntityFallingBlock(world, (double) ((float) i + 0.5F), (double) ((float) j + 0.5F), (double) ((float) k + 0.5F), this.id, world.getData(i, j, k));
 
                     this.a(entityfallingblock);
+
+                    // Craftbukkit begin
+                    org.bukkit.block.Block block = world.getWorld().getBlockAt(i, j, k);
+
+                    BlockFallEvent blockFallEvent = new BlockFallEvent(block, (org.bukkit.entity.FallingBlock) entityfallingblock.getBukkitEntity());
+                    world.getServer().getPluginManager().callEvent(blockFallEvent);
+                    
+                    if (blockFallEvent.isCancelled()) {
+                        return;
+                    }
+                    // Craftbukkit end
+
                     world.addEntity(entityfallingblock);
                 }
             } else {
-- 
1.7.9.6 (Apple Git-31.1)

