From 75f6422e8ed013d9524009dc955718aec6a2baaa Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Fri, 12 Apr 2013 21:34:01 -0700
Subject: [PATCH] Add BlockDispenseEntityEvent


diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorBoat.java b/src/main/java/net/minecraft/server/DispenseBehaviorBoat.java
index 5bd6ec7..e9a9ed1 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorBoat.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorBoat.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorBoat extends DispenseBehaviorItem {
@@ -37,8 +37,9 @@ final class DispenseBehaviorBoat extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        EntityBoat entityboat = new EntityBoat(world, d0, d1 + d3, d2);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2), entityboat.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -59,7 +60,7 @@ final class DispenseBehaviorBoat extends DispenseBehaviorItem {
             }
         }
 
-        EntityBoat entityboat = new EntityBoat(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+        entityboat.setPosition(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
         // CraftBukkit end
 
         world.addEntity(entityboat);
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorFireball.java b/src/main/java/net/minecraft/server/DispenseBehaviorFireball.java
index d16f035..11fbd73 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorFireball.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorFireball.java
@@ -4,7 +4,7 @@ import java.util.Random;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorFireball extends DispenseBehaviorItem {
@@ -27,8 +27,9 @@ final class DispenseBehaviorFireball extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        EntitySmallFireball entitysmallfireball = new EntitySmallFireball(world, d0, d1, d2, d3, d4, d5);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5), entitysmallfireball.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -49,8 +50,8 @@ final class DispenseBehaviorFireball extends DispenseBehaviorItem {
             }
         }
 
-        EntitySmallFireball entitysmallfireball = new EntitySmallFireball(world, d0, d1, d2, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
         entitysmallfireball.projectileSource = new org.bukkit.craftbukkit.projectiles.CraftBlockProjectileSource((TileEntityDispenser) isourceblock.getTileEntity());
+        entitysmallfireball.setDirection(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
 
         world.addEntity(entitysmallfireball);
         // itemstack.a(1); // Handled during event processing
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorFireworks.java b/src/main/java/net/minecraft/server/DispenseBehaviorFireworks.java
index 09a0842..e97b038 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorFireworks.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorFireworks.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorFireworks extends DispenseBehaviorItem {
@@ -20,8 +20,9 @@ final class DispenseBehaviorFireworks extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        EntityFireworks entityfireworks = new EntityFireworks(isourceblock.k(), d0, d1, d2, itemstack1);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2), entityfireworks.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -43,7 +44,8 @@ final class DispenseBehaviorFireworks extends DispenseBehaviorItem {
         }
 
         itemstack1 = CraftItemStack.asNMSCopy(event.getItem());
-        EntityFireworks entityfireworks = new EntityFireworks(isourceblock.k(), event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), itemstack1);
+        entityfireworks.setPosition(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+        entityfireworks.setEquipment(0, itemstack1);
 
         isourceblock.k().addEntity(entityfireworks);
         // itemstack.a(1); // Handled during event processing
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorItem.java b/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
index e7196db..9ad970f 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 public class DispenseBehaviorItem implements IDispenseBehavior {
@@ -52,7 +52,7 @@ public class DispenseBehaviorItem implements IDispenseBehavior {
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(entityitem.motX, entityitem.motY, entityitem.motZ));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(entityitem.motX, entityitem.motY, entityitem.motZ), entityitem.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorMinecart.java b/src/main/java/net/minecraft/server/DispenseBehaviorMinecart.java
index 3df54aa..06b4bb1 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorMinecart.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorMinecart.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorMinecart extends DispenseBehaviorItem {
@@ -37,8 +37,9 @@ final class DispenseBehaviorMinecart extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block2 = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, d0, d1 + d3, d2, ((ItemMinecart) itemstack1.getItem()).a);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block2, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block2, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2), entityminecartabstract.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -60,7 +61,8 @@ final class DispenseBehaviorMinecart extends DispenseBehaviorItem {
         }
 
         itemstack1 = CraftItemStack.asNMSCopy(event.getItem());
-        EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), ((ItemMinecart) itemstack1.getItem()).a);
+        entityminecartabstract.setPosition(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+        entityminecartabstract.setEquipment(((ItemMinecart) itemstack1.getItem()).a, itemstack1);
         // CraftBukkit end
 
         if (itemstack.hasName()) {
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorMonsterEgg.java b/src/main/java/net/minecraft/server/DispenseBehaviorMonsterEgg.java
index c345410..68c6a33 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorMonsterEgg.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorMonsterEgg.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorMonsterEgg extends DispenseBehaviorItem {
@@ -20,8 +20,9 @@ final class DispenseBehaviorMonsterEgg extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        Entity entity = ItemMonsterEgg.spawnCreature(isourceblock.k(), itemstack.getData(), d0, d1, d2, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2), entity.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -44,7 +45,7 @@ final class DispenseBehaviorMonsterEgg extends DispenseBehaviorItem {
 
         itemstack1 = CraftItemStack.asNMSCopy(event.getItem());
 
-        Entity entity = ItemMonsterEgg.spawnCreature(isourceblock.k(), itemstack.getData(), event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG);
+        entity.setPosition(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
 
         if (entity instanceof EntityLiving && itemstack.hasName()) {
             ((EntityInsentient) entity).setCustomName(itemstack.getName());
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorProjectile.java b/src/main/java/net/minecraft/server/DispenseBehaviorProjectile.java
index b906373..05ceb36 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorProjectile.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorProjectile.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 public abstract class DispenseBehaviorProjectile extends DispenseBehaviorItem {
@@ -20,7 +20,7 @@ public abstract class DispenseBehaviorProjectile extends DispenseBehaviorItem {
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) enumfacing.getAdjacentX(), (double) ((float) enumfacing.getAdjacentY() + 0.1F), (double) enumfacing.getAdjacentZ()));
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) enumfacing.getAdjacentX(), (double) ((float) enumfacing.getAdjacentY() + 0.1F), (double) enumfacing.getAdjacentZ()), ((Entity) iprojectile).getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorTNT.java b/src/main/java/net/minecraft/server/DispenseBehaviorTNT.java
index d949339..32fd28b 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorTNT.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorTNT.java
@@ -2,7 +2,7 @@ package net.minecraft.server;
 
 // CraftBukkit start
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
-import org.bukkit.event.block.BlockDispenseEvent;
+import org.bukkit.event.block.BlockDispenseEntityEvent;
 // CraftBukkit end
 
 final class DispenseBehaviorTNT extends DispenseBehaviorItem {
@@ -20,8 +20,10 @@ final class DispenseBehaviorTNT extends DispenseBehaviorItem {
         ItemStack itemstack1 = itemstack.a(1);
         org.bukkit.block.Block block = world.getWorld().getBlockAt(isourceblock.getBlockX(), isourceblock.getBlockY(), isourceblock.getBlockZ());
         CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+        EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, i + 0.5, j + 0.5, k + 0.5, (EntityLiving) null);
 
-        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(i + 0.5, j + 0.5, k + 0.5));
+
+        BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(i + 0.5, j + 0.5, k + 0.5), entitytntprimed.getBukkitEntity());
         if (!BlockDispenser.eventFired) {
             world.getServer().getPluginManager().callEvent(event);
         }
@@ -42,7 +44,7 @@ final class DispenseBehaviorTNT extends DispenseBehaviorItem {
             }
         }
 
-        EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), (EntityLiving) null);
+        entitytntprimed.setPosition(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
         // CraftBukkit end
 
         world.addEntity(entitytntprimed);
-- 
1.8.5.1

