From 9d5c9a8489eef766a6e1b834fe5ea65495fde796 Mon Sep 17 00:00:00 2001
From: Steve Anton <anxuiz.nx@gmail.com>
Date: Fri, 5 Oct 2012 12:00:26 -0700
Subject: [PATCH] Call BlockFormEvent for obsidian and cobblestone formation
 due to lava / water interacting.


diff --git a/src/main/java/net/minecraft/server/BlockFluids.java b/src/main/java/net/minecraft/server/BlockFluids.java
index da76b53..27179ae 100644
--- a/src/main/java/net/minecraft/server/BlockFluids.java
+++ b/src/main/java/net/minecraft/server/BlockFluids.java
@@ -163,7 +163,7 @@ public abstract class BlockFluids extends Block {
     }
 
     public void a(World world, int i, int j, int k, Entity entity, Vec3D vec3d) {
-        Vec3D vec3d1 = this.f(world, i, j, k);
+        Vec3D vec3d1 = this.f((IBlockAccess) world, i, j, k); // CraftBukkit - fix decompile error with cast
 
         vec3d.a += vec3d1.a;
         vec3d.b += vec3d1.b;
@@ -210,13 +210,27 @@ public abstract class BlockFluids extends Block {
                 if (flag) {
                     int l = world.getData(i, j, k);
 
+                    // CraftBukkit start
+                    Block formed = null;
                     if (l == 0) {
-                        world.setTypeUpdate(i, j, k, Blocks.OBSIDIAN);
+                        formed = Blocks.OBSIDIAN;
                     } else if (l <= 4) {
-                        world.setTypeUpdate(i, j, k, Blocks.COBBLESTONE);
+                        formed = Blocks.COBBLESTONE;
                     }
 
-                    this.fizz(world, i, j, k);
+                    if (formed != null) {
+                        org.bukkit.block.Block block = world.getWorld().getBlockAt(i, j, k);
+                        org.bukkit.block.BlockState newState = block.getState();
+                        newState.setType(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(formed));
+                        org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(block, newState);
+                        world.getServer().getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled()) {
+                            world.setTypeUpdate(i, j, k, formed);
+                            this.fizz(world, i, j, k);
+                        }
+                    }
+                    // CraftBukkit end
                 }
             }
         }
-- 
1.8.5.1

