From e70faa5a119889e3ff8de7905fa2c0473172d5cf Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 22 May 2013 23:14:40 -0500
Subject: [PATCH] Command Blocks shouldn't take players that can't collide with
 entites into account when doing calculations


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 9a90671..a9754e3 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -933,6 +933,7 @@ public abstract class PlayerList {
 
             for (int i2 = 0; i2 < this.players.size(); ++i2) {
                 EntityPlayer entityplayer = (EntityPlayer) this.players.get(i2);
+                if(!entityplayer.getBukkitEntity().getCollidesWithEntities()) continue; // CraftBukkit
 
                 if ((world == null || entityplayer.world == world) && (s == null || flag1 != s.equalsIgnoreCase(entityplayer.getName()))) {
                     if (s1 != null) {
-- 
1.8.5.1

