From dd81f9c085bf09bb84cc761e3d0629d8095d4674 Mon Sep 17 00:00:00 2001
From: Chad Waters <authorblues@gmail.com>
Date: Sun, 20 Jan 2013 22:04:58 -0500
Subject: [PATCH] Fire ServerCommandEvent for command block cmds. Addresses
 BUKKIT-3165


diff --git a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
index 2f5ee94..5a10775 100644
--- a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
+++ b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
@@ -161,7 +161,11 @@ public abstract class CommandBlockListenerAbstract implements ICommandListener {
             // Now dispatch all of the commands we ended up with
             for (int i = 0; i < commands.size(); i++) {
                 try {
-                    if (commandMap.dispatch(sender, joiner.join(java.util.Arrays.asList(commands.get(i))))) {
+                    String cmd = joiner.join(java.util.Arrays.asList(commands.get(i)));
+                    org.bukkit.event.server.ServerCommandEvent event = new org.bukkit.event.server.ServerCommandEvent(sender, cmd);
+                    world.getServer().getPluginManager().callEvent(event);
+
+                    if (commandMap.dispatch(sender, event.getCommand())) {
                         completed++;
                     }
                 } catch (Throwable exception) {
-- 
1.8.5.1

