From 8c712bef83b4cf7ea9ed2c234711f3011b0ea8e4 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 16 Jun 2013 17:03:07 -0500
Subject: [PATCH] Entity ID API


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 23590ac..754df42 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -59,7 +59,7 @@ public abstract class Entity implements ICommandListener {
     private static final Logger a = LogManager.getLogger();
     private static final AxisAlignedBB b = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
     private static double c = 1.0D;
-    private static int entityCount;
+    public static int entityCount; // SportBukkit - private -> public
     private int id;
     public boolean i;
     public final List<Entity> passengers;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 7cb0a18..e9ac9ab 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1608,6 +1608,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public int allocateEntityId() {
+        return net.minecraft.server.Entity.entityCount++;
+    }
+
+    @Override
     public CraftIconCache getServerIcon() {
         return icon;
     }
-- 
1.9.0

