From 0991a549df8071ade1d74e1d0c19ddc0ef6a5d3f Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 16 Jun 2013 17:03:07 -0500
Subject: [PATCH] Entity ID API


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 27f9bcd..1df2755 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -61,7 +61,7 @@ public abstract class Entity implements ICommandListener {
     private static final List<ItemStack> b = Collections.emptyList();
     private static final AxisAlignedBB c = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
     private static double f = 1.0D;
-    private static int entityCount;
+    public static int entityCount; // SportBukkit - private -> public
     private int id;
     public boolean i;
     public final List<Entity> passengers;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 7a3edca..c6eca7a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1611,6 +1611,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public int allocateEntityId() {
+        return net.minecraft.server.Entity.entityCount++;
+    }
+
+    @Override
     public CraftIconCache getServerIcon() {
         return icon;
     }
-- 
1.9.0

