From 4efe12fb315944f5467f2fd3e8339138be85c828 Mon Sep 17 00:00:00 2001
From: Chad Waters <authorblues@gmail.com>
Date: Thu, 2 Aug 2012 16:47:02 -0700
Subject: [PATCH] Fire ExplosionPrimeEvent properly. Fixes BUKKIT-770


diff --git a/src/main/java/net/minecraft/server/BlockTNT.java b/src/main/java/net/minecraft/server/BlockTNT.java
index 8cc7d5e..f1cd88c 100644
--- a/src/main/java/net/minecraft/server/BlockTNT.java
+++ b/src/main/java/net/minecraft/server/BlockTNT.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.Random;
 
 public class BlockTNT extends Block {
+    boolean removeBlock = false; // CraftBukkit
 
     public BlockTNT() {
         super(Material.TNT);
@@ -32,8 +33,17 @@ public class BlockTNT extends Block {
         if (!world.isStatic) {
             EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) i + 0.5F), (double) ((float) j + 0.5F), (double) ((float) k + 0.5F), explosion.c());
 
-            entitytntprimed.fuseTicks = world.random.nextInt(entitytntprimed.fuseTicks / 4) + entitytntprimed.fuseTicks / 8;
-            world.addEntity(entitytntprimed);
+            // CraftBukkit start - allow exploded TNT to cancel priming
+            org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+            world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                entitytntprimed.fuseTicks = world.random.nextInt(entitytntprimed.fuseTicks / 4) + entitytntprimed.fuseTicks / 8;
+                world.addEntity(entitytntprimed);
+            } else {
+                this.a(world, i, j, k, new ItemStack(this, 1, 0));
+            }
+            // CraftBukkit end
         }
     }
 
@@ -42,20 +52,34 @@ public class BlockTNT extends Block {
     }
 
     public void a(World world, int i, int j, int k, int l, EntityLiving entityliving) {
+        removeBlock = true; // CraftBukkit - TNT priming may be cancelled
         if (!world.isStatic) {
             if ((l & 1) == 1) {
                 EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) i + 0.5F), (double) ((float) j + 0.5F), (double) ((float) k + 0.5F), entityliving);
 
-                world.addEntity(entitytntprimed);
-                world.makeSound(entitytntprimed, "game.tnt.primed", 1.0F, 1.0F);
+                // CraftBukkit start - fire ExplosionPrimeEvent
+                org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    removeBlock = false;
+                } else {
+                    world.addEntity(entitytntprimed);
+                    world.makeSound(entitytntprimed, "game.tnt.primed", 1.0F, 1.0F);
+                }
+                // CraftBukkit end
             }
         }
     }
 
     public boolean interact(World world, int i, int j, int k, EntityHuman entityhuman, int l, float f, float f1, float f2) {
         if (entityhuman.bF() != null && entityhuman.bF().getItem() == Items.FLINT_AND_STEEL) {
-            this.a(world, i, j, k, 1, entityhuman);
-            world.setAir(i, j, k);
+            // CraftBukkit start - don't remove block if TNT priming is cancelled
+            this.a(world, i, j, k, 1, (EntityLiving) entityhuman);
+            if (removeBlock) {
+                world.setAir(i, j, k);
+            }
+            // CraftBukkit end
             entityhuman.bF().damage(1, entityhuman);
             return true;
         } else {
diff --git a/src/main/java/net/minecraft/server/EntityCreeper.java b/src/main/java/net/minecraft/server/EntityCreeper.java
index a8162c7..4a067b7 100644
--- a/src/main/java/net/minecraft/server/EntityCreeper.java
+++ b/src/main/java/net/minecraft/server/EntityCreeper.java
@@ -90,9 +90,20 @@ public class EntityCreeper extends EntityMonster {
 
             int i = this.cb();
 
+            // CraftBukkit start - fire ExplosionPrimeEvent
+            float radius = this.isPowered() ? 6.0F : 3.0F;
+
             if (i > 0 && this.fuseTicks == 0) {
-                this.makeSound("creeper.primed", 1.0F, 0.5F);
+                ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), radius, false);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    i = 0;
+                } else {
+                    this.makeSound("creeper.primed", 1.0F, 0.5F);
+                }
             }
+            // CraftBukkit end
 
             this.fuseTicks += i;
             if (this.fuseTicks < 0) {
diff --git a/src/main/java/net/minecraft/server/EntityLargeFireball.java b/src/main/java/net/minecraft/server/EntityLargeFireball.java
index 326f7ea..b3ab7ee 100644
--- a/src/main/java/net/minecraft/server/EntityLargeFireball.java
+++ b/src/main/java/net/minecraft/server/EntityLargeFireball.java
@@ -20,16 +20,8 @@ public class EntityLargeFireball extends EntityFireball {
                 movingobjectposition.entity.damageEntity(DamageSource.fireball(this, this.shooter), 6.0F);
             }
 
-            // CraftBukkit start - fire ExplosionPrimeEvent
-            ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.entity.CraftEntity.getEntity(this.world.getServer(), this));
-            this.world.getServer().getPluginManager().callEvent(event);
-
-            if (!event.isCancelled()) {
-                // give 'this' instead of (Entity) null so we know what causes the damage
-                this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), this.world.getGameRules().getBoolean("mobGriefing"));
-            }
-            // CraftBukkit end
-
+            // CraftBukkit - give 'this' instead of (Entity) null so we know what causes the damage
+            this.world.createExplosion(this, this.locX, this.locY, this.locZ, yield, isIncendiary, this.world.getGameRules().getBoolean("mobGriefing"));
             this.die();
         }
     }
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index 13cbc79..f557bc0 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -73,15 +73,10 @@ public class EntityTNTPrimed extends Entity {
         // float f = 4.0F;
 
         org.bukkit.craftbukkit.CraftServer server = this.world.getServer();
+        org.bukkit.entity.Explosive explosive = (org.bukkit.entity.Explosive) org.bukkit.craftbukkit.entity.CraftEntity.getEntity(server, this);
 
-        ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.entity.CraftEntity.getEntity(server, this));
-        server.getPluginManager().callEvent(event);
-
-        if (!event.isCancelled()) {
-            // give 'this' instead of (Entity) null so we know what causes the damage
-            this.world.createExplosion(this, this.locX, this.locY, this.locZ, event.getRadius(), event.getFire(), true);
-        }
-        // CraftBukkit end
+        // give 'this' instead of (Entity) null so we know what causes the damage
+        this.world.createExplosion(this, this.locX, this.locY, this.locZ, explosive.getYield(), explosive.isIncendiary(), true);
     }
 
     protected void b(NBTTagCompound nbttagcompound) {
-- 
1.8.5.2 (Apple Git-48)

