From 463261a08b6c3818c02d3c547f63e0fee510d7d3 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Jan 2013 14:12:03 -0500
Subject: [PATCH] Add onGround API


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 27e52ea..bddb407 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -224,6 +224,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                 Player player = this.getPlayer();
                 Location from = new Location(player.getWorld(), lastPosX, lastPosY, lastPosZ, lastYaw, lastPitch); // Get the Players previous Event location.
                 Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
+                boolean wasOnGround = this.player.onGround;
 
                 // If the packet contains movement information then we update the To location with the correct XYZ.
                 if (packetplayinflying.hasPos && !(packetplayinflying.hasPos && packetplayinflying.y == -999.0D)) {
@@ -437,6 +438,12 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
 
                     this.player.onGround = packetplayinflying.f();
                     this.minecraftServer.getPlayerList().d(this.player);
+                    // SportBukkit start
+                    if(wasOnGround != this.player.onGround) {
+                        org.bukkit.event.player.PlayerOnGroundEvent onGroundEvent = new org.bukkit.event.player.PlayerOnGroundEvent(player, this.player.onGround);
+                        this.server.getPluginManager().callEvent(onGroundEvent);
+                    }
+                    // SportBukkit end
                     this.player.a(this.player.locY - d10, packetplayinflying.f());
                 } else if (this.e - this.f > 20) {
                     this.a(this.o, this.p, this.q, this.player.yaw, this.player.pitch);
@@ -1113,7 +1120,6 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         if (movingobjectposition == null || movingobjectposition.type != MovingObjectPosition.EnumMovingObjectType.BLOCK) {
             CraftEventFactory.callPlayerInteractEvent(this.player, Action.LEFT_CLICK_AIR, this.player.inventory.getItemInHand());
         }
-
         // Arm swing animation
         PlayerAnimationEvent event = new PlayerAnimationEvent(this.getPlayer());
         this.server.getPluginManager().callEvent(event);
@@ -1785,7 +1791,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     }
 
     public void a(PacketPlayInKeepAlive packetplayinkeepalive) {
-        if (packetplayinkeepalive.a() == this.i) {
+        if(packetplayinkeepalive.a() == this.i) {
             int i = (int) (this.d() - this.j);
 
             this.player.ping = (this.player.ping * 3 + i) / 4;
-- 
1.9.0

