From 31230799a72aa436781bf92b9b3a088fb6dc71fd Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 16 May 2013 20:20:34 -0500
Subject: [PATCH] Add Ignorant Effects API


diff --git a/src/main/java/net/minecraft/server/DataWatcher.java b/src/main/java/net/minecraft/server/DataWatcher.java
index b05cfd8..86a1977 100644
--- a/src/main/java/net/minecraft/server/DataWatcher.java
+++ b/src/main/java/net/minecraft/server/DataWatcher.java
@@ -16,7 +16,7 @@ public class DataWatcher {
     private boolean b = true;
     private static final HashMap c = new HashMap();
     private final Map d = new HashMap();
-    private boolean e;
+    public boolean e; // CraftBukkit - private -> public
     private ReadWriteLock f = new ReentrantReadWriteLock();
 
     public DataWatcher(Entity entity) {
@@ -75,7 +75,8 @@ public class DataWatcher {
         return (ItemStack) this.i(i).b();
     }
 
-    private WatchableObject i(int i) {
+    // CraftBukkit - private -> public
+    public WatchableObject i(int i) {
         this.f.readLock().lock();
 
         WatchableObject watchableobject;
@@ -114,7 +115,7 @@ public class DataWatcher {
         return this.e;
     }
 
-    public static void a(List list, PacketDataSerializer packetdataserializer) {
+    public static void a(List list, PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         if (list != null) {
             Iterator iterator = list.iterator();
 
@@ -155,7 +156,7 @@ public class DataWatcher {
         return arraylist;
     }
 
-    public void a(PacketDataSerializer packetdataserializer) {
+    public void a(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         this.f.readLock().lock();
         Iterator iterator = this.d.values().iterator();
 
@@ -187,7 +188,7 @@ public class DataWatcher {
         return arraylist;
     }
 
-    private static void a(PacketDataSerializer packetdataserializer, WatchableObject watchableobject) {
+    private static void a(PacketDataSerializer packetdataserializer, WatchableObject watchableobject) throws java.io.IOException { // CraftBukkit - added throws
         int i = (watchableobject.c() << 5 | watchableobject.a() & 31) & 255;
 
         packetdataserializer.writeByte(i);
@@ -227,7 +228,7 @@ public class DataWatcher {
         }
     }
 
-    public static List b(PacketDataSerializer packetdataserializer) {
+    public static List b(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         ArrayList arraylist = null;
 
         for (byte b0 = packetdataserializer.readByte(); b0 != 127; b0 = packetdataserializer.readByte()) {
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 53b3e55..3f17a47 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -899,12 +899,18 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     protected void a(MobEffect mobeffect) {
         super.a(mobeffect);
-        this.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.getId(), mobeffect));
+        // CraftBukkit - account for ignorant effects
+        if (!this.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(mobeffect.getEffectId()))) {
+            this.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.getId(), mobeffect));
+        }
     }
 
     protected void a(MobEffect mobeffect, boolean flag) {
         super.a(mobeffect, flag);
-        this.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.getId(), mobeffect));
+        // CraftBukkit - account for ignorant effects
+        if (!this.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(mobeffect.getEffectId()))) {
+            this.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.getId(), mobeffect));
+        }
     }
 
     protected void b(MobEffect mobeffect) {
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 0cf9ad5..f0fb729 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import java.util.ArrayList; // CraftBukkit
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -99,7 +100,7 @@ public class EntityTrackerEntry {
                 }
             }
 
-            this.b();
+            this.sendMetadata(this.tracker.getDataWatcher(), false); // CraftBukkit - custom metadata sender
         } else if (this.m % this.c == 0 || this.tracker.al || this.tracker.getDataWatcher().a()) {
             int i;
             int j;
@@ -170,7 +171,7 @@ public class EntityTrackerEntry {
                     this.broadcast((Packet) object);
                 }
 
-                this.b();
+                this.sendMetadata(this.tracker.getDataWatcher(), false); // CraftBukkit - custom metadata sender
                 /* CraftBukkit start - Code moved up
                 if (flag) {
                     this.xLoc = i;
@@ -199,7 +200,7 @@ public class EntityTrackerEntry {
                 this.xLoc = this.tracker.as.a(this.tracker.locX);
                 this.yLoc = MathHelper.floor(this.tracker.locY * 32.0D);
                 this.zLoc = this.tracker.as.a(this.tracker.locZ);
-                this.b();
+                this.sendMetadata(this.tracker.getDataWatcher(), false); // CraftBukkit - custom metadata sender
                 this.x = true;
             }
 
@@ -281,6 +282,64 @@ public class EntityTrackerEntry {
         }
     }
 
+    // CraftBukkit start - custom metadata sender
+    public void sendMetadata(DataWatcher datawatcher, boolean flag) {
+        if (datawatcher.a()) {
+            this.sendMetadata(datawatcher, flag, null);
+        }
+
+        if (!flag && this.tracker instanceof EntityLiving) {
+            AttributeMapServer attributemapserver = (AttributeMapServer) ((EntityLiving) this.tracker).getAttributeMap();
+            Set set = attributemapserver.getAttributes();
+
+            if (!set.isEmpty()) {
+                this.broadcastIncludingSelf(new PacketPlayOutUpdateAttributes(this.tracker.getId(), set));
+            }
+
+            set.clear();
+        }
+    }
+
+    public void sendMetadata(DataWatcher datawatcher, boolean flag, EntityPlayer entityplayer) {
+        List dataNormal = flag ? datawatcher.c() : datawatcher.b();
+        List dataVisible = new ArrayList(dataNormal);
+
+        if(this.tracker.isInvisible()) {
+            WatchableObject normal = datawatcher.i(0);
+            byte b0 = ((Byte) normal.b()).byteValue();
+            WatchableObject visible = new WatchableObject(normal.c(), normal.a(), Byte.valueOf((byte) (b0 & ~(1 << 5))));
+
+            dataVisible.remove(normal);
+            dataVisible.add(visible);
+        }
+
+        PacketPlayOutEntityMetadata normal = new PacketPlayOutEntityMetadata(this.tracker.getId(), dataNormal);
+        PacketPlayOutEntityMetadata visible = new PacketPlayOutEntityMetadata(this.tracker.getId(), dataVisible);
+        Iterator iterator = this.trackedPlayers.iterator();
+
+        if(entityplayer == null) {
+            while (iterator.hasNext()) {
+                EntityPlayer entityplayer2 = (EntityPlayer) iterator.next();
+                this.sendMetadata(entityplayer2, normal, visible);
+            }
+        }
+
+        if (entityplayer == null && this.tracker instanceof EntityPlayer) {
+            this.sendMetadata((EntityPlayer) this.tracker, normal, visible);
+        } else if (entityplayer != null) {
+            this.sendMetadata(entityplayer, normal, visible);
+        }
+    }
+
+    private void sendMetadata(EntityPlayer entityplayer, PacketPlayOutEntityMetadata normal, PacketPlayOutEntityMetadata visible) {
+        if (entityplayer.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(MobEffectList.INVISIBILITY.getId()))) {
+            entityplayer.playerConnection.sendPacket(visible);
+        } else {
+            entityplayer.playerConnection.sendPacket(normal);
+        }
+    }
+    // CraftBukkit end
+
     public void a() {
         Iterator iterator = this.trackedPlayers.iterator();
 
@@ -321,7 +380,7 @@ public class EntityTrackerEntry {
 
                     entityplayer.playerConnection.sendPacket(packet);
                     if (!this.tracker.getDataWatcher().d()) {
-                        entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityMetadata(this.tracker.getId(), this.tracker.getDataWatcher(), true));
+                        this.sendMetadata(this.tracker.getDataWatcher(), true, entityplayer); // CraftBukkit - custom metadata sender
                     }
 
                     if (this.tracker instanceof EntityLiving) {
@@ -388,8 +447,10 @@ public class EntityTrackerEntry {
 
                         while (iterator.hasNext()) {
                             MobEffect mobeffect = (MobEffect) iterator.next();
-
-                            entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.tracker.getId(), mobeffect));
+                            // CraftBukkit - account for ignorant effects
+                            if (!entityplayer.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(mobeffect.getEffectId()))) {
+                                entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.tracker.getId(), mobeffect));
+                            }
                         }
                     }
                 }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java b/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
index a95877f..42c99a1 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntityMetadata.java
@@ -18,12 +18,19 @@ public class PacketPlayOutEntityMetadata extends Packet {
         }
     }
 
-    public void a(PacketDataSerializer packetdataserializer) {
+    // CraftBukkit start - add constructor
+    public PacketPlayOutEntityMetadata(int i, List list) {
+        this.a = i;
+        this.b = list;
+    }
+    // CraftBukkit end
+
+    public void a(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         this.a = packetdataserializer.readInt();
         this.b = DataWatcher.b(packetdataserializer);
     }
 
-    public void b(PacketDataSerializer packetdataserializer) {
+    public void b(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         packetdataserializer.writeInt(this.a);
         DataWatcher.a(this.b, packetdataserializer);
     }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
index a58dc60..cfb98ab 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -14,13 +14,13 @@ public class PacketPlayOutPlayerInfo extends Packet {
         this.c = i;
     }
 
-    public void a(PacketDataSerializer packetdataserializer) {
+    public void a(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         this.a = packetdataserializer.c(16);
         this.b = packetdataserializer.readBoolean();
         this.c = packetdataserializer.readShort();
     }
 
-    public void b(PacketDataSerializer packetdataserializer) {
+    public void b(PacketDataSerializer packetdataserializer) throws java.io.IOException { // CraftBukkit - added throws
         packetdataserializer.a(this.a);
         packetdataserializer.writeBoolean(this.b);
         packetdataserializer.writeShort(this.c);
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 7121e9c..e29a5ca 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -148,7 +148,10 @@ public abstract class PlayerList {
         while (iterator.hasNext()) {
             MobEffect mobeffect = (MobEffect) iterator.next();
 
-            playerconnection.sendPacket(new PacketPlayOutEntityEffect(entityplayer.getId(), mobeffect));
+            // CraftBukkit - account for ignorant effects
+            if (!entityplayer.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(mobeffect.getEffectId()))) {
+                playerconnection.sendPacket(new PacketPlayOutEntityEffect(entityplayer.getId(), mobeffect));
+            }
         }
 
         entityplayer.syncInventory();
@@ -538,7 +541,10 @@ public abstract class PlayerList {
         while (iterator.hasNext()) {
             MobEffect mobeffect = (MobEffect) iterator.next();
 
-            entityplayer1.playerConnection.sendPacket(new PacketPlayOutEntityEffect(entityplayer1.getId(), mobeffect));
+            // CraftBukkit - account for ignorant effects
+            if (!entityplayer1.getBukkitEntity().hasIgnorantEffect(org.bukkit.potion.PotionEffectType.getById(mobeffect.getEffectId()))) {
+                entityplayer1.playerConnection.sendPacket(new PacketPlayOutEntityEffect(entityplayer1.getId(), mobeffect));
+            }
         }
         // entityplayer1.syncInventory();
         // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/WatchableObject.java b/src/main/java/net/minecraft/server/WatchableObject.java
index 678aa91..cb84fef 100644
--- a/src/main/java/net/minecraft/server/WatchableObject.java
+++ b/src/main/java/net/minecraft/server/WatchableObject.java
@@ -38,7 +38,8 @@ public class WatchableObject {
         this.d = flag;
     }
 
-    static boolean a(WatchableObject watchableobject, boolean flag) {
+    // CraftBukkit - public
+    public static boolean a(WatchableObject watchableobject, boolean flag) {
         return watchableobject.d = flag;
     }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index bbf310b..6a7ccdb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -57,6 +57,7 @@ import org.bukkit.map.MapView;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.StandardMessenger;
+import org.bukkit.potion.PotionEffectType;
 import org.bukkit.scoreboard.Scoreboard;
 
 @DelegateDeserialization(CraftOfflinePlayer.class)
@@ -67,6 +68,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private final ConversationTracker conversationTracker = new ConversationTracker();
     private final Set<String> channels = new HashSet<String>();
     private final Set<UUID> hiddenPlayers = new HashSet<UUID>();
+    private final Set<PotionEffectType> ignorantEffects = new HashSet<PotionEffectType>();
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
@@ -1287,4 +1289,43 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         collection.add(new AttributeModifiable(getHandle().getAttributeMap(), (new AttributeRanged("generic.maxHealth", scaledHealth ? healthScale : getMaxHealth(), 0.0D, Float.MAX_VALUE)).a("Max Health").a(true)));
     }
+
+    public Set<PotionEffectType> getIgnorantEffects() {
+        return ImmutableSet.copyOf(this.ignorantEffects);
+    }
+
+    public void clearIgnorantEffects() {
+        this.ignorantEffects.clear();
+
+        refreshTrackedMetadata();
+    }
+
+    public void addIgnorantEffect(PotionEffectType effect) {
+        Validate.notNull(effect, "Effect cannot be null");
+
+        this.ignorantEffects.add(effect);
+
+        refreshTrackedMetadata();
+    }
+
+    public void removeIgnorantEffect(PotionEffectType effect) {
+        Validate.notNull(effect, "Effect cannot be null");
+
+        this.ignorantEffects.remove(effect);
+
+        refreshTrackedMetadata();
+    }
+
+    public boolean hasIgnorantEffect(PotionEffectType effect) {
+        Validate.notNull(effect, "Effect cannot be null");
+
+        return this.ignorantEffects.contains(effect);
+    }
+
+    private void refreshTrackedMetadata() {
+        for(Object obj : this.getHandle().world.players) {
+            EntityPlayer player = (EntityPlayer) obj;
+            player.setInvisible(player.isInvisible());
+        }
+    }
 }
-- 
1.8.5.2 (Apple Git-48)

