From 53d8f51cb43fad42084e01d7e3b78af589633238 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Jan 2013 14:12:03 -0500
Subject: [PATCH] Add PlayerOnGroundEvent


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index c5184ea..02e9336 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -99,6 +99,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     private int G;
     private boolean processedDisconnect; // CraftBukkit - Added
 
+    boolean clientOnGround; // SportBukkit - only set by client
+
     public PlayerConnection(MinecraftServer minecraftserver, NetworkManager networkmanager, EntityPlayer entityplayer) {
         this.minecraftServer = minecraftserver;
         this.networkManager = networkmanager;
@@ -623,6 +625,13 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                         this.B &= !this.player.hasEffect(MobEffects.LEVITATION) && !this.player.cC() && !worldserver.d(this.player.getBoundingBox().g(0.0625D).a(0.0D, -0.55D, 0.0D));
                         this.player.onGround = packetplayinflying.a();
                         this.minecraftServer.getPlayerList().d(this.player);
+                        // SportBukkit start
+                        if(clientOnGround != this.player.onGround) {
+                            clientOnGround = this.player.onGround;
+                            org.bukkit.event.player.PlayerOnGroundEvent onGroundEvent = new org.bukkit.event.player.PlayerOnGroundEvent(player, this.player.onGround);
+                            this.server.getPluginManager().callEvent(onGroundEvent);
+                        }
+                        // SportBukkit end
                         this.player.a(this.player.locY - d3, packetplayinflying.a());
                         this.o = this.player.locX;
                         this.p = this.player.locY;
-- 
1.9.0

