From 8b9738a92f699f1c28378280a5db26243fd008ee Mon Sep 17 00:00:00 2001
From: Steve Anton <anxuiz.nx@gmail.com>
Date: Fri, 31 Aug 2012 12:30:14 -0700
Subject: [PATCH] More complex logic when looking for nearby players


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 8f9aabc..411439a 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2545,7 +2545,7 @@ public abstract class World implements IBlockAccess {
         for (int i = 0; i < this.players.size(); ++i) {
             EntityHuman entityhuman1 = (EntityHuman) this.players.get(i);
             // CraftBukkit start - Fixed an NPE
-            if (entityhuman1 == null || entityhuman1.dead) {
+            if (entityhuman1 == null || !entityhuman1.R()) { // CraftBukkit allow for more complex logic by using R()
                 continue;
             }
             // CraftBukkit end
-- 
1.8.5.2 (Apple Git-48)

