From 2b4efe97739a505d045f2646185b4fbd48d9b539 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 13 Nov 2012 19:38:11 -0600
Subject: [PATCH] Enable multiworld maps


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 03b4435..6525118 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -722,7 +722,7 @@ public final class CraftServer implements Server {
             return null;
         }
 
-        internal.worldMaps = console.worlds.get(0).worldMaps;
+        // internal.worldMaps = console.worlds.get(0).worldMaps; // CraftBukkit - enable multiworld maps
         internal.scoreboard = getScoreboardManager().getMainScoreboard().getHandle();
 
         internal.tracker = new EntityTracker(internal);
-- 
1.7.9.6 (Apple Git-31.1)

