From 9c935a3110f7f30356c95e3eec8dfefad5453024 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 30 Mar 2013 19:22:48 -0500
Subject: [PATCH] Don't filter negative damage values


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index f9aaf4a..d6cca2e 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -270,9 +270,10 @@ public final class ItemStack {
         // CraftBukkit end
 
         this.damage = i;
-        if (this.damage < -1) { // CraftBukkit - don't filter -1, we use it
-            this.damage = 0;
-        }
+        //if (this.damage < -1) { // CraftBukkit - don't filter negative data values
+        //    this.damage = 0;
+        //}
+        // CraftBukkit end
     }
 
     public int l() {
-- 
1.8.5.2 (Apple Git-48)

