From 1072cad0fc5f5f3e825720bcabb7125331217a8b Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Jan 2013 14:12:03 -0500
Subject: [PATCH] Add onGround API


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index c965557..e3dde17 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -205,6 +205,7 @@ public class PlayerConnection implements PacketPlayInListener {
             Player player = this.getPlayer();
             Location from = new Location(player.getWorld(), lastPosX, lastPosY, lastPosZ, lastYaw, lastPitch); // Get the Players previous Event location.
             Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
+            boolean wasOnGround = this.player.onGround;
 
             // If the packet contains movement information then we update the To location with the correct XYZ.
             if (packetplayinflying.hasPos && !(packetplayinflying.hasPos && packetplayinflying.y == -999.0D && packetplayinflying.stance == -999.0D)) {
@@ -416,6 +417,12 @@ public class PlayerConnection implements PacketPlayInListener {
 
                 this.player.onGround = packetplayinflying.i();
                 this.minecraftServer.getPlayerList().d(this.player);
+                // CraftBukkit start
+                if (wasOnGround != this.player.onGround) {
+                    org.bukkit.event.player.PlayerOnGroundEvent onGroundEvent = new org.bukkit.event.player.PlayerOnGroundEvent(player, this.player.onGround);
+                    this.server.getPluginManager().callEvent(onGroundEvent);
+                }
+                // CraftBukkit end
                 this.player.b(this.player.locY - d0, packetplayinflying.i());
             } else if (this.e % 20 == 0) {
                 this.a(this.y, this.z, this.q, this.player.yaw, this.player.pitch);
-- 
1.8.5.2 (Apple Git-48)

