From 67f157a7fb43856aed38dcaf59fa4f531b723fed Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 30 Mar 2013 19:22:48 -0500
Subject: [PATCH] Don't filter negative damage values


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index b308e03..77e0154 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -164,12 +164,12 @@ public final class ItemStack {
                 i = 0;
             }
         }
-        // CraftBukkit end
 
         this.damage = i;
-        if (this.damage < -1) { // CraftBukkit - don't filter -1, we use it
-            this.damage = 0;
-        }
+        //if (this.damage < -1) { // CraftBukkit - don't filter negative data values
+        //    this.damage = 0;
+        //}
+        // CraftBukkit end
     }
 
     public int l() {
-- 
1.8.5.1

