From 8bc5409575d6ec70ac4a510c45779bd7ba5cab18 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 23 Apr 2013 19:45:55 -0500
Subject: [PATCH] Call AsyncPlayerPreLoginEvent regardless of online-mode


diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index 44a6bfe..3a6f269 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -106,7 +106,8 @@ public class LoginListener implements PacketLoginInListener {
             this.g = EnumProtocolState.KEY;
             this.networkManager.handle(new PacketLoginOutEncryptionBegin(this.j, this.server.J().getPublic(), this.e), new GenericFutureListener[0]);
         } else {
-            this.g = EnumProtocolState.READY_TO_ACCEPT;
+            // CraftBukkit - call async login event no matter what
+            (new ThreadPlayerLookupUUID(this, "User Authenticator #" + b.incrementAndGet())).start();
         }
     }
 
diff --git a/src/main/java/net/minecraft/server/ThreadPlayerLookupUUID.java b/src/main/java/net/minecraft/server/ThreadPlayerLookupUUID.java
index 1aa2ce4..ab470dd 100644
--- a/src/main/java/net/minecraft/server/ThreadPlayerLookupUUID.java
+++ b/src/main/java/net/minecraft/server/ThreadPlayerLookupUUID.java
@@ -22,15 +22,21 @@ class ThreadPlayerLookupUUID extends Thread {
     }
 
     public void run() {
-        boolean allowed = false; // CraftBukkit
-        try {
-            String s = (new BigInteger(MinecraftEncryption.a(LoginListener.a(this.a), LoginListener.b(this.a).J().getPublic(), LoginListener.c(this.a)))).toString(16);
+        // CraftBukkit start
+        final org.bukkit.craftbukkit.CraftServer server = LoginListener.b(this.a).server;
+        boolean allowed = false;
 
-            LoginListener.a(this.a, LoginListener.b(this.a).at().hasJoinedServer(new GameProfile((String) null, LoginListener.d(this.a).getName()), s));
+        if (server.getOnlineMode()) {
+            try {
+                String s = (new BigInteger(MinecraftEncryption.a(LoginListener.a(this.a), LoginListener.b(this.a).J().getPublic(), LoginListener.c(this.a)))).toString(16);
 
-            // CraftBukkit start - fire PlayerPreLoginEvent
-            allowed = LoginListener.d(this.a) != null;
-        } catch(Exception exception) {}
+                LoginListener.a(this.a, LoginListener.b(this.a).at().hasJoinedServer(new GameProfile((String) null, LoginListener.d(this.a).getName()), s));
+
+                allowed = LoginListener.d(this.a) != null;
+            } catch(Exception exception) {}
+        } else {
+            allowed = true;
+        }
 
         if (!this.a.networkManager.isConnected()) {
             return;
@@ -38,7 +44,6 @@ class ThreadPlayerLookupUUID extends Thread {
 
         String playerName = LoginListener.d(this.a).getName();
         java.net.InetAddress address = ((java.net.InetSocketAddress) a.networkManager.getSocketAddress()).getAddress();
-        final org.bukkit.craftbukkit.CraftServer server = LoginListener.b(this.a).server;
 
         AsyncPlayerPreLoginEvent asyncEvent = new AsyncPlayerPreLoginEvent(playerName, address, allowed);
         server.getPluginManager().callEvent(asyncEvent);
-- 
1.8.5.1

