From 713232b36eb40e4c89fb406f05eeddda2255c233 Mon Sep 17 00:00:00 2001
From: Chad Waters <authorblues@gmail.com>
Date: Wed, 6 Feb 2013 14:45:43 -0500
Subject: [PATCH] Add safe methods from @md-5's offlineUtil


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index dd6bcf9..3ab69d5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -4,19 +4,27 @@ import java.io.File;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.UUID;
 
 import net.minecraft.server.BanEntry;
 import net.minecraft.server.EntityPlayer;
+import net.minecraft.server.NBTCompressedStreamTools;
 import net.minecraft.server.NBTTagCompound;
+import net.minecraft.server.NBTTagDouble;
+import net.minecraft.server.NBTTagFloat;
+import net.minecraft.server.NBTTagList;
 import net.minecraft.server.WorldNBTStorage;
 
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.OfflinePlayer;
 import org.bukkit.Server;
+import org.bukkit.World;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.configuration.serialization.SerializableAs;
+import org.bukkit.craftbukkit.inventory.CraftInventoryPlayer;
 import org.bukkit.entity.Player;
+import org.bukkit.inventory.PlayerInventory;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 
@@ -25,6 +33,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     private final String name;
     private final CraftServer server;
     private final WorldNBTStorage storage;
+    private CraftInventoryPlayer inventory;
 
     protected CraftOfflinePlayer(CraftServer server, String name) {
         this.server = server;
@@ -225,4 +234,35 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     public void removeMetadata(String metadataKey, Plugin plugin) {
         server.getPlayerMetadata().removeMetadata(this, metadataKey, plugin);
     }
+
+    public PlayerInventory getInventory() {
+        if (inventory == null) {
+            NBTTagList nbttaglist = getData().getList("Inventory");
+            net.minecraft.server.PlayerInventory inv = new net.minecraft.server.PlayerInventory(null);
+            inv.b(nbttaglist);
+            inventory = new CraftInventoryPlayer(inv);
+        }
+        return inventory;
+    }
+
+    public Location getLocation() {
+        NBTTagCompound data = getData();
+        Location location = null;
+        if (data.hasKey("Pos") && data.hasKey("Rotation")) {
+            NBTTagList pos = data.getList("Pos");
+            NBTTagList rot = data.getList("Rotation");
+            double locX = ((NBTTagDouble) pos.get(0)).data;
+            double locY = ((NBTTagDouble) pos.get(1)).data;
+            double locZ = ((NBTTagDouble) pos.get(2)).data;
+            float yaw = ((NBTTagFloat) rot.get(0)).data;
+            float pitch = ((NBTTagFloat) rot.get(1)).data;
+            World world = null;
+            if (data.hasKey("WorldUUIDMost") && data.hasKey("WorldUUIDLeast")) {
+                UUID uid = new UUID(data.getLong("WorldUUIDMost"), data.getLong("WorldUUIDLeast"));
+                world = server.getWorld(uid);
+            }
+            location = new Location(world, locX, locY, locZ, yaw, pitch);
+        }
+        return location;
+    }
 }
-- 
1.7.9.6 (Apple Git-31.1)

