From dc5c35a09f15ae2e74cd32db91c96a16e16015c3 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Jan 2013 14:12:03 -0500
Subject: [PATCH] Add onGround API


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index bc246b9..acb2336 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -43,6 +43,7 @@ import org.bukkit.event.player.PlayerInteractEntityEvent;
 import org.bukkit.event.player.PlayerItemHeldEvent;
 import org.bukkit.event.player.PlayerKickEvent;
 import org.bukkit.event.player.PlayerMoveEvent;
+import org.bukkit.event.player.PlayerOnGroundEvent;
 import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.event.player.PlayerToggleFlightEvent;
 import org.bukkit.event.player.PlayerToggleSneakEvent;
@@ -196,6 +197,7 @@ public class PlayerConnection extends Connection {
             Player player = this.getPlayer();
             Location from = new Location(player.getWorld(), lastPosX, lastPosY, lastPosZ, lastYaw, lastPitch); // Get the Players previous Event location.
             Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
+            boolean wasOnGround = this.player.onGround;
 
             // If the packet contains movement information then we update the To location with the correct XYZ.
             if (packet10flying.hasPos && !(packet10flying.hasPos && packet10flying.y == -999.0D && packet10flying.stance == -999.0D)) {
@@ -415,6 +417,14 @@ public class PlayerConnection extends Connection {
 
                 this.player.onGround = packet10flying.g;
                 this.minecraftServer.getPlayerList().d(this.player);
+
+                // CraftBukkit start
+                if (wasOnGround != this.player.onGround) {
+                    PlayerOnGroundEvent onGroundEvent = new PlayerOnGroundEvent(player, this.player.onGround);
+                    this.server.getPluginManager().callEvent(onGroundEvent);
+                }
+                // CraftBukkit end
+
                 if (this.player.playerInteractManager.isCreative()) return; // CraftBukkit - fixed fall distance accumulating while being in Creative mode.
                 this.player.b(this.player.locY - d0, packet10flying.g);
             } else if (this.e % 20 == 0) {
-- 
1.7.12.4 (Apple Git-37)

