From ca54e3d39f5081e1a056bed3d7db457a9606fed5 Mon Sep 17 00:00:00 2001
From: agentk20 <agentkid20@gmail.com>
Date: Sat, 3 Aug 2013 19:28:48 +1000
Subject: [PATCH] Fully Disable Snooper When Not Required


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f3d4c45..ea93849 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -673,11 +673,11 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.h[this.ticks % 100] = System.nanoTime() - i;
         this.methodProfiler.b();
         this.methodProfiler.a("snooper");
-        if (!this.m.d() && this.ticks > 100) {
+        if (getSnooperEnabled() && !this.m.d() && this.ticks > 100) { // SportBukkit - Disable snooper
             this.m.a();
         }
 
-        if (this.ticks % 6000 == 0) {
+        if (getSnooperEnabled() && this.ticks % 6000 == 0) { // SportBukkit - Disable snooper
             this.m.b();
         }
 
-- 
1.9.0

