From daeb7725366d8ff59fe061ed9788004332c07bc2 Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sun, 26 May 2013 18:43:56 -0700
Subject: [PATCH] Add PlayerSpawnEntityEvent


diff --git a/src/main/java/net/minecraft/server/ItemMonsterEgg.java b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
index 0adb4e4..5f5c3ff 100644
--- a/src/main/java/net/minecraft/server/ItemMonsterEgg.java
+++ b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
@@ -1,5 +1,11 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerSpawnEntityEvent;
+// CraftBukkit end
+
 public class ItemMonsterEgg extends Item {
 
     public ItemMonsterEgg() {
@@ -34,7 +40,29 @@ public class ItemMonsterEgg extends Item {
                 d0 = 0.5D;
             }
 
-            Entity entity = a(world, itemstack.getData(), (double) i + 0.5D, (double) j + d0, (double) k + 0.5D);
+            // CraftBukkit start
+            CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack);
+            Entity entity = EntityTypes.a(itemstack.getData(), world);
+
+            PlayerSpawnEntityEvent event = new PlayerSpawnEntityEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), (double) i + 0.5D, (double) j + d0, (double) k + 0.5D, MathHelper.g(world.random.nextFloat() * 360.0F), 0.0F), craftItem.clone());
+            world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                event.getEntity().remove();
+                return false;
+            }
+
+            if (entity != null && entity instanceof EntityLiving) {
+                EntityInsentient entityinsentient = (EntityInsentient) entity;
+
+                entity.setPositionRotation(event.getLocation().getX(), event.getLocation().getY(), event.getLocation().getZ(), event.getLocation().getYaw(), event.getLocation().getPitch());
+                entityinsentient.aO = entityinsentient.yaw;
+                entityinsentient.aM = entityinsentient.yaw;
+                entityinsentient.a((GroupDataEntity) null);
+                world.addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+                entityinsentient.r();
+            }
+            // CraftBukkit end
 
             if (entity != null) {
                 if (entity instanceof EntityLiving && itemstack.hasName()) {
-- 
1.8.5.1

