From 39badcf140859faf5a8f26dd8f75a491a6874b78 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 25 Nov 2012 13:56:45 -0600
Subject: [PATCH] Remove the entity destroy queue


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 78dc3c4..bcf73e1 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -37,7 +37,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public double d;
     public double e;
     public final List chunkCoordIntPairQueue = new LinkedList();
-    public final List removeQueue = new LinkedList(); // CraftBukkit - private -> public
+    // public final List removeQueue = new LinkedList(); CraftBukkit - remove the removeQueue
     private final ServerStatisticManager bO;
     private float bP = Float.MIN_VALUE;
     private float bQ = -1.0E8F;
@@ -192,6 +192,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.activeContainer = this.defaultContainer;
         }
 
+        /* CraftBukkit - remove the removeQueue
         while (!this.removeQueue.isEmpty()) {
             int i = Math.min(this.removeQueue.size(), 127);
             int[] aint = new int[i];
@@ -205,6 +206,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
             this.playerConnection.sendPacket(new PacketPlayOutEntityDestroy(aint));
         }
+        */
 
         if (!this.chunkCoordIntPairQueue.isEmpty()) {
             ArrayList arraylist = new ArrayList();
@@ -908,7 +910,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.lastSentExp = -1;
         this.bQ = -1.0F;
         this.bR = -1;
-        this.removeQueue.addAll(((EntityPlayer) entityhuman).removeQueue);
+        // this.removeQueue.addAll(((EntityPlayer) entityhuman).removeQueue);  CraftBukkit - remove the removeQueue
     }
 
     protected void a(MobEffect mobeffect) {
@@ -1032,6 +1034,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         return this.bO;
     }
 
+    /* CraftBukkit - remove destroy queue
     public void d(Entity entity) {
         if (entity instanceof EntityHuman) {
             this.playerConnection.sendPacket(new PacketPlayOutEntityDestroy(new int[] { entity.getId()}));
@@ -1039,6 +1042,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.removeQueue.add(Integer.valueOf(entity.getId()));
         }
     }
+    */
 
     public long x() {
         return this.bX;
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 7dbbcc6..9aede8b 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -341,18 +341,12 @@ public class EntityTrackerEntry {
     // CraftBukkit end
 
     public void a() {
-        Iterator iterator = this.trackedPlayers.iterator();
-
-        while (iterator.hasNext()) {
-            EntityPlayer entityplayer = (EntityPlayer) iterator.next();
-
-            entityplayer.d(this.tracker);
-        }
+        this.broadcast(new PacketPlayOutEntityDestroy(this.tracker.getId())); // CraftBukkit - remove destroy queue
     }
 
     public void a(EntityPlayer entityplayer) {
         if (this.trackedPlayers.contains(entityplayer)) {
-            entityplayer.d(this.tracker);
+            entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityDestroy(this.tracker.getId())); // CraftBukkit - remove destroy queue
             this.trackedPlayers.remove(entityplayer);
         }
     }
@@ -372,7 +366,7 @@ public class EntityTrackerEntry {
                         }
                     }
 
-                    entityplayer.removeQueue.remove(Integer.valueOf(this.tracker.getId()));
+                    // entityplayer.removeQueue.remove(Integer.valueOf(this.tracker.getId())); CraftBukkit - remove destroy queue
                     // CraftBukkit end
 
                     this.trackedPlayers.add(entityplayer);
@@ -456,7 +450,7 @@ public class EntityTrackerEntry {
                 }
             } else if (this.trackedPlayers.contains(entityplayer)) {
                 this.trackedPlayers.remove(entityplayer);
-                entityplayer.d(this.tracker);
+                entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityDestroy(this.tracker.getId())); // CraftBukkit - remove destroy queue
             }
         }
     }
@@ -578,7 +572,7 @@ public class EntityTrackerEntry {
     public void clear(EntityPlayer entityplayer) {
         if (this.trackedPlayers.contains(entityplayer)) {
             this.trackedPlayers.remove(entityplayer);
-            entityplayer.d(this.tracker);
+            entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityDestroy(this.tracker.getId())); // CraftBukkit - remove destroy queue
         }
     }
 }
-- 
1.8.5.2 (Apple Git-48)

