From 64d44f5ab4beb5c63d46233f6764e7edd1eed6f8 Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sun, 26 May 2013 18:43:56 -0700
Subject: [PATCH] Add PlayerSpawnEntityEvent


diff --git a/src/main/java/net/minecraft/server/ItemMonsterEgg.java b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
index 06967d8..2b2ff26 100644
--- a/src/main/java/net/minecraft/server/ItemMonsterEgg.java
+++ b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
@@ -1,5 +1,11 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerSpawnEntityEvent;
+// CraftBukkit end
+
 public class ItemMonsterEgg extends Item {
 
     public ItemMonsterEgg(int i) {
@@ -34,7 +40,31 @@ public class ItemMonsterEgg extends Item {
                 d0 = 0.5D;
             }
 
-            Entity entity = a(world, itemstack.getData(), (double) i + 0.5D, (double) j + d0, (double) k + 0.5D);
+            // CraftBukkit start
+            CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack);
+            Entity entity = EntityTypes.a(itemstack.getData(), world);
+
+            PlayerSpawnEntityEvent event = new PlayerSpawnEntityEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), (double) i + 0.5D, (double) j + d0, (double) k + 0.5D, MathHelper.g(world.random.nextFloat() * 360.0F), 0.0F), craftItem.clone());
+            world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                event.getEntity().remove();
+                return false;
+            }
+
+            entity.setPositionRotation(event.getLocation().getX(), event.getLocation().getY(), event.getLocation().getZ(), event.getLocation().getYaw(), event.getLocation().getPitch());
+            if (entity instanceof EntityLiving) {
+                EntityLiving entityliving = (EntityLiving) entity;
+                entityliving.aP = entityliving.yaw;
+                entityliving.aN = entityliving.yaw;
+                entityliving.a((EntityInsentient) null);
+            }
+            world.addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+            if (entity instanceof EntityInsentient) {
+                ((EntityInsentient) entity).p();
+
+            }
+            // CraftBukkit end
 
             if (entity != null) {
                 if (entity instanceof EntityLiving && itemstack.hasName()) {
-- 
1.7.9.6 (Apple Git-31.1)

