From 90b846e0f5d40eeec64dc7d47569e3ce1e7fa5df Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Thu, 4 Jul 2013 10:11:46 +1000
Subject: [PATCH] Colour Console Messages


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f673d64..e0574df 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -399,7 +399,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
             if (this.worldServer != null) {
                 i.info("Saving worlds");
                 this.saveChunks(false);
-            
+
                 /* CraftBukkit start - Handled in saveChunks
                 for (int i = 0; i < this.worldServer.length; ++i) {
                     WorldServer worldserver = this.worldServer[i];
@@ -986,7 +986,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
     }
 
     public void sendMessage(IChatBaseComponent ichatbasecomponent) {
-        i.info(ichatbasecomponent.c());
+        this.console.sendMessage(ichatbasecomponent.c()); // CraftBukkit - we want coloured and pretty messages too!
     }
 
     public boolean a(int i, String s) {
-- 
1.8.5.2 (Apple Git-48)

