From f94816463b483ed41e12f53b024713b666349137 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 2 Sep 2012 12:28:34 -0500
Subject: [PATCH] Prevent buffer overflow when writing response data on RCON
 query, fixes BUKKIT-2459


diff --git a/src/main/java/net/minecraft/server/RemoteStatusListener.java b/src/main/java/net/minecraft/server/RemoteStatusListener.java
index 12815ee..451f1e2 100644
--- a/src/main/java/net/minecraft/server/RemoteStatusListener.java
+++ b/src/main/java/net/minecraft/server/RemoteStatusListener.java
@@ -73,7 +73,11 @@ public class RemoteStatusListener extends RemoteConnectionThread {
     }
 
     private void send(byte[] abyte, DatagramPacket datagrampacket) {
-        this.socket.send(new DatagramPacket(abyte, abyte.length, datagrampacket.getSocketAddress()));
+        // CraftBukkit start - catch exceptions
+        try {
+            this.socket.send(new DatagramPacket(abyte, abyte.length, datagrampacket.getSocketAddress()));
+        } catch(Exception e) {}
+        // CraftBukkit end
     }
 
     private boolean parsePacket(DatagramPacket datagrampacket) {
@@ -167,10 +171,12 @@ public class RemoteStatusListener extends RemoteConnectionThread {
             this.cachedReply.write("player_");
             this.cachedReply.write((int) 0);
             String[] astring = this.server.getPlayers();
-            byte b0 = (byte) astring.length;
+            // CraftBukkit start - use int instead of byte to prevent overflow
+            int i0 = astring.length;
 
-            for (byte b1 = (byte) (b0 - 1); b1 >= 0; --b1) {
-                this.cachedReply.write(astring[b1]);
+            for (int i1 = i0 - 1; i1 >= 0; --i1) {
+                // CraftBukkit end
+                this.cachedReply.write(astring[i1]);
             }
 
             this.cachedReply.write((int) 0);
-- 
1.7.9.6 (Apple Git-31.1)

