From 435da92cb6e35d502958d152846ea10aa354a5a1 Mon Sep 17 00:00:00 2001
From: Isaac Moore <rmsy@me.com>
Date: Wed, 25 Dec 2013 20:11:34 -0600
Subject: [PATCH] Add EntityDespawnInVoidEvent


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 754df42..b0d1760 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -25,6 +25,7 @@ import org.bukkit.entity.Hanging;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Vehicle;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityDespawnInVoidEvent;
 import org.bukkit.event.hanging.HangingBreakByEntityEvent;
 import org.bukkit.event.vehicle.VehicleBlockCollisionEvent;
 import org.bukkit.event.vehicle.VehicleEnterEvent;
@@ -437,6 +438,9 @@ public abstract class Entity implements ICommandListener {
     }
 
     protected void Y() {
+        // SportBukkit start - call EntityDespawnInVoidEvent
+        this.world.getServer().getPluginManager().callEvent(new EntityDespawnInVoidEvent(this.bukkitEntity));
+        // SportBukkit end
         this.die();
     }
 
-- 
1.9.0

