From 6723e1d92342dc7552e3af9ea6b7233f331358f4 Mon Sep 17 00:00:00 2001
From: Isaac Moore <rmsy@me.com>
Date: Sun, 5 Jan 2014 20:35:57 -0600
Subject: [PATCH] Allow plugins to access player locale


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 4794521..8bd8105 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -28,7 +28,7 @@ import org.bukkit.inventory.MainHand;
 public class EntityPlayer extends EntityHuman implements ICrafting {
 
     private static final Logger bS = LogManager.getLogger();
-    private String locale = "en_US";
+    public String locale = null; // SportBukkit - make public and default to null
     public PlayerConnection playerConnection;
     public final MinecraftServer server;
     public final PlayerInteractManager playerInteractManager;
@@ -1058,7 +1058,15 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.server.server.getPluginManager().callEvent(event);
         }
         // CraftBukkit end
+        String oldLocale = this.locale; // SportBukkit
         this.locale = packetplayinsettings.a();
+        // SportBukkit start - add PlayerLocaleChangeEvent
+        // Since the field is initialized to null, this event should always fire the first time the packet is received
+        if (!this.locale.equals(oldLocale)) {
+            CraftEventFactory.callPlayerLocaleChangeEvent(this, oldLocale, this.locale);
+        }
+        // SportBukkit end
+
         this.ch = packetplayinsettings.c();
         this.ci = packetplayinsettings.d();
         this.getDataWatcher().set(EntityPlayer.br, Byte.valueOf((byte) packetplayinsettings.e()));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 97cc888..a6498fc 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1443,4 +1443,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public String getHostname() {
         return getHandle().hostname;
     }
+
+    @Override
+    public String getLocale() {
+        final String locale = getHandle().locale;
+        return locale != null ? locale : "en_US";
+    }
+
+    @Override
+    public void setLocale(String locale) {
+        getHandle().locale = locale;
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f8c98b7..1c60b02 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -650,7 +650,7 @@ public class CraftEventFactory {
         if (!event.isCancelled()) {
             state.update(true);
         }
-        
+
         return !event.isCancelled();
     }
 
@@ -1016,4 +1016,11 @@ public class CraftEventFactory {
 		cloud.world.getServer().getPluginManager().callEvent(event);
 		return event;
 	}
+
+    public static PlayerLocaleChangeEvent callPlayerLocaleChangeEvent(EntityHuman who, String oldLocale, String newLocale) {
+        Player player = (Player) who.getBukkitEntity();
+        PlayerLocaleChangeEvent event = new PlayerLocaleChangeEvent(player, oldLocale, newLocale);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
 }
-- 
1.9.0

