From 42aa80d0281c7c1bdbe84397732503060518a70d Mon Sep 17 00:00:00 2001
From: Marcos Vives Del Sol <socram8888@gmail.com>
Date: Sun, 28 Jul 2013 18:04:19 +0200
Subject: [PATCH] Update Furnace block when using Bukkit setBurnTime


diff --git a/src/main/java/net/minecraft/server/TileEntityFurnace.java b/src/main/java/net/minecraft/server/TileEntityFurnace.java
index 8f8771c..f1a83a5 100644
--- a/src/main/java/net/minecraft/server/TileEntityFurnace.java
+++ b/src/main/java/net/minecraft/server/TileEntityFurnace.java
@@ -25,6 +25,7 @@ public class TileEntityFurnace extends TileEntity implements IWorldInventory {
     private int lastTick = MinecraftServer.currentTick;
     private int maxStack = MAX_STACK;
     public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private boolean burningBlock;
 
     public ItemStack[] getContents() {
         return this.items;
@@ -124,6 +125,7 @@ public class TileEntityFurnace extends TileEntity implements IWorldInventory {
         }
 
         this.burnTime = nbttagcompound.getShort("BurnTime");
+        this.burningBlock = this.burnTime > 0;
         this.cookTime = nbttagcompound.getShort("CookTime");
         this.ticksForCurrentFuel = fuelTime(this.items[1]);
         if (nbttagcompound.hasKeyOfType("CustomName", 8)) {
@@ -162,7 +164,6 @@ public class TileEntityFurnace extends TileEntity implements IWorldInventory {
     }
 
     public void h() {
-        boolean flag = this.burnTime > 0;
         boolean flag1 = false;
 
         // CraftBukkit start - Use wall time instead of ticks for cooking
@@ -229,9 +230,10 @@ public class TileEntityFurnace extends TileEntity implements IWorldInventory {
                 */
             }
 
-            if (flag != this.burnTime > 0) {
+            if (burningBlock != this.burnTime > 0) {
                 flag1 = true;
-                BlockFurnace.a(this.burnTime > 0, this.world, this.x, this.y, this.z);
+                burningBlock = this.burnTime > 0;
+                BlockFurnace.a(burningBlock, this.world, this.x, this.y, this.z);
             }
         }
 
-- 
1.8.5.2 (Apple Git-48)

