From 0c3e3ddceb5c7b62412a306c9232799e0e9e9d02 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 25 Nov 2012 13:56:45 -0600
Subject: [PATCH] Remove the entity destroy queue


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index fab36ca..653cd61 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -29,7 +29,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public double d;
     public double e;
     public final List chunkCoordIntPairQueue = new LinkedList();
-    public final List removeQueue = new LinkedList();
+    // public final List removeQueue = new LinkedList(); CraftBukkit - remove the removeQueue
     private float bO = Float.MIN_VALUE;
     private float bP = -1.0E8F;
     private int bQ = -99999999;
@@ -179,6 +179,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         }
         // CraftBukkit end
 
+        /* CraftBukkit - remove the removeQueue
         while (!this.removeQueue.isEmpty()) {
             int i = Math.min(this.removeQueue.size(), 127);
             int[] aint = new int[i];
@@ -192,6 +193,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
             this.playerConnection.sendPacket(new Packet29DestroyEntity(aint));
         }
+        */
 
         if (!this.chunkCoordIntPairQueue.isEmpty()) {
             ArrayList arraylist = new ArrayList();
@@ -779,7 +781,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.lastSentExp = -1;
         this.bP = -1.0F;
         this.bQ = -1;
-        this.removeQueue.addAll(((EntityPlayer) entityhuman).removeQueue);
+        // this.removeQueue.addAll(((EntityPlayer) entityhuman).removeQueue); CraftBukkit - remove the removeQueue
     }
 
     protected void a(MobEffect mobeffect) {
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 712895e..9879831 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -334,18 +334,12 @@ public class EntityTrackerEntry {
     // CraftBukkit end
 
     public void a() {
-        Iterator iterator = this.trackedPlayers.iterator();
-
-        while (iterator.hasNext()) {
-            EntityPlayer entityplayer = (EntityPlayer) iterator.next();
-
-            entityplayer.removeQueue.add(Integer.valueOf(this.tracker.id));
-        }
+        this.broadcast(new Packet29DestroyEntity(this.tracker.id)); // CraftBukkit - remove destroy queue
     }
 
     public void a(EntityPlayer entityplayer) {
         if (this.trackedPlayers.contains(entityplayer)) {
-            entityplayer.removeQueue.add(Integer.valueOf(this.tracker.id));
+            entityplayer.playerConnection.sendPacket(new Packet29DestroyEntity(this.tracker.id)); // CraftBukkit - remove destroy queue
             this.trackedPlayers.remove(entityplayer);
         }
     }
@@ -365,7 +359,7 @@ public class EntityTrackerEntry {
                         }
                     }
 
-                    entityplayer.removeQueue.remove(Integer.valueOf(this.tracker.id));
+                    // entityplayer.removeQueue.remove(Integer.valueOf(this.tracker.id));
                     // CraftBukkit end
 
                     this.trackedPlayers.add(entityplayer);
@@ -443,7 +437,7 @@ public class EntityTrackerEntry {
                 }
             } else if (this.trackedPlayers.contains(entityplayer)) {
                 this.trackedPlayers.remove(entityplayer);
-                entityplayer.removeQueue.add(Integer.valueOf(this.tracker.id));
+                entityplayer.playerConnection.sendPacket(new Packet29DestroyEntity(this.tracker.id));
             }
         }
     }
@@ -565,7 +559,7 @@ public class EntityTrackerEntry {
     public void clear(EntityPlayer entityplayer) {
         if (this.trackedPlayers.contains(entityplayer)) {
             this.trackedPlayers.remove(entityplayer);
-            entityplayer.removeQueue.add(Integer.valueOf(this.tracker.id));
+            entityplayer.playerConnection.sendPacket(new Packet29DestroyEntity(this.tracker.id));
         }
     }
 }
-- 
1.7.9.6 (Apple Git-31.1)

