From 0d7c7b91afac86ce300cd8aa782c91457335d146 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 15 Apr 2014 01:12:54 -0500
Subject: [PATCH] Stop server on failed port binding


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 566c295..1511a3d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -564,7 +564,8 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
                 }
                 // Spigot end
             } else {
-                this.a((CrashReport) null);
+                // CraftBukkit - if init fails, stop the server
+                // this.a((CrashReport) null);
             }
         } catch (Throwable throwable) {
             MinecraftServer.LOGGER.error("Encountered an unexpected exception", throwable);
-- 
1.9.0

