From 2428659dc81bc872204aed739136069e2d37130a Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 21 Jan 2013 18:12:27 -0600
Subject: [PATCH] Remove useless debug message


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index cb2b114..5bf29a2 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -282,7 +282,7 @@ public class EntityItem extends Entity {
 
         if (itemstack == null) {
             if (this.world != null) {
-                this.world.getLogger().severe("Item entity " + this.id + " has no item?!");
+               // this.world.getLogger().severe("Item entity " + this.id + " has no item?!"); CraftBukkit - remove useless debug message
             }
 
             return new ItemStack(Block.STONE);
-- 
1.7.12.4 (Apple Git-37)

