From b83e090642dde8a2f75af0972c82945a9123cd9a Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sun, 10 Nov 2013 12:26:30 -0800
Subject: [PATCH] Add support for opening dropper inventory


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 1709da6..6b57bf0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -219,6 +219,13 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
                 openCustomInventory(inventory, player, "minecraft:hopper");
             }
             break;
+        case DROPPER:
+            if (craftinv.getInventory() instanceof TileEntityDropper) {
+                getHandle().openTileEntity((TileEntityDropper) craftinv.getInventory());
+            } else {
+                openCustomInventory(inventory, player, "minecraft:dropper");
+            }
+            break;
         case BEACON:
             if (craftinv.getInventory() instanceof TileEntityBeacon) {
                 getHandle().openContainer((TileEntityBeacon) craftinv.getInventory());
@@ -268,9 +275,9 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     }
 
     public InventoryView openWorkbench(Location location, boolean force) {
-        if (!force) {
+        if(!force) {
             Block block = location.getBlock();
-            if (block.getType() != Material.WORKBENCH) {
+            if(block.getType() != Material.WORKBENCH) {
                 return null;
             }
         }
-- 
1.9.0

