From b13838458d913b0227fe8b79e782724bd44b1481 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 17 Apr 2014 19:56:31 -0500
Subject: [PATCH] Make ticksFlown in EntityFireworks public


diff --git a/src/main/java/net/minecraft/server/EntityFireworks.java b/src/main/java/net/minecraft/server/EntityFireworks.java
index d8fb786..f83f67b 100644
--- a/src/main/java/net/minecraft/server/EntityFireworks.java
+++ b/src/main/java/net/minecraft/server/EntityFireworks.java
@@ -3,7 +3,7 @@ package net.minecraft.server;
 public class EntityFireworks extends Entity {
 
     public static final DataWatcherObject<ItemStack> FIREWORK_ITEM = DataWatcher.a(EntityFireworks.class, DataWatcherRegistry.f);
-    private int ticksFlown;
+    public int ticksFlown; // SportBukkit - private -> public
     public int expectedLifespan;
 
     public EntityFireworks(World world) {
-- 
1.9.0

