From 944f2a0293f4e6f21380eebabfe206d0c2a6bd87 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 10 Nov 2012 12:12:54 -0600
Subject: [PATCH] Add more whitelisting methods to CraftBukkit


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e5705fb..725dac5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1169,6 +1169,18 @@ public final class CraftServer implements Server {
         playerList.reloadWhitelist();
     }
 
+    public boolean isWhitelisted(OfflinePlayer player) {
+        return playerList.getWhitelisted().contains(player.getName().toLowerCase());
+    }
+
+    public void addWhitelist(OfflinePlayer player) {
+        playerList.addWhitelist(player.getName());
+    }
+
+    public void removeWhitelist(OfflinePlayer player) {
+        playerList.removeWhitelist(player.getName());
+    }
+
     public GameMode getDefaultGameMode() {
         return GameMode.getByValue(console.worlds.get(0).getWorldData().getGameType().a());
     }
-- 
1.8.5.1

