From 44a584477ea9b2f851ff46ad63da607d00dcff15 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 9 May 2014 22:38:54 -0400
Subject: [PATCH] Fix wrong potion particles after respawn


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 5e8bfc1..50a73cd 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -596,6 +596,7 @@ public abstract class EntityLiving extends Entity {
         return false;
     }
 
+    public void clearPotionEffectMetadata() { this.bM(); } // SportBukkit - alias for following method
     protected void bM() {
         this.datawatcher.set(EntityLiving.g, Boolean.valueOf(false));
         this.datawatcher.set(EntityLiving.f, Integer.valueOf(0));
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 02b7526..9b0590c 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1276,6 +1276,14 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.deathTicks = 0;
         this.removeAllEffects();
         this.updateEffects = true;
+
+        // SportBukkit - clear potion metadata now, because new effects might
+        // get added before the update in the tick has a chance to run, and if they
+        // match the old effects, the metadata will never be marked dirty and will
+        // go out of sync with the client.
+        this.clearPotionEffectMetadata();
+        this.setInvisible(false);
+
         this.activeContainer = this.defaultContainer;
         this.killer = null;
         this.lastDamager = null;
-- 
1.9.0

