From 1e6e28bd5ececdf8180c5ba89f894b9524c1c900 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 17 Apr 2014 19:52:45 -0500
Subject: [PATCH] Make a bunch of packet fields public


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutAttachEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutAttachEntity.java
index a4dd6ea..b900086 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutAttachEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutAttachEntity.java
@@ -4,9 +4,10 @@ import java.io.IOException;
 
 public class PacketPlayOutAttachEntity implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private int b;
-    private int c;
+    // CraftBukkit - private -> public
+    public int a;
+    public int b;
+    public int c;
 
     public PacketPlayOutAttachEntity() {}
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntityHeadRotation.java b/src/main/java/net/minecraft/server/PacketPlayOutEntityHeadRotation.java
index b8daa72..77d49ca 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntityHeadRotation.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntityHeadRotation.java
@@ -4,8 +4,9 @@ import java.io.IOException;
 
 public class PacketPlayOutEntityHeadRotation implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private byte b;
+    // CraftBukkit - private -> public
+    public int a;
+    public byte b;
 
     public PacketPlayOutEntityHeadRotation() {}
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
index a4626cc..bd2d71e 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
@@ -4,17 +4,18 @@ import java.io.IOException;
 
 public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private int b;
-    private int c;
-    private int d;
-    private int e;
-    private int f;
-    private int g;
-    private int h;
-    private int i;
-    private int j;
-    private int k;
+    // CraftBukkit - private -> public
+    public int a;
+    public int b;
+    public int c;
+    public int d;
+    public int e;
+    public int f;
+    public int g;
+    public int h;
+    public int i;
+    public int j;
+    public int k;
 
     public PacketPlayOutSpawnEntity() {}
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
index 89499ea..9d19be5 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -5,19 +5,21 @@ import java.util.List;
 
 public class PacketPlayOutSpawnEntityLiving implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private int b;
-    private int c;
-    private int d;
-    private int e;
-    private int f;
-    private int g;
-    private int h;
-    private byte i;
-    private byte j;
-    private byte k;
-    private DataWatcher l;
-    private List<DataWatcher.WatchableObject> m;
+    // SportBukkit start - private -> public
+    public int a;
+    public int b;
+    public int c;
+    public int d;
+    public int e;
+    public int f;
+    public int g;
+    public int h;
+    public byte i;
+    public byte j;
+    public byte k;
+    public DataWatcher l;
+    public List<DataWatcher.WatchableObject> m;
+    // SportBukkit end
 
     public PacketPlayOutSpawnEntityLiving() {}
 
-- 
1.9.0

