From cd537c123a853d7584a301eb998f13248ad5df99 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 17 Apr 2014 19:56:31 -0500
Subject: [PATCH] Make ticksFlown in EntityFireworks public


diff --git a/src/main/java/net/minecraft/server/EntityFireworks.java b/src/main/java/net/minecraft/server/EntityFireworks.java
index aa36bf2..81104e5 100644
--- a/src/main/java/net/minecraft/server/EntityFireworks.java
+++ b/src/main/java/net/minecraft/server/EntityFireworks.java
@@ -2,7 +2,8 @@ package net.minecraft.server;
 
 public class EntityFireworks extends Entity {
 
-    private int ticksFlown;
+    // CraftBukkit - private -> public
+    public int ticksFlown;
     public int expectedLifespan;
 
     public EntityFireworks(World world) {
-- 
1.9.0

