From cb4fa11f87bdb130d19d1793efb618c19fa92a9b Mon Sep 17 00:00:00 2001
From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sun, 10 Nov 2013 12:26:30 -0800
Subject: [PATCH] Add support for opening dropper inventory


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 806fbed..2717a81 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -10,6 +10,7 @@ import net.minecraft.server.PacketPlayInCloseWindow;
 import net.minecraft.server.PacketPlayOutOpenWindow;
 import net.minecraft.server.TileEntityBrewingStand;
 import net.minecraft.server.TileEntityDispenser;
+import net.minecraft.server.TileEntityDropper;
 import net.minecraft.server.TileEntityFurnace;
 import net.minecraft.server.TileEntityHopper;
 
@@ -228,6 +229,13 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
                 openCustomInventory(inventory, player, 9);
             }
             break;
+        case DROPPER:
+            if (craftinv.getInventory() instanceof TileEntityDropper) {
+                getHandle().openDispenser((TileEntityDropper) craftinv.getInventory());
+            } else {
+                openCustomInventory(inventory, player, 3);
+            }
+            break;
         case CREATIVE:
         case CRAFTING:
             throw new IllegalArgumentException("Can't open a " + type + " inventory!");
-- 
1.8.5.2 (Apple Git-48)

