From 3426cd2643f2f2ae7b532cb9905813d3dfebe2d8 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 9 Jul 2013 17:06:59 -0500
Subject: [PATCH] Fix invalid positions kicking players


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 82bb796..4db5cdb 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -216,6 +216,13 @@ public class PlayerConnection extends Connection {
             double delta = Math.pow(this.lastPosX - to.getX(), 2) + Math.pow(this.lastPosY - to.getY(), 2) + Math.pow(this.lastPosZ - to.getZ(), 2);
             float deltaAngle = Math.abs(this.lastYaw - to.getYaw()) + Math.abs(this.lastPitch - to.getPitch());
 
+            if (Double.isNaN(packet10flying.x) || Double.isNaN(packet10flying.y) || Double.isNaN(packet10flying.z) || Double.isNaN(packet10flying.stance) || (packet10flying.hasLook && (Double.isNaN(packet10flying.yaw) || Double.isNaN(packet10flying.pitch)))) {
+                player.teleport(to, PlayerTeleportEvent.TeleportCause.UNKNOWN);
+
+                System.err.println(player.getName() + " somehow got to an invalid position, so we teleported them to their previous position");
+                return;
+            }
+
             if ((delta > 1f / 256 || deltaAngle > 10f) && (this.checkMovement && !this.player.dead)) {
                 this.lastPosX = to.getX();
                 this.lastPosY = to.getY();
@@ -251,13 +258,6 @@ public class PlayerConnection extends Connection {
                 }
             }
 
-            if (Double.isNaN(packet10flying.x) || Double.isNaN(packet10flying.y) || Double.isNaN(packet10flying.z) || Double.isNaN(packet10flying.stance)) {
-                player.teleport(player.getWorld().getSpawnLocation(), PlayerTeleportEvent.TeleportCause.UNKNOWN);
-                System.err.println(player.getName() + " was caught trying to crash the server with an invalid position.");
-                player.kickPlayer("Nope!");
-                return;
-            }
-
             if (this.checkMovement && !this.player.dead) {
                 // CraftBukkit end
                 double d1;
@@ -466,6 +466,18 @@ public class PlayerConnection extends Connection {
             f1 = 0;
         }
 
+        if(Double.isNaN(d0)) {
+            d0 = this.y;
+        }
+
+        if(Double.isNaN(d1)) {
+            d1 = this.z;
+        }
+
+        if(Double.isNaN(d2)) {
+            d2 = this.p;
+        }
+
         this.lastPosX = d0;
         this.lastPosY = d1;
         this.lastPosZ = d2;
-- 
1.7.12.4 (Apple Git-37)

