From f72497f98f0f209c73e6cfdbc06098692c221fbe Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 1 Sep 2014 15:37:16 -0500
Subject: [PATCH] Remove this shit


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index acced03..20b9fea 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -125,7 +125,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         DedicatedServer.LOGGER.info("Loading properties");
         this.propertyManager = new PropertyManager(this.options); // CraftBukkit - CLI argument support
         this.q = new EULA(new File("eula.txt"));
-        if (!this.q.a()) {
+        if (false && !this.q.a()) { // SportBukkit - disable check
             DedicatedServer.LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
             this.q.b();
             return false;
-- 
1.9.0

