From 4680791e0c757a33d7946a4ff576498ac657476a Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Wed, 10 Jul 2013 23:46:09 +1000
Subject: [PATCH] Client Crash Cannot Send Message


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 8637ef9..bd149a5 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -854,7 +854,7 @@ public class PlayerConnection extends Connection {
 
                 // CraftBukkit start
                 if (this.player.getChatFlags() == 1 && !s.startsWith("/")) {
-                    this.sendPacket(new Packet3Chat("Cannot send chat message."));
+                    this.player.sendMessage(ChatMessage.d("Cannot send chat message.")); // Spigot - fix client crash
                     return;
                 }
 
-- 
1.7.9.6 (Apple Git-31.1)

