From 4d2e6a9a0f1f09bb90338d70242efcb8a5057b10 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 17 Apr 2014 19:56:31 -0500
Subject: [PATCH] Make ticksFlown in EntityFireworks public


diff --git a/src/main/java/net/minecraft/server/EntityFireworks.java b/src/main/java/net/minecraft/server/EntityFireworks.java
index 759d46c..ecce081 100644
--- a/src/main/java/net/minecraft/server/EntityFireworks.java
+++ b/src/main/java/net/minecraft/server/EntityFireworks.java
@@ -2,8 +2,9 @@ package net.minecraft.server;
 
 public class EntityFireworks extends Entity {
 
-    private int ticksFlown;
-    public int expectedLifespan; // CraftBukkit - private -> public
+    // CraftBukkit - private -> public
+    public int ticksFlown;
+    public int expectedLifespan;
 
     public EntityFireworks(World world) {
         super(world);
-- 
1.8.5.2 (Apple Git-48)

