From 25f771adf6e70fbebc38731b350a2a68dc108fab Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 9 May 2014 22:38:54 -0400
Subject: [PATCH] Fix wrong potion particles after respawn


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index de45a92..3b03166 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1117,6 +1117,15 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.deathTicks = 0;
         this.removeAllEffects();
         this.updateEffects = true;
+
+        // Clear potion metadata now, because new effects might get added
+        // before the update in the tick has a chance to run, and if they
+        // match the old effects, the metadata will never be marked dirty
+        // and will go out of sync with the client.
+        this.datawatcher.watch(8, Byte.valueOf((byte) 0));
+        this.datawatcher.watch(7, Integer.valueOf(0));
+        this.setInvisible(false);
+
         this.activeContainer = this.defaultContainer;
         this.killer = null;
         this.lastDamager = null;
-- 
1.8.5.2 (Apple Git-48)

