From 84d7243602e76c668005be9d17e10e93822e7243 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 9 May 2014 23:48:01 -0400
Subject: [PATCH] Add API to disable potion particles


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index bd9d07e..04d985c 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -14,6 +14,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
 import org.bukkit.event.entity.EntityRegainHealthEvent;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 // CraftBukkit end
 
 public abstract class EntityLiving extends Entity {
@@ -463,7 +464,7 @@ public abstract class EntityLiving extends Entity {
                     this.datawatcher.watch(7, Integer.valueOf(0));
                     this.setInvisible(false);
                 } else {
-                    i = PotionBrewer.a(this.effects.values());
+                    i = ((CraftLivingEntity) this.getBukkitEntity()).getPotionParticles() ? PotionBrewer.a(this.effects.values()) : 0;
                     this.datawatcher.watch(8, Byte.valueOf((byte) (PotionBrewer.b(this.effects.values()) ? 1 : 0)));
                     this.datawatcher.watch(7, Integer.valueOf(i));
                     this.setInvisible(this.hasEffect(MobEffectList.INVISIBILITY.id));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index e678bd8..806116a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -63,6 +63,7 @@ import org.bukkit.util.Vector;
 
 public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     private CraftEntityEquipment equipment;
+    private boolean potionParticles = true;
 
     public CraftLivingEntity(final CraftServer server, final EntityLiving entity) {
         super(server, entity);
@@ -294,6 +295,21 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return effects;
     }
 
+    @Override
+    public void setPotionParticles(boolean enabled) {
+        this.potionParticles = enabled;
+        if(enabled) {
+            this.getHandle().updateEffects = true;
+        } else {
+            this.getHandle().getDataWatcher().watch(7, (int) 0); // particle color
+            this.getHandle().getDataWatcher().watch(8, (byte) 0); // ambient flag
+        }
+    }
+
+    public boolean getPotionParticles() {
+        return this.potionParticles;
+    }
+
     public <T extends Projectile> T launchProjectile(Class<? extends T> projectile) {
         return launchProjectile(projectile, null);
     }
-- 
1.8.5.2 (Apple Git-48)

