From b205b038f3a640e5c7f79e2b34e07e72b4764238 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 12 May 2014 01:55:46 -0400
Subject: [PATCH] Fix spurious bucket empty events


diff --git a/src/main/java/net/minecraft/server/ItemBucket.java b/src/main/java/net/minecraft/server/ItemBucket.java
index a58d4e7..402f9b1 100644
--- a/src/main/java/net/minecraft/server/ItemBucket.java
+++ b/src/main/java/net/minecraft/server/ItemBucket.java
@@ -108,16 +108,18 @@ public class ItemBucket extends Item {
                     }
 
                     // CraftBukkit start
-                    PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(entityhuman, clickedX, clickedY, clickedZ, movingobjectposition.face, itemstack);
+                    // Check that the bucket can be emptied before firing the event
+                    if (world.isEmpty(i, j, k) || !world.getType(i, j, k).getMaterial().isBuildable()) {
+                        PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(entityhuman, clickedX, clickedY, clickedZ, movingobjectposition.face, itemstack);
+                        if(event.isCancelled()) {
+                            return itemstack;
+                        }
 
-                    if (event.isCancelled()) {
-                        return itemstack;
+                        if (this.a(world, i, j, k) && !entityhuman.abilities.canInstantlyBuild) {
+                            return CraftItemStack.asNMSCopy(event.getItemStack());
+                        }
                     }
                     // CraftBukkit end
-
-                    if (this.a(world, i, j, k) && !entityhuman.abilities.canInstantlyBuild) {
-                        return CraftItemStack.asNMSCopy(event.getItemStack()); // CraftBukkit
-                    }
                 }
             }
 
-- 
1.8.5.2 (Apple Git-48)

