From e97828367e3fcdb919f990d85dcbcfea6e84f09e Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 12 Feb 2015 22:38:42 -0500
Subject: [PATCH] Player list header/footer API


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
index 5f5b382..34d0783 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
@@ -4,6 +4,8 @@ import java.io.IOException;
 
 public class PacketPlayOutPlayerListHeaderFooter implements Packet<PacketListenerPlayOut> {
 
+    public net.md_5.bungee.api.chat.BaseComponent[] header, footer; // SportBukkit
+
     private IChatBaseComponent a;
     private IChatBaseComponent b;
 
@@ -19,8 +21,19 @@ public class PacketPlayOutPlayerListHeaderFooter implements Packet<PacketListene
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.a(this.a);
-        packetdataserializer.a(this.b);
+        // SportBukkit start
+        if(this.header != null) {
+            packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(this.header));
+        } else {
+            packetdataserializer.a(this.a);
+        }
+
+        if(this.footer != null) {
+            packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(this.footer));
+        } else {
+            packetdataserializer.a(this.b);
+        }
+        // SportBukkit end
     }
 
     public void a(PacketListenerPlayOut packetlistenerplayout) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index c3839e4..bc43323 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1731,4 +1731,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         packet.components = components;
         getHandle().playerConnection.sendPacket(packet);
     }
+
+    @Override
+    public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
+        PacketPlayOutPlayerListHeaderFooter packet = new PacketPlayOutPlayerListHeaderFooter();
+        packet.header = header;
+        packet.footer = footer;
+        getHandle().playerConnection.sendPacket(packet);
+    }
+
+    @Override
+    public void setPlayerListHeaderFooter(BaseComponent header, BaseComponent footer) {
+        this.setPlayerListHeaderFooter(header == null ? null : new BaseComponent[]{ header },
+                                       footer == null ? null : new BaseComponent[]{ footer });
+    }
 }
-- 
1.9.0

