From e8d967dd0ac653aba7f29f5d007516d71df12b67 Mon Sep 17 00:00:00 2001
From: Marcos Vives Del Sol <socram8888@gmail.com>
Date: Sun, 11 Aug 2013 20:55:40 +0200
Subject: [PATCH] Call BlockFadeEvent for Locked Chest decay


diff --git a/src/main/java/net/minecraft/server/BlockLockedChest.java b/src/main/java/net/minecraft/server/BlockLockedChest.java
index 7aa9b1d..65c9c22 100644
--- a/src/main/java/net/minecraft/server/BlockLockedChest.java
+++ b/src/main/java/net/minecraft/server/BlockLockedChest.java
@@ -13,6 +13,9 @@ public class BlockLockedChest extends Block {
     }
 
     public void a(World world, int i, int j, int k, Random random) {
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world.getWorld().getBlockAt(i, j, k), 0).isCancelled()) {
+            return;
+        }
         world.setAir(i, j, k);
     }
 }
-- 
1.7.12.4 (Apple Git-37)

