From 8d3625babbb6fb3df84590d16429879b3bf43b7f Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 18 May 2014 15:09:16 -0400
Subject: [PATCH] Add knockback reductio


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 04d985c..05300fb 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -818,9 +818,16 @@ public abstract class EntityLiving extends Entity {
             float f1 = MathHelper.sqrt(d0 * d0 + d1 * d1);
             float f2 = 0.4F;
 
-            this.motX /= 2.0D;
-            this.motY /= 2.0D;
-            this.motZ /= 2.0D;
+            // CraftBukkit start
+            double knockbackReduction = this.getBukkitEntity().getKnockbackReduction();
+            double friction = 2.0d - knockbackReduction;
+            f2 *= (1d - knockbackReduction);
+
+            this.motX /= friction;
+            this.motY /= friction;
+            this.motZ /= friction;
+            // CraftBukkit end
+
             this.motX -= d0 / (double) f1 * (double) f2;
             this.motY += (double) f2;
             this.motZ -= d1 / (double) f1 * (double) f2;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index b8a9a14..00cf3b8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -21,6 +21,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     protected final CraftServer server;
     protected Entity entity;
     private EntityDamageEvent lastDamageEvent;
+    private float knockbackReduction;
 
     public CraftEntity(final CraftServer server, final Entity entity) {
         this.server = server;
@@ -185,6 +186,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         entity.velocityChanged = true;
     }
 
+    @Override
+    public void setKnockbackReduction(float n) {
+        this.knockbackReduction = n;
+    }
+
+    @Override
+    public float getKnockbackReduction() {
+        return this.knockbackReduction;
+    }
+
     public boolean isOnGround() {
         if (entity instanceof EntityArrow) {
             return ((EntityArrow) entity).isInGround();
-- 
1.8.5.2 (Apple Git-48)

