From cebe31d85b3d3f63d6ba30b8f5a884296c294013 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 11 Feb 2014 22:12:50 -0600
Subject: [PATCH] Stop server on failed port binding


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 3936588..39c1098 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -415,7 +415,8 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
                 }
                 // Spigot end
             } else {
-                this.a((CrashReport) null);
+                // CraftBukkit - if init fails, stop the server
+                // this.a((CrashReport) null);
             }
         } catch (Throwable throwable) {
             throwable.printStackTrace();
-- 
1.8.5.1

