From 9a18bcb370d04b1072f9b8246cf23dcabdd1f960 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 9 Jul 2014 21:35:52 -0500
Subject: [PATCH] Fix denied lily pad and water bucket interact events not
 displaying the denied result to the client


diff --git a/src/main/java/net/minecraft/server/Item.java b/src/main/java/net/minecraft/server/Item.java
index 4382f1a..18bdb82 100644
--- a/src/main/java/net/minecraft/server/Item.java
+++ b/src/main/java/net/minecraft/server/Item.java
@@ -455,7 +455,13 @@ public class Item {
         return this.getMaxStackSize() == 1 && this.usesDurability();
     }
 
+    // CraftBukkit start - make method public and static
     protected MovingObjectPosition a(World world, EntityHuman entityhuman, boolean flag) {
+        return rayTrace(world, entityhuman, flag);
+    }
+
+    public static MovingObjectPosition rayTrace(World world, EntityHuman entityhuman, boolean flag) {
+        // CraftBukkit end
         float f = 1.0F;
         float f1 = entityhuman.lastPitch + (entityhuman.pitch - entityhuman.lastPitch) * f;
         float f2 = entityhuman.lastYaw + (entityhuman.yaw - entityhuman.lastYaw) * f;
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 5a80716..2f92cb5 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -635,23 +635,48 @@ public class PlayerConnection implements PacketPlayInListener {
         int k = packetplayinblockplace.e();
         int l = packetplayinblockplace.getFace();
 
+        // CraftBukkit - water lilies and empty buckets fire interact events
+        // if the interact event is denied, we need to always make sure to update the player inventory
+        // we also need to update the block they attempted to click
+        boolean lily = false;
+        boolean special = false;
+        // CraftBukkit end
+
         this.player.v();
         if (packetplayinblockplace.getFace() == 255) {
             if (itemstack == null) {
                 return;
             }
 
+            // CraftBukkit - we can assign these variables once we know the itemstack is not null
+            lily = itemstack.getItem() == Item.getItemOf(Blocks.WATER_LILY);
+            special = lily || itemstack.getItem() == Items.BUCKET;
+
             // CraftBukkit start
             int itemstackAmount = itemstack.count;
             org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(this.player, Action.RIGHT_CLICK_AIR, itemstack);
             if (event.useItemInHand() != Event.Result.DENY) {
                 this.player.playerInteractManager.useItem(this.player, this.player.world, itemstack);
+            } else if (special) {
+                MovingObjectPosition movingobjectposition = Item.rayTrace(this.player.world, this.player, true);
+
+                if (movingobjectposition != null) {
+                    flag = true;
+                    i = movingobjectposition.b;
+                    j = movingobjectposition.c;
+                    k = movingobjectposition.d;
+                }
+
+                // for lilies we need to update the block above the block we clicked
+                if (lily) {
+                    j += 1;
+                }
             }
 
             // CraftBukkit - notch decrements the counter by 1 in the above method with food,
             // snowballs and so forth, but he does it in a place that doesn't cause the
             // inventory update packet to get sent
-            always = (itemstack.count != itemstackAmount) || itemstack.getItem() == Item.getItemOf(Blocks.WATER_LILY);
+            always = (itemstack.count != itemstackAmount) || special;
             // CraftBukkit end
         } else if (packetplayinblockplace.d() >= this.minecraftServer.getMaxBuildHeight() - 1 && (packetplayinblockplace.getFace() == 1 || packetplayinblockplace.d() >= this.minecraftServer.getMaxBuildHeight())) {
             ChatMessage chatmessage = new ChatMessage("build.tooHigh", new Object[] { Integer.valueOf(this.minecraftServer.getMaxBuildHeight())});
-- 
1.8.5.2 (Apple Git-48)

