From d3de3ddec933f20d16bed95e0347f3a8d96ef88d Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 20 Feb 2015 15:23:57 -0500
Subject: [PATCH] Fix playEffect params


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 4ac9cac..554ea57 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -864,7 +864,7 @@ public class CraftWorld implements World {
                 particleFullName.append('_').append(data);
             }
 
-            packet = new PacketPlayOutWorldParticles(CraftEffect.getNMSParticle(effect), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, particleCount, radius);
+            packet = new PacketPlayOutWorldParticles(CraftEffect.getNMSParticle(effect), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, speed, particleCount);
         }
 
         int distance;
-- 
1.9.0

