From a7fcc54978e2587012118e698fa5e7e8b67e6bd8 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 12 Feb 2015 04:28:58 -0500
Subject: [PATCH] Expose the fields of several packets


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java b/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
index fe3002d..06a93d5 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
@@ -4,13 +4,15 @@ import java.io.IOException;
 
 public class PacketPlayOutEntityTeleport implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private int b;
-    private int c;
-    private int d;
-    private byte e;
-    private byte f;
-    private boolean g;
+    // CraftBukkit start - make public
+    public int a;
+    public int b;
+    public int c;
+    public int d;
+    public byte e;
+    public byte f;
+    public boolean g;
+    // CraftBukkit end
 
     public PacketPlayOutEntityTeleport() {}
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
index 953d6ef..8c817a9 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
@@ -6,16 +6,18 @@ import java.util.UUID;
 
 public class PacketPlayOutNamedEntitySpawn implements Packet<PacketListenerPlayOut> {
 
-    private int a;
-    private UUID b;
-    private int c;
-    private int d;
-    private int e;
-    private byte f;
-    private byte g;
-    private int h;
-    private DataWatcher i;
-    private List<DataWatcher.WatchableObject> j;
+     // SportBukkit start - make public
+    public int a;
+    public UUID b;
+    public int c;
+    public int d;
+    public int e;
+    public byte f;
+    public byte g;
+    public int h;
+    public DataWatcher i;
+    public List<DataWatcher.WatchableObject> j;
+    // SportBukkit end
 
     public PacketPlayOutNamedEntitySpawn() {}
 
-- 
1.9.0

