From ba068a3f2102cad4a4cbc0cbd58f409e6b7a2832 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 12 Apr 2014 22:42:11 -0400
Subject: [PATCH] Fix BlockFromToEvents firing when flow is obstructed


diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index a98dada..4c8209e 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -137,7 +137,7 @@ public class BlockFlowing extends BlockFluids {
             int index = 0;
 
             for (BlockFace currentFace : faces) {
-                if (aboolean[index]) {
+                if (aboolean[index] && this.o(world, i + currentFace.getModX(), j, k + currentFace.getModZ())) {
                     BlockFromToEvent event = new BlockFromToEvent(source, currentFace);
 
                     if (server != null) {
-- 
1.8.5.2 (Apple Git-48)

