From 578e7c4e61d575dd1585c84a2dbd56746d26fe9a Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Mar 2015 04:29:17 -0400
Subject: [PATCH] Relative teleport API


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 8bf8fd6..010e436 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -672,6 +672,12 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     }
 
     public void a(double d0, double d1, double d2, float f, float f1, Set<PacketPlayOutPosition.EnumPlayerTeleportFlags> set) {
+        // SportBukkit start - add cause parameter
+        this.teleportRelative(d0, d1, d2, f, f1, set, PlayerTeleportEvent.TeleportCause.UNKNOWN);
+    }
+
+    public boolean teleportRelative(double d0, double d1, double d2, float f, float f1, Set<PacketPlayOutPosition.EnumPlayerTeleportFlags> set, PlayerTeleportEvent.TeleportCause cause) {
+        // SportBukkit end
         // CraftBukkit start - Delegate to teleport(Location)
         Player player = this.getPlayer();
         Location from = player.getLocation();
@@ -699,7 +705,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
 
 
         Location to = new Location(this.getPlayer().getWorld(), x, y, z, yaw, pitch);
-        PlayerTeleportEvent event = new PlayerTeleportEvent(player, from.clone(), to.clone(), PlayerTeleportEvent.TeleportCause.UNKNOWN);
+        PlayerTeleportEvent event = new PlayerTeleportEvent(player, from.clone(), to.clone(), cause);
         this.server.getPluginManager().callEvent(event);
 
         if (event.isCancelled() || !to.equals(event.getTo())) {
@@ -713,6 +719,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
         }
 
         this.internalTeleport(d0, d1, d2, f, f1, set);
+        return !event.isCancelled(); // SportBukkit - add return value
     }
 
     public void teleport(Location dest) {
@@ -729,15 +736,15 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
         }
 
         if(Double.isNaN(d0)) {
-            d0 = this.o;
+            d0 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.X) ? 0 : this.o;
         }
 
         if(Double.isNaN(d1)) {
-            d1 = this.p;
+            d1 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.Y) ? 0 : this.p;
         }
 
         if(Double.isNaN(d2)) {
-            d2 = this.q;
+            d2 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.Z) ? 0 : this.q;
         }
 
         this.justTeleported = true;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0bfcf48..06a2f71 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1833,4 +1833,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void hideTitle() {
         getHandle().playerConnection.sendPacket(new PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.CLEAR, (BaseComponent[]) null, 0, 0, 0));
     }
+
+    @Override
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch) {
+        return teleportRelative(deltaPos, deltaYaw, deltaPitch, PlayerTeleportEvent.TeleportCause.PLUGIN);
+    }
+
+    @Override
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch, PlayerTeleportEvent.TeleportCause cause) {
+        return !(getHandle().playerConnection == null || getHandle().playerConnection.isDisconnected()) &&
+               getHandle().playerConnection.teleportRelative(deltaPos.getX(), deltaPos.getY(), deltaPos.getZ(),
+                                                             deltaYaw, deltaPitch,
+                                                             java.util.EnumSet.allOf(PacketPlayOutPosition.EnumPlayerTeleportFlags.class),
+                                                             cause);
+    }
 }
-- 
1.9.0

