From 0a555fbb889265743e2de6b93274ed3784cc9d53 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 11 Aug 2014 10:58:32 -0500
Subject: [PATCH] Fix skull bedrock breaking bug


diff --git a/src/main/java/net/minecraft/server/ItemSkull.java b/src/main/java/net/minecraft/server/ItemSkull.java
index 4a40068..0de7aef 100644
--- a/src/main/java/net/minecraft/server/ItemSkull.java
+++ b/src/main/java/net/minecraft/server/ItemSkull.java
@@ -42,6 +42,11 @@ public class ItemSkull extends Item {
             }
 
             if (!world.isStatic) {
+                // CraftBukkit start
+                if (!Blocks.SKULL.canPlace(world, i, j, k)) {
+                    return false;
+                }
+                // CraftBukkit end
                 world.setTypeAndData(i, j, k, Blocks.SKULL, l, 2);
                 int i1 = 0;
 
-- 
1.8.4.msysgit.0

