From 693b451f842d3f312ea8c676f577fd0943c105b4 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 21 Aug 2014 02:21:45 -0400
Subject: [PATCH] Remove Overmapped and legacy methods


diff --git a/maps.yml b/maps.yml
deleted file mode 100644
index 5384024..0000000
--- a/maps.yml
+++ /dev/null
@@ -1,46 +0,0 @@
-members:
-  "org/bukkit/entity/Damageable _INVALID_damage (I)V": damage
-  "org/bukkit/entity/Damageable _INVALID_damage (ILorg/bukkit/entity/Entity;)V": damage
-  "org/bukkit/entity/Damageable _INVALID_getHealth ()I": getHealth
-  "org/bukkit/entity/Damageable _INVALID_setHealth (I)V": setHealth
-  "org/bukkit/entity/Damageable _INVALID_getMaxHealth ()I": getMaxHealth
-  "org/bukkit/entity/Damageable _INVALID_setMaxHealth (I)V": setMaxHealth
-  "org/bukkit/entity/LivingEntity _INVALID_getLastDamage ()I": getLastDamage
-  "org/bukkit/entity/LivingEntity _INVALID_setLastDamage (I)V": setLastDamage
-  "org/bukkit/event/entity/EntityDamageEvent _INVALID_getDamage ()I": getDamage
-  "org/bukkit/event/entity/EntityDamageEvent _INVALID_setDamage (I)V": setDamage
-  "org/bukkit/event/vehicle/VehicleDamageEvent _INVALID_getDamage ()I": getDamage
-  "org/bukkit/event/vehicle/VehicleDamageEvent _INVALID_setDamage (I)V": setDamage
-  "org/bukkit/event/entity/EntityRegainHealthEvent _INVALID_getAmount ()I": getAmount
-  "org/bukkit/event/entity/EntityRegainHealthEvent _INVALID_setAmount (I)V": setAmount
-  "org/bukkit/entity/Minecart _INVALID_getDamage ()I": getDamage
-  "org/bukkit/entity/Minecart _INVALID_setDamage (I)V": setDamage
-  "org/bukkit/entity/Projectile _INVALID_getShooter ()Lorg/bukkit/entity/LivingEntity;": getShooter
-  "org/bukkit/entity/Projectile _INVALID_setShooter (Lorg/bukkit/entity/LivingEntity;)V": setShooter
-  "org/bukkit/Bukkit _INVALID_getOnlinePlayers ()[Lorg/bukkit/entity/Player;": getOnlinePlayers
-  "org/bukkit/Server _INVALID_getOnlinePlayers ()[Lorg/bukkit/entity/Player;": getOnlinePlayers
-flags:
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity getHealth ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart getHealth ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity getMaxHealth ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart getMaxHealth ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity setHealth (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart setHealth (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity setMaxHealth (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart setMaxHealth (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity damage (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart damage (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity damage (ILorg/bukkit/entity/Entity;)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftEnderDragonPart damage (ILorg/bukkit/entity/Entity;)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity getLastDamage ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftLivingEntity setLastDamage (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftMinecart setDamage (I)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftMinecart getDamage ()I": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftProjectile getShooter ()Lorg/bukkit/entity/LivingEntity;": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftProjectile setShooter (Lorg/bukkit/entity/LivingEntity;)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftArrow getShooter ()Lorg/bukkit/entity/LivingEntity;": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftArrow setShooter (Lorg/bukkit/entity/LivingEntity;)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftFireball getShooter ()Lorg/bukkit/entity/LivingEntity;": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftFireball setShooter (Lorg/bukkit/entity/LivingEntity;)V": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftFish getShooter ()Lorg/bukkit/entity/LivingEntity;": 0x1001
-  "org/bukkit/craftbukkit/v${minecraft_version}/entity/CraftFish setShooter (Lorg/bukkit/entity/LivingEntity;)V": 0x1001
diff --git a/pom.xml b/pom.xml
index f80ca4d..90aa870 100644
--- a/pom.xml
+++ b/pom.xml
@@ -319,24 +319,6 @@
           </execution>
         </executions>
       </plugin>
-      <plugin>
-        <groupId>com.wolvereness</groupId>
-        <artifactId>overmapped</artifactId>
-        <version>0.0.2</version>
-        <executions>
-          <execution>
-            <phase>package</phase>
-            <goals>
-              <goal>map</goal>
-            </goals>
-            <configuration>
-              <maps>${basedir}/target/maps.yml</maps>
-              <input>${basedir}/target/${project.artifactId}-${project.version}.jar</input>
-              <original>${basedir}/target/unmapped-${project.artifactId}-${project.version}.jar</original>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
     </plugins>
     <pluginManagement>
       <plugins>
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6499fa9..7f2a5f1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -489,13 +489,6 @@ public final class CraftServer implements Server {
     }
 
     @Override
-    @Deprecated
-    @SuppressWarnings("unchecked")
-    public Player[] _INVALID_getOnlinePlayers() {
-        return getOnlinePlayers().toArray(EMPTY_PLAYER_ARRAY);
-    }
-
-    @Override
     public List<CraftPlayer> getOnlinePlayers() {
         return this.playerView;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
index 4f8d47d..18a1c2d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
@@ -58,17 +58,4 @@ public class CraftArrow extends AbstractProjectile implements Arrow {
     public EntityType getType() {
         return EntityType.ARROW;
     }
-
-    @Deprecated
-    public LivingEntity _INVALID_getShooter() {
-        if (getHandle().shooter == null) {
-            return null;
-        }
-        return (LivingEntity) getHandle().shooter.getBukkitEntity();
-    }
-
-    @Deprecated
-    public void _INVALID_setShooter(LivingEntity shooter) {
-        getHandle().shooter = ((CraftLivingEntity) shooter).getHandle();
-    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
index 736a460..4767ae2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
@@ -54,34 +54,4 @@ public class CraftEnderDragonPart extends CraftComplexPart implements EnderDrago
     public void resetMaxHealth() {
         getParent().resetMaxHealth();
     }
-
-    @Deprecated
-    public void _INVALID_damage(int amount) {
-        damage(amount);
-    }
-
-    @Deprecated
-    public void _INVALID_damage(int amount, Entity source) {
-        damage(amount, source);
-    }
-
-    @Deprecated
-    public int _INVALID_getHealth() {
-        return NumberConversions.ceil(getHealth());
-    }
-
-    @Deprecated
-    public void _INVALID_setHealth(int health) {
-        setHealth(health);
-    }
-
-    @Deprecated
-    public int _INVALID_getMaxHealth() {
-        return NumberConversions.ceil(getMaxHealth());
-    }
-
-    @Deprecated
-    public void _INVALID_setMaxHealth(int health) {
-        setMaxHealth(health);
-    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index 6f0b942..7655540 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -73,17 +73,4 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
     public EntityType getType() {
         return EntityType.UNKNOWN;
     }
-
-    @Deprecated
-    public void _INVALID_setShooter(LivingEntity shooter) {
-        setShooter(shooter);
-    }
-
-    @Deprecated
-    public LivingEntity _INVALID_getShooter() {
-        if (getHandle().shooter != null) {
-            return (LivingEntity) getHandle().shooter.getBukkitEntity();
-        }
-        return null;
-    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
index edb30e7..a48c35d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFish.java
@@ -62,14 +62,4 @@ public class CraftFish extends AbstractProjectile implements Fish {
         Validate.isTrue(chance >= 0 && chance <= 1, "The bite chance must be between 0 and 1.");
         this.biteChance = chance;
     }
-
-    @Deprecated
-    public LivingEntity _INVALID_getShooter() {
-        return (LivingEntity) getShooter();
-    }
-
-    @Deprecated
-    public void _INVALID_setShooter(LivingEntity shooter) {
-        setShooter(shooter);
-    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 806116a..d1bfb11 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -481,46 +481,6 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return true;
     }
 
-    @Deprecated
-    public int _INVALID_getLastDamage() {
-        return NumberConversions.ceil(getLastDamage());
-    }
-
-    @Deprecated
-    public void _INVALID_setLastDamage(int damage) {
-        setLastDamage(damage);
-    }
-
-    @Deprecated
-    public void _INVALID_damage(int amount) {
-        damage(amount);
-    }
-
-    @Deprecated
-    public void _INVALID_damage(int amount, Entity source) {
-        damage(amount, source);
-    }
-
-    @Deprecated
-    public int _INVALID_getHealth() {
-        return NumberConversions.ceil(getHealth());
-    }
-
-    @Deprecated
-    public void _INVALID_setHealth(int health) {
-        setHealth(health);
-    }
-
-    @Deprecated
-    public int _INVALID_getMaxHealth() {
-        return NumberConversions.ceil(getMaxHealth());
-    }
-
-    @Deprecated
-    public void _INVALID_setMaxHealth(int health) {
-        setMaxHealth(health);
-    }
-
     public int getArrowsStuck() {
         return getHandle().aZ();
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java
index daf6bb4..6401fa7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java
@@ -58,14 +58,4 @@ public abstract class CraftMinecart extends CraftVehicle implements Minecart {
     public EntityMinecartAbstract getHandle() {
         return (EntityMinecartAbstract) entity;
     }
-
-    @Deprecated
-    public void _INVALID_setDamage(int damage) {
-        setDamage(damage);
-    }
-
-    @Deprecated
-    public int _INVALID_getDamage() {
-        return NumberConversions.ceil(getDamage());
-    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
index 2f29f2f..629a191 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
@@ -39,24 +39,4 @@ public abstract class CraftProjectile extends AbstractProjectile implements Proj
     public String toString() {
         return "CraftProjectile";
     }
-
-
-    @Deprecated
-    public LivingEntity _INVALID_getShooter() {
-        if (getHandle().shooter == null) {
-            return null;
-        }
-        return (LivingEntity) getHandle().shooter.getBukkitEntity();
-    }
-
-    @Deprecated
-    public void _INVALID_setShooter(LivingEntity shooter) {
-        if (shooter == null) {
-            return;
-        }
-        getHandle().shooter = ((CraftLivingEntity) shooter).getHandle();
-        if (shooter instanceof CraftHumanEntity) {
-            getHandle().shooterName = ((CraftHumanEntity) shooter).getName();
-        }
-    }
 }
-- 
1.8.4.msysgit.0

