From 4e87642097e4da39c86f89d0e1205a0f4d4079e4 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 1 Sep 2014 15:37:16 -0500
Subject: [PATCH] Remove this shit


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 83f2dad..6f1f0d4 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -76,7 +76,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
         i.info("Loading properties");
         this.propertyManager = new PropertyManager(this.options); // CraftBukkit - CLI argument support
         this.n = new EULA(new File("eula.txt"));
-        if (!this.n.a()) {
+        if (false && !this.n.a()) {
             i.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
             this.n.b();
             return false;
-- 
1.8.4.msysgit.0

