From d2cdbd78d05a82cc39203a5bd88616d851312ed9 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 6 Sep 2014 14:51:47 -0400
Subject: [PATCH] Support nameTagVisibility option


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutScoreboardTeam.java b/src/main/java/net/minecraft/server/PacketPlayOutScoreboardTeam.java
index 77be15c..27afcf1 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutScoreboardTeam.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutScoreboardTeam.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.scoreboard.NameTagVisibility; // CraftBukkit
+
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
@@ -15,6 +17,8 @@ public class PacketPlayOutScoreboardTeam extends Packet {
     public int f;
     public int g;
 
+    public NameTagVisibility nameTagVisibility = NameTagVisibility.ALWAYS; // CraftBukkit
+
     public PacketPlayOutScoreboardTeam() {}
 
     public PacketPlayOutScoreboardTeam(ScoreboardTeam scoreboardteam, int i) {
@@ -25,6 +29,7 @@ public class PacketPlayOutScoreboardTeam extends Packet {
             this.c = scoreboardteam.getPrefix();
             this.d = scoreboardteam.getSuffix();
             this.g = scoreboardteam.packOptionData();
+            this.nameTagVisibility = scoreboardteam.nameTagVisibility;
         }
 
         if (i == 0) {
@@ -74,7 +79,7 @@ public class PacketPlayOutScoreboardTeam extends Packet {
             // Spigot start - protocol patch
             if ( packetdataserializer.version >= 16 )
             {
-                packetdataserializer.a( "always" );
+                packetdataserializer.a( this.nameTagVisibility.token() );
                 packetdataserializer.writeByte( EnumChatFormat.WHITE.ordinal() );
             }
             // Spigot end
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index d8cb4da..950019f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -853,6 +853,7 @@ public class PlayerConnection implements PacketPlayInListener {
                         newPacket.e = Arrays.asList(fakeName);
                         newPacket.f = oldPacket.f;
                         newPacket.g = oldPacket.g;
+                        newPacket.nameTagVisibility = oldPacket.nameTagVisibility;
 
                         packet = newPacket;
                     }
@@ -894,6 +895,7 @@ public class PlayerConnection implements PacketPlayInListener {
                         newPacket.e = newNames;
                         newPacket.f = oldPacket.f;
                         newPacket.g = oldPacket.g;
+                        newPacket.nameTagVisibility = oldPacket.nameTagVisibility;
 
                         packet = newPacket;
                     }
diff --git a/src/main/java/net/minecraft/server/ScoreboardTeam.java b/src/main/java/net/minecraft/server/ScoreboardTeam.java
index 31ef831..72f055f 100644
--- a/src/main/java/net/minecraft/server/ScoreboardTeam.java
+++ b/src/main/java/net/minecraft/server/ScoreboardTeam.java
@@ -1,12 +1,17 @@
 package net.minecraft.server;
 
+import org.bukkit.scoreboard.NameTagVisibility; // CraftBukkit
+
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
 public class ScoreboardTeam extends ScoreboardTeamBase {
 
-    private final Scoreboard a;
+    // CraftBukkit start
+    public NameTagVisibility nameTagVisibility = NameTagVisibility.ALWAYS;
+    public final Scoreboard a; // private -> public
+    //CraftBukkit end
     private final String b;
     private final Set c = new HashSet();
     private String d;
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index 59bff0b..915ea7d 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -6,6 +6,7 @@ import org.apache.commons.lang.Validate;
 import org.bukkit.Bukkit;
 import org.bukkit.entity.Player;
 import org.bukkit.OfflinePlayer;
+import org.bukkit.scoreboard.NameTagVisibility;
 import org.bukkit.scoreboard.Team;
 
 import com.google.common.collect.ImmutableSet;
@@ -93,6 +94,19 @@ public final class CraftTeam extends CraftScoreboardComponent implements Team {
         team.setCanSeeFriendlyInvisibles(enabled);
     }
 
+    @Override
+    public NameTagVisibility nameTagVisibility() throws IllegalStateException {
+        checkState();
+        return team.nameTagVisibility;
+    }
+
+    @Override
+    public void setNameTagVisibility(NameTagVisibility visibility) throws IllegalStateException {
+        checkState();
+        team.nameTagVisibility = visibility;
+        team.a.handleTeamChanged(team);
+    }
+
     public Set<OfflinePlayer> getPlayers() throws IllegalStateException {
         Validate.isTrue(false, CraftScoreboard.warningMessage);
 
-- 
1.8.5.2 (Apple Git-48)

