From 649cde02bfd5720a29fb8468561610fffba78d5c Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 18 Feb 2015 06:36:47 -0500
Subject: [PATCH] Include stack traces in packet encoding exceptions


diff --git a/src/main/java/net/minecraft/server/PacketEncoder.java b/src/main/java/net/minecraft/server/PacketEncoder.java
index 5c91610..f9d513a 100644
--- a/src/main/java/net/minecraft/server/PacketEncoder.java
+++ b/src/main/java/net/minecraft/server/PacketEncoder.java
@@ -40,7 +40,7 @@ public class PacketEncoder extends MessageToByteEncoder<Packet> {
 
                 packet.b(packetdataserializer);
             } catch (Throwable throwable) {
-                PacketEncoder.a.error(throwable);
+                PacketEncoder.a.error("Exception serializing packet " + packet, throwable);
             }
 
         }
-- 
1.9.0

