From 45313aaafabb01d35a7fd05fcc104227fc134b1c Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 22 Sep 2014 05:35:35 -0400
Subject: [PATCH] Fix NPE in Player.playEffect


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 5bfbced..880823e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -668,7 +668,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         } else {
             Validate.isTrue(effect.getData() == null, "Wrong kind of data for this effect!");
         }
-        if (data.getClass().equals(MaterialData.class)) {
+        if (data != null && data.getClass().equals(MaterialData.class)) {
             MaterialData materialData = (MaterialData) data;
             Validate.isTrue(!materialData.getItemType().isBlock(), "Material must be block");
             playEffect(loc, effect, materialData.getItemType().getId(), materialData.getData(), 0, 0, 0, 1, 1, 64);
-- 
1.8.5.2 (Apple Git-48)

