From af02cd45d7404213822ff22dcf513174ca403d4e Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Mar 2015 04:29:17 -0400
Subject: [PATCH] Relative teleport API


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index e246636..332e4e9 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -515,23 +515,16 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
         }
 
         if(Double.isNaN(d0)) {
-            d0 = this.o;
+            d0 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.X) ? 0 : this.o;
         }
 
         if(Double.isNaN(d1)) {
-            d1 = this.p;
+            d1 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.Y) ? 0 : this.p;
         }
 
         if(Double.isNaN(d2)) {
-            d2 = this.q;
+            d2 = set.contains(PacketPlayOutPosition.EnumPlayerTeleportFlags.Z) ? 0 : this.q;
         }
-
-        this.lastPosX = d0;
-        this.lastPosY = d1;
-        this.lastPosZ = d2;
-        this.lastYaw = f;
-        this.lastPitch = f1;
-        this.justTeleported = true;
         // CraftBukkit end
         this.checkMovement = false;
         this.o = d0;
@@ -560,6 +553,15 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             f3 = f1 + this.player.pitch;
         }
 
+        // CraftBukkit start
+        this.lastPosX = this.o;
+        this.lastPosY = this.p;
+        this.lastPosZ = this.q;
+        this.lastYaw = f2;
+        this.lastPitch = f3;
+        this.justTeleported = true;
+        // CraftBukkit end
+
         this.player.setLocation(this.o, this.p, this.q, f2, f3);
         this.player.playerConnection.sendPacket(new PacketPlayOutPosition(d0, d1, d2, f, f1, set));
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0ed5b73..4049e36 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1771,4 +1771,35 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             this.getHandle().playerConnection.sendPacket(new PacketPlayOutEntityVelocity(this.getHandle()));
         }
     }
+
+    @Override
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch) {
+        return teleportRelative(deltaPos, deltaYaw, deltaPitch, PlayerTeleportEvent.TeleportCause.PLUGIN);
+    }
+
+    @Override
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch, PlayerTeleportEvent.TeleportCause cause) {
+        if (getHandle().playerConnection == null || getHandle().playerConnection.isDisconnected()) {
+           return false;
+        }
+
+        // From = Players current Location
+        Location from = this.getLocation();
+        // To = Players new Location if Teleport is Successful
+        Location to = from.clone().add(deltaPos);
+        to.setYaw(to.getYaw() + deltaYaw);
+        to.setPitch(to.getPitch() + deltaPitch);
+
+        // Create & Call the Teleport Event.
+        PlayerTeleportEvent event = new PlayerTeleportEvent(this, from, to, cause);
+        server.getPluginManager().callEvent(event);
+
+        // Return False to inform the Plugin that the Teleport was unsuccessful/cancelled.
+        if (event.isCancelled()) {
+            return false;
+        }
+
+        getHandle().playerConnection.teleport(deltaPos.toLocation(getWorld(), deltaYaw, deltaPitch), java.util.EnumSet.allOf(PacketPlayOutPosition.EnumPlayerTeleportFlags.class));
+        return true;
+    }
 }
-- 
1.9.0

