From 117b591256c59a583c426542f10ab79315740095 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 16 Oct 2014 22:46:43 -0400
Subject: [PATCH] Replace formerly invisible iron door blocks with barriers for
 1.8 clients


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutBlockChange.java b/src/main/java/net/minecraft/server/PacketPlayOutBlockChange.java
index 25f45a0..0cc42a5 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutBlockChange.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutBlockChange.java
@@ -8,6 +8,8 @@ public class PacketPlayOutBlockChange extends Packet {
     public Block block; // CraftBukkit - public
     public int data; // CraftBukkit - public
 
+    public boolean barrier; // CraftBukkit - true if the block should be changed to barrier for 1.8+ clients
+
     public PacketPlayOutBlockChange() {}
 
     public PacketPlayOutBlockChange(int i, int j, int k, World world) {
@@ -16,6 +18,11 @@ public class PacketPlayOutBlockChange extends Packet {
         this.c = k;
         this.block = world.getType(i, j, k);
         this.data = world.getData(i, j, k);
+
+        // CraftBukkit start - change iron door blocks to barrier if the other half is missing
+        this.barrier = this.block == Blocks.IRON_DOOR_BLOCK &&
+                       world.getType(i, j + ((this.data & 0x8) == 0 ? 1 : -1), k) != Blocks.IRON_DOOR_BLOCK;
+        // CraftBukkit end
     }
 
     public void a(PacketDataSerializer packetdataserializer) {
@@ -40,6 +47,10 @@ public class PacketPlayOutBlockChange extends Packet {
             packetdataserializer.writePosition( a, b, c );
             int id = Block.getId( this.block );
             data = org.spigotmc.SpigotBlockDebreakifier.getCorrectedData( id, data );
+            if(this.barrier) {
+                id = 166; // barrier block
+                data = 0;
+            }
             packetdataserializer.b( (id << 4) | this.data );
         }
         // Spigot end
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
index 0ed277e..b133a34 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
@@ -152,6 +152,10 @@ public class PacketPlayOutMapChunk extends Packet {
                 }
             }
         } else {
+            int belowIds[] = new int[256];
+            int belowData[] = new int[256];
+            final int ironDoorId = Block.getId(Blocks.IRON_DOOR_BLOCK);
+
             for ( l = 0; l < achunksection.length; ++l )
             {
                 if ( achunksection[ l ] != null && ( !flag || !achunksection[ l ].isEmpty() ) && ( i & 1 << l ) != 0 )
@@ -172,9 +176,28 @@ public class PacketPlayOutMapChunk extends Packet {
                         {
                             data = org.spigotmc.SpigotBlockDebreakifier.getCorrectedData( id, data );
                         }
+
+                        int layerIndex = ind & 0xff;
+                        if(id == ironDoorId) {
+                            if((data & 0x8) != 0 && belowIds[layerIndex] != ironDoorId) {
+                                // Current block is an iron door top and block below it was not
+                                // an iron door, so replace this block with a barrier (166)
+                                id = 166;
+                                data = 0;
+                            }
+                        } else if(belowIds[layerIndex] == ironDoorId && (belowData[layerIndex] & 0x8) == 0) {
+                            // Current block is not an iron door and block below it was an
+                            // iron door bottom, so replace the block below with a barrier (166)
+                            abyte[j - 512] = (byte) ((166 << 4) & 0xff);
+                            abyte[j - 511] = (byte) ((166 >> 4) & 0xff);
+                        }
+
                         char val = (char) ( id << 4 | data );
                         abyte[ j++ ] = (byte) ( val & 0xFF );
                         abyte[ j++ ] = (byte) ( ( val >> 8 ) & 0xFF );
+
+                        belowIds[layerIndex] = id;
+                        belowData[layerIndex] = data;
                     }
                 }
             }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMultiBlockChange.java b/src/main/java/net/minecraft/server/PacketPlayOutMultiBlockChange.java
index 0e3419f..3a87da6 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutMultiBlockChange.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMultiBlockChange.java
@@ -43,11 +43,18 @@ public class PacketPlayOutMultiBlockChange extends Packet {
                 int j1 = ashort[k] & 255;
 
                 dataoutputstream.writeShort(ashort[k]);
-                int blockId = Block.getId( chunk.getType( l, j1, i1 ) );
+                Block blockType = chunk.getType( l, j1, i1 );
+                int blockId = Block.getId( blockType );
                 int data = chunk.getData( l, j1, i1 );
                 data = org.spigotmc.SpigotBlockDebreakifier.getCorrectedData( blockId, data );
                 int id = ( blockId & 4095 ) << 4 | data & 15;
                 dataoutputstream.writeShort((short) id);
+
+                // change iron door blocks to barrier if the other half is missing
+                if(blockType == Blocks.IRON_DOOR_BLOCK && chunk.getType(l, j1 + ((data & 0x8) == 0 ? 1 : -1), i1) != Blocks.IRON_DOOR_BLOCK) {
+                    id = 166 << 4; // barrier block
+                }
+
                 blocks[k] = id;
             }
             // Spigot end
-- 
1.8.5.2 (Apple Git-48)

