From c3cb736fbe48eee0e63771e1ce953f806c5fc951 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 13 Nov 2014 14:02:49 -0500
Subject: [PATCH] Fix Ender Pearl null shooter crash


diff --git a/src/main/java/net/minecraft/server/EntityEnderPearl.java b/src/main/java/net/minecraft/server/EntityEnderPearl.java
index 6cf6c5f..14fcc6e 100644
--- a/src/main/java/net/minecraft/server/EntityEnderPearl.java
+++ b/src/main/java/net/minecraft/server/EntityEnderPearl.java
@@ -44,14 +44,14 @@ public class EntityEnderPearl extends EntityProjectile {
                     Bukkit.getPluginManager().callEvent(teleEvent);
 
                     if (!teleEvent.isCancelled() && !entityplayer.playerConnection.isDisconnected()) {
-                        if (this.getShooter().am()) {
-                            this.getShooter().mount((Entity) null);
+                        if (entityplayer.am()) {
+                            entityplayer.mount((Entity) null);
                         }
 
                         entityplayer.playerConnection.teleport(teleEvent.getTo());
-                        this.getShooter().fallDistance = 0.0F;
+                        entityplayer.fallDistance = 0.0F;
                         CraftEventFactory.entityDamage = this;
-                        this.getShooter().damageEntity(DamageSource.FALL, 5.0F);
+                        entityplayer.damageEntity(DamageSource.FALL, 5.0F);
                         CraftEventFactory.entityDamage = null;
                     }
                     // CraftBukkit end
-- 
1.9.0

