From 9f8386ebc04cfb0166467a6bfaf623aac33c76d4 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 15 Nov 2014 16:17:49 -0500
Subject: [PATCH] Add PlayerAttackEntityEvent


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 6baae0c..aaeabbf 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -49,6 +49,7 @@ import org.bukkit.event.inventory.InventoryCreativeEvent;
 import org.bukkit.event.inventory.InventoryType.SlotType;
 import org.bukkit.event.player.AsyncPlayerChatEvent;
 import org.bukkit.event.player.PlayerAnimationEvent;
+import org.bukkit.event.player.PlayerAttackEntityEvent;
 import org.bukkit.event.player.PlayerChatEvent;
 import org.bukkit.event.player.PlayerCommandPreprocessEvent;
 import org.bukkit.event.player.PlayerInteractEntityEvent;
@@ -1308,6 +1309,12 @@ public class PlayerConnection implements PacketPlayInListener {
                         return;
                     }
 
+                    // CraftBukkit start
+                    PlayerAttackEntityEvent event = new PlayerAttackEntityEvent(this.getPlayer(), entity.getBukkitEntity());
+                    this.server.getPluginManager().callEvent(event);
+                    if(event.isCancelled()) return;
+                    // CraftBukkit end
+
                     this.player.attack(entity);
 
                     // CraftBukkit start
-- 
1.9.0

