From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 8 Sep 2015 15:49:30 -0400
Subject: [PATCH] Plugin isolation


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -633,6 +633,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public boolean getIsolatePlugins() {
+        return this.configuration.getBoolean("settings.isolate-plugins", false);
+    }
+
+    @Override
     public int getTicksPerAnimalSpawns() {
         return this.configuration.getInt("ticks-per.animal-spawns");
     }
