From 71d379548c9966f038f1c9d1cdc4dce924a7334a Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 17 Dec 2014 17:26:31 -0600
Subject: [PATCH] Allow access to ItemStack handle


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 704be69..b09dbf1 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -86,7 +86,10 @@ public final class CraftItemStack extends ItemStack {
         return new CraftItemStack(CraftMagicNumbers.getMaterial(item), amount, (short) 0, null);
     }
 
-    net.minecraft.server.ItemStack handle;
+    public net.minecraft.server.ItemStack handle;
+    public net.minecraft.server.ItemStack getHandle() {
+        return this.handle;
+    }
 
     /**
      * Mirror
-- 
1.8.5.1

