From 396f18be5fa79c65e310834cf2c861b6203c86ae Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 4 Feb 2015 03:39:46 -0500
Subject: [PATCH] Add setting to disable skull fetching


diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 2a50db9..cacf87a 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -72,7 +72,13 @@ public class TileEntitySkull extends TileEntity {
                     Property property = (Property) Iterables.getFirst(gameprofile.getProperties().get("textures"), null);
 
                     if (property == null) {
-                        gameprofile = MinecraftServer.getServer().av().fillProfileProperties(gameprofile, true);
+                        // CraftBukkit start - implement settings.fetch-skulls
+                        if(org.bukkit.Bukkit.getFetchSkulls()) {
+                            gameprofile = MinecraftServer.getServer().av().fillProfileProperties(gameprofile, true);
+                        } else {
+                            org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for \"" + this.j.getName() + "\"");
+                        }
+                        // CraftBukkit end
                     }
 
                     this.j = gameprofile;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index fcce975..8985e6a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -697,6 +697,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public boolean getFetchSkulls() {
+        return this.configuration.getBoolean("settings.fetch-skulls", true);
+    }
+
+    @Override
     public int getTicksPerAnimalSpawns() {
         return this.configuration.getInt("ticks-per.animal-spawns");
     }
diff --git a/src/main/resources/configurations/bukkit.yml b/src/main/resources/configurations/bukkit.yml
index 896a1b9..53668f1 100644
--- a/src/main/resources/configurations/bukkit.yml
+++ b/src/main/resources/configurations/bukkit.yml
@@ -27,6 +27,7 @@ settings:
     deprecated-verbose: default
     shutdown-message: Server closed
     bungeecord: false
+    fetch-skulls: true
 spawn-limits:
     monsters: 70
     animals: 15
-- 
1.9.0

