From d470f923ba7cd9a2492f6edb84e83ac4aea75192 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 8 Sep 2015 15:49:30 -0400
Subject: [PATCH] Plugin isolation


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 4800e47..b033214 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -632,6 +632,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public boolean getIsolatePlugins() {
+        return this.configuration.getBoolean("settings.isolate-plugins", false);
+    }
+
+    @Override
     public int getTicksPerAnimalSpawns() {
         return this.configuration.getInt("ticks-per.animal-spawns");
     }
-- 
1.9.0

