From abd0df2fbc759eed62ab672b2125edb4c7006e77 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 03:45:49 -0500
Subject: [PATCH] Fix invalid items in villager trades


diff --git a/src/main/java/net/minecraft/server/MerchantRecipe.java b/src/main/java/net/minecraft/server/MerchantRecipe.java
index e3a8b1d..dcb9954 100644
--- a/src/main/java/net/minecraft/server/MerchantRecipe.java
+++ b/src/main/java/net/minecraft/server/MerchantRecipe.java
@@ -96,6 +96,16 @@ public class MerchantRecipe {
         NBTTagCompound nbttagcompound2 = nbttagcompound.getCompound("sell");
 
         this.sellingItem = ItemStack.createStack(nbttagcompound2);
+
+        // SportBukkit start - replace invalid items with barrier, to prevent NPE later
+        if(this.buyingItem1 == null) {
+            this.buyingItem1 = new ItemStack(Blocks.BARRIER, 1);
+        }
+        if(this.sellingItem == null) {
+            this.sellingItem = new ItemStack(Blocks.BARRIER, 1);
+        }
+        // SportBukkit end
+
         if (nbttagcompound.hasKeyOfType("buyB", 10)) {
             this.buyingItem2 = ItemStack.createStack(nbttagcompound.getCompound("buyB"));
         }
-- 
1.9.0

