From 2d39f0bc7e1cd92daf53a8ef156ae927c76ca29c Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 10 Jul 2015 23:32:00 -0400
Subject: [PATCH] Fix self-damage being cancelled by FF/PvP rules


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index cb0627d..08cf87f 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -825,6 +825,9 @@ public abstract class EntityHuman extends EntityLiving {
     public boolean a(EntityHuman entityhuman) {
         // CraftBukkit start - Change to check OTHER player's scoreboard team according to API
         // To summarize this method's logic, it's "Can parameter hurt this"
+
+        if(this == entityhuman) return true; // SportBukkit - self-damage is always allowed
+
         org.bukkit.scoreboard.Team team;
         if (entityhuman instanceof EntityPlayer) {
             EntityPlayer thatPlayer = (EntityPlayer) entityhuman;
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index e9d3d91..1ac8740 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -504,6 +504,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     public boolean a(EntityHuman entityhuman) {
+        if(this == entityhuman) return true; // SportBukkit - self-damage is always allowed
         return !this.cr() ? false : super.a(entityhuman);
     }
 
-- 
1.9.0

