From 74485c973891cecad4130e5858810bdb3943fa3b Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 6 Mar 2016 08:48:16 -0500
Subject: [PATCH] Team collision API


diff --git a/src/main/java/net/minecraft/server/ScoreboardTeam.java b/src/main/java/net/minecraft/server/ScoreboardTeam.java
index 71d372f..bc84b17 100644
--- a/src/main/java/net/minecraft/server/ScoreboardTeam.java
+++ b/src/main/java/net/minecraft/server/ScoreboardTeam.java
@@ -116,10 +116,12 @@ public class ScoreboardTeam extends ScoreboardTeamBase {
         this.a.handleTeamChanged(this);
     }
 
+    public ScoreboardTeamBase.EnumTeamPush getCollisionRule() { return this.k(); } // SportBukkit - alias for following method
     public ScoreboardTeamBase.EnumTeamPush k() {
         return this.l;
     }
 
+    public void setCollisionRule(ScoreboardTeamBase.EnumTeamPush rule) { this.a(rule); } // SportBukkit - alias for following method
     public void a(ScoreboardTeamBase.EnumTeamPush scoreboardteambase_enumteampush) {
         this.l = scoreboardteambase_enumteampush;
         this.a.handleTeamChanged(this);
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index 763ddca..074e58c 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -2,10 +2,12 @@ package org.bukkit.craftbukkit.scoreboard;
 
 import java.util.Set;
 
+import net.minecraft.server.ScoreboardTeamBase;
 import net.minecraft.server.ScoreboardTeamBase.EnumNameTagVisibility;
 import org.apache.commons.lang.Validate;
 import org.bukkit.Bukkit;
 import org.bukkit.OfflinePlayer;
+import org.bukkit.scoreboard.CollisionRule;
 import org.bukkit.scoreboard.NameTagVisibility;
 import org.bukkit.scoreboard.Team;
 
@@ -106,6 +108,18 @@ public final class CraftTeam extends CraftScoreboardComponent implements Team {
         team.setNameTagVisibility(bukkitToNotch(visibility));
     }
 
+    @Override
+    public CollisionRule getCollisionRule() {
+        checkState();
+        return CollisionRule.byKey(team.getCollisionRule().e);
+    }
+
+    @Override
+    public void setCollisionRule(CollisionRule rule) {
+        checkState();
+        team.setCollisionRule(ScoreboardTeamBase.EnumTeamPush.a(rule.getKey()));
+    }
+
     public Set<OfflinePlayer> getPlayers() throws IllegalStateException {
         CraftScoreboard scoreboard = checkState();
 
-- 
1.9.0

