From 75390302814f3d28a70bd49de92669bd77847fa0 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 19 Jul 2015 16:51:38 -0400
Subject: [PATCH] Set health before death event


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 07578f7..cf8bce7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -102,11 +102,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
             throw new IllegalArgumentException("Health must be between 0 and " + getMaxHealth());
         }
 
+        getHandle().setHealth((float) health);
+
         if (entity instanceof EntityPlayer && health == 0) {
             ((EntityPlayer) entity).die(DamageSource.GENERIC);
         }
-
-        getHandle().setHealth((float) health);
     }
 
     public double getMaxHealth() {
-- 
1.9.0

