From 6a10b4796cdaba161fcebe479fbe561f315283a8 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index f46fd49..fa41b3e 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -66,6 +66,8 @@ public abstract class EntityHuman extends EntityLiving {
     private final ItemCooldown bW = this.l();
     public EntityFishingHook hookedFish;
 
+    public int clientTicksSinceLastAttack; // SportBukkit
+
     // SportBukkit start - natural regen API
     public boolean fastNaturalRegeneration = true;
     public boolean slowNaturalRegeneration = true;
@@ -1913,11 +1915,12 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     public float o(float f) {
-        return MathHelper.a(((float) this.aF + f) / this.dd(), 0.0F, 1.0F);
+        return MathHelper.a(((float) Math.max(this.aF, this.clientTicksSinceLastAttack) + f) / this.dd(), 0.0F, 1.0F); // SportBukkit - check client and server ticks
     }
 
     public void de() {
         this.aF = 0;
+        this.clientTicksSinceLastAttack = 0; // SportBukkit
     }
 
     public ItemCooldown df() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 9b35fdf..e267557 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -497,6 +497,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.d();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // SportBukkit
+
                 // SportBukkit start - velocity prediction
                 this.player.handleClientTick(
                     packetplayinflying.hasPos ? new Vector(packetplayinflying.x,
-- 
1.9.0

