From 65330f505ed336e02c705fcb3827436d0ab84a38 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 29 Mar 2016 02:13:07 -0400
Subject: [PATCH] Attack cooldown info API


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index fa41b3e..49e9ae4 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -976,7 +976,7 @@ public abstract class EntityHuman extends EntityLiving {
 
                 f *= 0.2F + f2 * f2 * 0.8F;
                 f1 *= f2;
-                this.de();
+                // this.de();   SportBukkit - don't reset cooldown here, let the arm swing packet do it
                 if (f > 0.0F || f1 > 0.0F) {
                     boolean flag = f2 > 0.9F;
                     boolean flag1 = false;
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index ab59c3f..2795be7 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -58,7 +58,7 @@ public abstract class EntityLiving extends Entity {
     public int deathTicks;
     public float aD;
     public float aE;
-    protected int aF;
+    protected int aF; public int ticksSinceLastAttack() { return aF; } // SportBukkit - add accessor
     public float aG;
     public float aH;
     public float aI;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 74fbb3c..9dd6cbf 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -481,4 +481,15 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
     public int getExpToLevel() {
         return getHandle().getExpToLevel();
     }
+
+    @Override
+    public int getAttackCooldownTicks() {
+        return Math.max(getHandle().ticksSinceLastAttack(), getHandle().clientTicksSinceLastAttack);
+    }
+
+    @Override
+    public float getAttackCooldownCoefficient() {
+        final float f = getHandle().o(0.5f);
+        return 0.2f + f * f * 0.8f;
+    }
 }
-- 
1.9.0

