From da0f9b66bf6b12658db662d5e19d53f0ef67243f Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 26 Jun 2016 01:57:17 -0400
Subject: [PATCH] Item cooldown API


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 3528710..fe17a06 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1913,6 +1913,7 @@ public abstract class EntityHuman extends EntityLiving {
         this.clientTicksSinceLastAttack = 0; // SportBukkit
     }
 
+    public ItemCooldown getItemCooldown() { return di(); } // SportBukkit - method alias
     public ItemCooldown di() {
         return this.bV;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0d012f3..b1e72b6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2037,4 +2037,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public int getProtocolVersion() {
         return getHandle().protocolVersion;
     }
+
+    @Override
+    public void startItemCooldown(Material item, int ticks) {
+        getHandle().getItemCooldown().a(CraftMagicNumbers.getItem(item), ticks);
+    }
+
+    @Override
+    public float getRemainingItemCooldown(Material item) {
+        return getHandle().getItemCooldown().a(CraftMagicNumbers.getItem(item), 0F);
+    }
 }
-- 
1.9.0

