From 146ce828ea30954e5bbdc69d8dac01555603ded3 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 26 Jun 2016 01:57:17 -0400
Subject: [PATCH] Item cooldown API


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index fbfe329..b48aaa7 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1908,6 +1908,7 @@ public abstract class EntityHuman extends EntityLiving {
         this.clientTicksSinceLastAttack = 0; // SportBukkit
     }
 
+    public ItemCooldown getItemCooldown() { return di(); } // SportBukkit - method alias
     public ItemCooldown di() {
         return this.bV;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 699017c..820802d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2035,4 +2035,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public int getProtocolVersion() {
         return getHandle().protocolVersion;
     }
+
+    @Override
+    public void startItemCooldown(Material item, int ticks) {
+        getHandle().getItemCooldown().a(CraftMagicNumbers.getItem(item), ticks);
+    }
+
+    @Override
+    public float getRemainingItemCooldown(Material item) {
+        return getHandle().getItemCooldown().a(CraftMagicNumbers.getItem(item), 0F);
+    }
 }
-- 
1.9.0

