From af5d97f97609e772ab6a5b22190c94698e77d2df Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 22 Jun 2016 15:17:26 -0400
Subject: [PATCH] Allow empty ItemStack


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 097bc98..f2b12df 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -150,11 +150,7 @@ public final class CraftItemStack extends ItemStack {
         if (handle == null) {
             return;
         }
-        if (amount == 0) {
-            handle = null;
-        } else {
-            handle.setCount(amount);
-        }
+        handle.setCount(amount);
     }
 
     @Override
-- 
1.9.0

