From 580c16d5d5c01d460d4787bdb26fad4196064537 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 22 Jun 2016 15:17:26 -0400
Subject: [PATCH] Allow empty ItemStack


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index f05d9c0..1b12514 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -154,11 +154,7 @@ public final class CraftItemStack extends ItemStack {
         if (handle == null) {
             return;
         }
-        if (amount == 0) {
-            handle = null;
-        } else {
-            handle.count = amount;
-        }
+        handle.count = amount;
     }
 
     @Override
-- 
1.9.0

