From aae135770f456279b70cf575fc83e67a1dfbffe0 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 23 Nov 2016 04:38:06 -0500
Subject: [PATCH] Fix no knockback from zero-damage projectiles


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 047ca32..a274c3b 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1379,12 +1379,9 @@ public abstract class EntityLiving extends Entity {
                 if (!human) {
                     this.setAbsorptionHearts(this.getAbsorptionHearts() - f);
                 }
-
-                return true;
-            } else {
-                return originalDamage > 0;
-                // CraftBukkit end
             }
+            return true;
+            // CraftBukkit end
         }
         return false; // CraftBukkit
     }
-- 
1.9.0

