From b6ffbf8480b9f7b5e4d4a5d534eed55006c898c4 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 23 Dec 2016 13:36:50 -0500
Subject: [PATCH] Fix NPE riding non-vehicle entity


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index bdcdd19..9e56489 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2698,7 +2698,10 @@ public abstract class Entity implements ICommandListener {
 
     @Nullable
     public Entity bw() {
-        return null;
+        // SportBukkit start
+        final List<Entity> passengers = getPassengers();
+        return passengers.isEmpty() ? null : passengers.get(0);
+        // SportBukkit end
     }
 
     public List<Entity> getPassengers() { return bx(); } // SportBukkit - alias
-- 
1.9.0

