From eeda9863aaf24daf115827d35e5dfb1c0a2fe4fd Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 14 Nov 2012 15:51:08 -0600
Subject: [PATCH] Add support for anvil damage events

---
 .../net/minecraft/server/EntityFallingBlock.java   |   20 +++++++++++++++++++-
 1 files changed, 19 insertions(+), 1 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index dd8b64f..8e74699 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -3,6 +3,11 @@ package net.minecraft.server;
 import java.util.ArrayList;
 import java.util.Iterator;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.Bukkit;
+// CraftBukkit end
+
 public class EntityFallingBlock extends Entity {
 
     public int id;
@@ -124,7 +129,20 @@ public class EntityFallingBlock extends Entity {
                 while (iterator.hasNext()) {
                     Entity entity = (Entity) iterator.next();
 
-                    entity.damageEntity(damagesource, Math.min(MathHelper.d((float) i * this.fallHurtAmount), this.fallHurtMax));
+                    // CraftBukkit start - entity damage event for anvils
+                    int damage = Math.min(MathHelper.d((float) i * this.fallHurtAmount), this.fallHurtMax);
+                    if(this.id == Block.ANVIL.id) {
+                        EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(entity.getBukkitEntity(), this.getBukkitEntity(), EntityDamageByEntityEvent.DamageCause.ANVIL, damage);
+                        Bukkit.getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled()) {
+                            entity.getBukkitEntity().setLastDamageCause(event);
+                            entity.damageEntity(damagesource, event.getDamage());
+                        }
+                    } else {
+                        entity.damageEntity(damagesource, damage);
+                    }
+                    // CraftBukkit end
                 }
 
                 if (this.id == Block.ANVIL.id && (double) this.random.nextFloat() < 0.05000000074505806D + (double) i * 0.05D) {
-- 
1.7.4.4

