From 409278c113fa275a0ea4b3400369e6291714968c Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 25 Nov 2012 13:43:39 -0600
Subject: [PATCH] Add methods for working with arrows stuck in living entities

---
 .../craftbukkit/entity/CraftLivingEntity.java      |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 096de37..331d2f8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -303,4 +303,12 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     public boolean hasLineOfSight(Entity other) {
         return getHandle().aA().canSee(((CraftEntity) other).getHandle()); // az should be getEntitySenses
     }
+
+    public int getArrowsStuck() {
+        return getHandle().bM();
+    }
+
+    public void setArrowsStuck(int arrows) {
+        getHandle().r(arrows);
+    }
 }
--
1.7.4.4

