From e87afd5c92e452cd35ec41831e24d97470039601 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 21 Feb 2013 19:21:29 -0600
Subject: [PATCH] Be less strict with illegal stances, fixes BUKKIT-2578

---
 .../net/minecraft/server/PlayerConnection.java     |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 6b7a5a8..16140a6 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -333,7 +333,7 @@ public class PlayerConnection extends Connection {
                     d2 = packet10flying.y;
                     d3 = packet10flying.z;
                     d4 = packet10flying.stance - packet10flying.y;
-                    if (!this.player.isSleeping() && (d4 > 1.65D || d4 < 0.1D)) {
+                    if (!this.player.isSleeping() && (d4 > 1.65D /** || d4 < 0.1D **/)) { // CraftBukkit - be less strict
                         this.disconnect("Illegal stance");
                         logger.warning(this.player.name + " had an illegal stance: " + d4);
                         return;
-- 
1.7.4.4

