From 4edffa18da6a8e45ef6f099438ddf5ddebe5495a Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 17 Nov 2012 01:24:18 -0600
Subject: [PATCH] Allow placing of ender portals in any dimension

---
 .../net/minecraft/server/BlockEnderPortal.java     |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/main/java/net/minecraft/server/BlockEnderPortal.java b/src/main/java/net/minecraft/server/BlockEnderPortal.java
index d55e41d..10436c8 100644
--- a/src/main/java/net/minecraft/server/BlockEnderPortal.java
+++ b/src/main/java/net/minecraft/server/BlockEnderPortal.java
@@ -49,7 +49,7 @@ public class BlockEnderPortal extends BlockContainer {
     public void onPlace(World world, int i, int j, int k) {
         if (!a) {
             if (world.worldProvider.dimension != 0) {
-                world.setTypeId(i, j, k, 0);
+                // world.setTypeId(i, j, k, 0); // CraftBukkit - allow placing of portals in any dimension
             }
         }
     }
-- 
1.7.4.4

