From b55f597c2185ecb2f81a6196b759d5d59d0fa967 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Tue, 18 Dec 2012 22:12:55 -0600
Subject: [PATCH] Fix record players without a record placed before 1.4.4

---
 .../minecraft/server/TileEntityRecordPlayer.java   |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/src/main/java/net/minecraft/server/TileEntityRecordPlayer.java b/src/main/java/net/minecraft/server/TileEntityRecordPlayer.java
index 4504ae0..d83f545 100644
--- a/src/main/java/net/minecraft/server/TileEntityRecordPlayer.java
+++ b/src/main/java/net/minecraft/server/TileEntityRecordPlayer.java
@@ -11,7 +11,12 @@ public class TileEntityRecordPlayer extends TileEntity {
         if (nbttagcompound.hasKey("RecordItem")) {
             this.record = ItemStack.a(nbttagcompound.getCompound("RecordItem"));
         } else {
-            this.record = new ItemStack(nbttagcompound.getInt("Record"), 1, 0);
+            // CraftBukkit start
+            int id = nbttagcompound.getInt("Record");
+            if(id != 0) {
+                this.record = new ItemStack(id, 1, 0);
+            }
+            // CraftBukkit end
         }
     }

--
1.7.4.4

