--- a/net/minecraft/server/BlockTNT.java
+++ b/net/minecraft/server/BlockTNT.java
@@ -1,6 +1,8 @@
 package net.minecraft.server;
 
 public class BlockTNT extends Block {
+    private boolean isPrimed = false; // SportBukkit
+    Entity primer; // SportBukkit
 
     public static final BlockStateBoolean EXPLODE = BlockStateBoolean.of("explode");
 
@@ -14,7 +16,11 @@
         super.onPlace(world, blockposition, iblockdata);
         if (world.isBlockIndirectlyPowered(blockposition)) {
             this.postBreak(world, blockposition, iblockdata.set(BlockTNT.EXPLODE, Boolean.valueOf(true)));
-            world.setAir(blockposition);
+            // SportBukkit start - don't remove block if TNT priming is cancelled
+            if (isPrimed) {
+                world.setAir(blockposition);
+            }
+            // SportBukkit end
         }
 
     }
@@ -22,7 +28,11 @@
     public void a(IBlockData iblockdata, World world, BlockPosition blockposition, Block block, BlockPosition blockposition1) {
         if (world.isBlockIndirectlyPowered(blockposition)) {
             this.postBreak(world, blockposition, iblockdata.set(BlockTNT.EXPLODE, Boolean.valueOf(true)));
-            world.setAir(blockposition);
+            // SportBukkit start - don't remove block if TNT priming is cancelled
+            if (isPrimed) {
+                world.setAir(blockposition);
+            }
+            // SportBukkit end
         }
 
     }
@@ -31,8 +41,23 @@
         if (!world.isClientSide) {
             EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) blockposition.getX() + 0.5F), (double) blockposition.getY(), (double) ((float) blockposition.getZ() + 0.5F), explosion.getSource());
 
-            entitytntprimed.setFuseTicks((short) (world.random.nextInt(entitytntprimed.getFuseTicks() / 4) + entitytntprimed.getFuseTicks() / 8));
-            world.addEntity(entitytntprimed);
+            // SportBukkit start - allow exploded TNT to cancel priming
+            final org.bukkit.event.entity.ExplosionPrimeEvent event;
+            if(explosion.source != null) {
+                event = new org.bukkit.event.entity.ExplosionPrimeByEntityEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity(), explosion.source.getBukkitEntity());
+            } else {
+                event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+            }
+            world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                isPrimed = true;
+                entitytntprimed.setFuseTicks(world.random.nextInt(entitytntprimed.getFuseTicks() / 4) + entitytntprimed.getFuseTicks() / 8);
+                world.addEntity(entitytntprimed);
+            } else {
+                a(world, blockposition, new ItemStack(this, 1, 0));
+            }
+            // SportBukkit end
         }
     }
 
@@ -45,6 +70,18 @@
             if (((Boolean) iblockdata.get(BlockTNT.EXPLODE)).booleanValue()) {
                 EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) blockposition.getX() + 0.5F), (double) blockposition.getY(), (double) ((float) blockposition.getZ() + 0.5F), entityliving);
 
+                // SportBukkit start - fire ExplosionPrimeEvent
+                org.bukkit.event.entity.ExplosionPrimeEvent event;
+                if(primer != null) {
+                    event = new org.bukkit.event.entity.ExplosionPrimeByEntityEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity(), primer.getBukkitEntity());
+                } else {
+                    event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+                }
+                world.getServer().getPluginManager().callEvent(event);
+                if(event.isCancelled()) return;
+                isPrimed = true;
+                // SportBukkit end
+
                 world.addEntity(entitytntprimed);
                 world.a((EntityHuman) null, entitytntprimed.locX, entitytntprimed.locY, entitytntprimed.locZ, SoundEffects.gV, SoundCategory.BLOCKS, 1.0F, 1.0F);
             }
@@ -56,8 +93,14 @@
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (!itemstack.isEmpty() && (itemstack.getItem() == Items.FLINT_AND_STEEL || itemstack.getItem() == Items.FIRE_CHARGE)) {
+            try { primer = entityhuman; // SportBukkit
             this.a(world, blockposition, iblockdata.set(BlockTNT.EXPLODE, Boolean.valueOf(true)), (EntityLiving) entityhuman);
-            world.setTypeAndData(blockposition, Blocks.AIR.getBlockData(), 11);
+            } finally { primer = null; } // SportBukkit
+            // SportBukkit start - don't remove block if TNT priming is cancelled
+            if (isPrimed) {
+                world.setTypeAndData(blockposition, Blocks.AIR.getBlockData(), 11);
+            }
+            // SportBukkit end
             if (itemstack.getItem() == Items.FLINT_AND_STEEL) {
                 itemstack.damage(1, entityhuman);
             } else if (!entityhuman.abilities.canInstantlyBuild) {
@@ -75,8 +118,19 @@
             EntityArrow entityarrow = (EntityArrow) entity;
 
             if (entityarrow.isBurning()) {
+                // CraftBukkit start
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entityarrow, blockposition, Blocks.AIR, 0).isCancelled()) {
+                    return;
+                }
+                // CraftBukkit end
+                try { primer = entityarrow; // SportBukkit
                 this.a(world, blockposition, world.getType(blockposition).set(BlockTNT.EXPLODE, Boolean.valueOf(true)), entityarrow.shooter instanceof EntityLiving ? (EntityLiving) entityarrow.shooter : null);
-                world.setAir(blockposition);
+                } finally { primer = null; } // SportBukkit
+                // SportBukkit start - don't remove block if TNT priming is cancelled
+                if (isPrimed) {
+                    world.setAir(blockposition);
+                }
+                // SportBukkit end
             }
         }
 
