--- a/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
+++ b/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
@@ -4,6 +4,8 @@
 
 public class PacketPlayOutPlayerListHeaderFooter implements Packet<PacketListenerPlayOut> {
 
+    public net.md_5.bungee.api.chat.BaseComponent[] header, footer; // SportBukkit
+
     private IChatBaseComponent a;
     private IChatBaseComponent b;
 
@@ -19,8 +21,19 @@
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.a(this.a);
-        packetdataserializer.a(this.b);
+        // SportBukkit start
+        if(this.header != null) {
+            packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(this.header));
+        } else {
+            packetdataserializer.a(this.a);
+        }
+
+        if(this.footer != null) {
+            packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(this.footer));
+        } else {
+            packetdataserializer.a(this.b);
+        }
+        // SportBukkit end
     }
 
     public void a(PacketListenerPlayOut packetlistenerplayout) {
