--- a/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
+++ b/net/minecraft/server/PacketPlayOutPlayerListHeaderFooter.java
@@ -2,7 +2,25 @@
 
 import java.io.IOException;
 
-public class PacketPlayOutPlayerListHeaderFooter implements Packet<PacketListenerPlayOut> {
+// SportBukkit start
+import net.md_5.bungee.api.chat.BaseComponent;
+import org.bukkit.craftbukkit.protocol.RenderableComponent;
+import org.bukkit.craftbukkit.protocol.Renderable;
+import org.bukkit.entity.Player;
+
+public class PacketPlayOutPlayerListHeaderFooter implements Packet<PacketListenerPlayOut>, Renderable {
+
+    @Override
+    public void render(Player viewer) {
+        RenderableComponent.render(a, viewer);
+        RenderableComponent.render(b, viewer);
+    }
+
+    public PacketPlayOutPlayerListHeaderFooter(BaseComponent header, BaseComponent footer) {
+        a = new RenderableComponent(header);
+        b = new RenderableComponent(footer);
+    }
+    // SportBukkit end
 
     private IChatBaseComponent a;
     private IChatBaseComponent b;
