--- a/net/minecraft/server/PacketPlayOutTitle.java
+++ b/net/minecraft/server/PacketPlayOutTitle.java
@@ -4,7 +4,23 @@
 import java.util.Locale;
 import javax.annotation.Nullable;
 
-public class PacketPlayOutTitle implements Packet<PacketListenerPlayOut> {
+// SportBukkit start
+import net.md_5.bungee.api.chat.BaseComponent;
+import org.bukkit.craftbukkit.protocol.RenderableComponent;
+import org.bukkit.craftbukkit.protocol.Renderable;
+import org.bukkit.entity.Player;
+
+public class PacketPlayOutTitle implements Packet<PacketListenerPlayOut>, Renderable {
+
+    @Override
+    public void render(Player viewer) {
+        RenderableComponent.render(b, viewer);
+    }
+
+    public PacketPlayOutTitle(EnumTitleAction action, @Nullable BaseComponent text, int fadeIn, int stay, int fadeOut) {
+        this(action, text == null ? null : new RenderableComponent(text), fadeIn, stay, fadeOut);
+    }
+    // SportBukkit end
 
     private PacketPlayOutTitle.EnumTitleAction a;
     private IChatBaseComponent b;
@@ -24,7 +40,7 @@
 
     public PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction packetplayouttitle_enumtitleaction, @Nullable IChatBaseComponent ichatbasecomponent, int i, int j, int k) {
         this.a = packetplayouttitle_enumtitleaction;
-        this.b = ichatbasecomponent;
+        this.b = RenderableComponent.wrap(ichatbasecomponent); // SportBukkit
         this.c = i;
         this.d = j;
         this.e = k;
@@ -45,7 +61,14 @@
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.a((Enum) this.a);
+        // SportBukkit start
+        if(packetdataserializer.protocolVersion < 305 && this.a.ordinal() >= 2) {
+            // ACTIONBAR doesn't exist
+            packetdataserializer.writeVarInt(this.a.ordinal() - 1);
+        } else {
+            packetdataserializer.writeOrdinal(this.a);
+        }
+        // SportBukkit end
         if (this.a == PacketPlayOutTitle.EnumTitleAction.TITLE || this.a == PacketPlayOutTitle.EnumTitleAction.SUBTITLE || this.a == PacketPlayOutTitle.EnumTitleAction.ACTIONBAR) {
             packetdataserializer.a(this.b);
         }
