--- a/net/minecraft/server/PacketPlayOutBoss.java
+++ b/net/minecraft/server/PacketPlayOutBoss.java
@@ -3,7 +3,18 @@
 import java.io.IOException;
 import java.util.UUID;
 
-public class PacketPlayOutBoss implements Packet<PacketListenerPlayOut> {
+// SportBukkit start
+import org.bukkit.craftbukkit.protocol.RenderableComponent;
+import org.bukkit.craftbukkit.protocol.Renderable;
+import org.bukkit.entity.Player;
+
+public class PacketPlayOutBoss implements Packet<PacketListenerPlayOut>, Renderable {
+
+    @Override
+    public void render(Player viewer) {
+        RenderableComponent.render(c, viewer);
+    }
+    // SportBukkit end
 
     private UUID a;
     private PacketPlayOutBoss.Action b;
@@ -20,7 +31,7 @@
     public PacketPlayOutBoss(PacketPlayOutBoss.Action packetplayoutboss_action, BossBattle bossbattle) {
         this.b = packetplayoutboss_action;
         this.a = bossbattle.d();
-        this.c = bossbattle.e();
+        this.c = RenderableComponent.wrap(bossbattle.e()); // SportBukkit
         this.d = bossbattle.getProgress();
         this.e = bossbattle.g();
         this.f = bossbattle.h();
