--- a/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -14,6 +14,20 @@
     private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a;
     private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList();
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction a) {
+        this.a = a;
+    }
+
+    public void add(PacketPlayOutPlayerInfo.PlayerInfoData info) {
+        this.b.add(info);
+    }
+
+    public boolean isEmpty() {
+        return this.b.isEmpty();
+    }
+    // SportBukkit end
+
     public PacketPlayOutPlayerInfo() {}
 
     public PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, EntityPlayer... aentityplayer) {
@@ -130,6 +144,15 @@
 
                 packetdataserializer.d(packetplayoutplayerinfo_playerinfodata.c().getId());
                 packetdataserializer.d(packetplayoutplayerinfo_playerinfodata.b());
+
+                // SportBukkit start
+                if (packetplayoutplayerinfo_playerinfodata.displayName != null) {
+                    packetdataserializer.writeBoolean(true);
+                    packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(packetplayoutplayerinfo_playerinfodata.displayName));
+                    break;
+                }
+                // SportBukkit end
+
                 if (packetplayoutplayerinfo_playerinfodata.d() == null) {
                     packetdataserializer.writeBoolean(false);
                 } else {
@@ -150,6 +173,15 @@
 
             case UPDATE_DISPLAY_NAME:
                 packetdataserializer.a(packetplayoutplayerinfo_playerinfodata.a().getId());
+
+                // SportBukkit start
+                if (packetplayoutplayerinfo_playerinfodata.displayName != null) {
+                    packetdataserializer.writeBoolean(true);
+                    packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(packetplayoutplayerinfo_playerinfodata.displayName));
+                    break;
+                }
+                // SportBukkit end
+
                 if (packetplayoutplayerinfo_playerinfodata.d() == null) {
                     packetdataserializer.writeBoolean(false);
                 } else {
@@ -175,6 +207,8 @@
 
     public class PlayerInfoData {
 
+        public net.md_5.bungee.api.chat.BaseComponent[] displayName; // SportBukkit
+
         private final int b;
         private final EnumGamemode c;
         private final GameProfile d;
