--- a/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -9,7 +9,36 @@
 import java.util.List;
 import javax.annotation.Nullable;
 
-public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
+// SportBukkit start
+import java.util.UUID;
+
+import net.md_5.bungee.api.chat.BaseComponent;
+import org.bukkit.GameMode;
+import org.bukkit.Skin;
+import org.bukkit.craftbukkit.protocol.RenderableComponent;
+import org.bukkit.craftbukkit.protocol.Renderable;
+import org.bukkit.craftbukkit.util.Skins;
+import org.bukkit.entity.Player;
+
+public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut>, Renderable {
+
+    @Override
+    public void render(Player viewer) {
+        b.forEach(info -> RenderableComponent.render(info.e, viewer));
+    }
+
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction a) {
+        this.a = a;
+    }
+
+    public void add(PacketPlayOutPlayerInfo.PlayerInfoData info) {
+        this.b.add(info);
+    }
+
+    public boolean isEmpty() {
+        return this.b.isEmpty();
+    }
+    // SportBukkit end
 
     private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a;
     private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList();
@@ -175,6 +204,21 @@
 
     public class PlayerInfoData {
 
+        // SportBukkit start
+        public PlayerInfoData(UUID uuid, String name, @Nullable BaseComponent displayName, GameMode gamemode, int ping, @Nullable Skin skin) {
+            this(new GameProfile(uuid, name),
+                 ping,
+                 EnumGamemode.getById(gamemode.getValue()),
+                 displayName == null ? null : new RenderableComponent(displayName));
+
+            if(skin != null) {
+                Skins.toProperties(skin).asMap().forEach(
+                    (k, v) -> d.getProperties().putAll(k, v)
+                );
+            }
+        }
+        // SportBukkit end
+
         private final int b;
         private final EnumGamemode c;
         private final GameProfile d;
@@ -184,7 +228,7 @@
             this.d = gameprofile;
             this.b = i;
             this.c = enumgamemode;
-            this.e = ichatbasecomponent;
+            this.e = RenderableComponent.wrap(ichatbasecomponent); // SportBukkit
         }
 
         public GameProfile a() {
