--- a/net/minecraft/server/PacketPlayOutTitle.java
+++ b/net/minecraft/server/PacketPlayOutTitle.java
@@ -12,6 +12,18 @@
     private int d;
     private int e;
 
+    // SportBukkit start
+    public net.md_5.bungee.api.chat.BaseComponent[] components;
+
+    public PacketPlayOutTitle(EnumTitleAction action, net.md_5.bungee.api.chat.BaseComponent[] components, int fadeIn, int stay, int fadeOut) {
+        this.a = action;
+        this.components = components;
+        this.c = fadeIn;
+        this.d = stay;
+        this.e = fadeOut;
+    }
+    // SportBukkit end
+
     public PacketPlayOutTitle() {}
 
     public PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction packetplayouttitle_enumtitleaction, IChatBaseComponent ichatbasecomponent) {
@@ -45,9 +57,22 @@
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.a((Enum) this.a);
+        // SportBukkit start
+        if(packetdataserializer.protocolVersion < 305 && this.a.ordinal() >= 2) {
+            // ACTIONBAR doesn't exist
+            packetdataserializer.writeVarInt(this.a.ordinal() - 1);
+        } else {
+            packetdataserializer.writeOrdinal(this.a);
+        }
+        // SportBukkit end
         if (this.a == PacketPlayOutTitle.EnumTitleAction.TITLE || this.a == PacketPlayOutTitle.EnumTitleAction.SUBTITLE || this.a == PacketPlayOutTitle.EnumTitleAction.ACTIONBAR) {
-            packetdataserializer.a(this.b);
+            // SportBukkit start
+            if(this.components != null) {
+                packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+            } else {
+                packetdataserializer.a(this.b);
+            }
+            // SportBukkit end
         }
 
         if (this.a == PacketPlayOutTitle.EnumTitleAction.TIMES) {
