--- a/net/minecraft/server/SoundEffectType.java
+++ b/net/minecraft/server/SoundEffectType.java
@@ -11,7 +11,7 @@
     public static final SoundEffectType g = new SoundEffectType(1.0F, 1.0F, SoundEffects.ai, SoundEffects.am, SoundEffects.al, SoundEffects.ak, SoundEffects.aj);
     public static final SoundEffectType h = new SoundEffectType(1.0F, 1.0F, SoundEffects.fm, SoundEffects.fq, SoundEffects.fp, SoundEffects.fo, SoundEffects.fn);
     public static final SoundEffectType i = new SoundEffectType(1.0F, 1.0F, SoundEffects.gt, SoundEffects.gx, SoundEffects.gw, SoundEffects.gv, SoundEffects.gu);
-    public static final SoundEffectType j = new SoundEffectType(1.0F, 1.0F, SoundEffects.dl, SoundEffects.dp, SoundEffects.do, SoundEffects.dn, SoundEffects.dm);
+    public static final SoundEffectType j = new SoundEffectType(1.0F, 1.0F, SoundEffects.dl, SoundEffects.dp, SoundEffects.BLOCK_LADDER_PLACE, SoundEffects.dn, SoundEffects.dm); // SportBukkit - fix decomp
     public static final SoundEffectType k = new SoundEffectType(0.3F, 1.0F, SoundEffects.b, SoundEffects.h, SoundEffects.g, SoundEffects.e, SoundEffects.d);
     public static final SoundEffectType l = new SoundEffectType(1.0F, 1.0F, SoundEffects.fY, SoundEffects.gg, SoundEffects.ge, SoundEffects.gb, SoundEffects.ga);
     public final float m;
@@ -40,15 +40,23 @@
         return this.n;
     }
 
+    public SoundEffect stepSound() { return this.d(); } // SportBukkit - alias for following method
     public SoundEffect d() {
         return this.p;
     }
 
+    public SoundEffect placeSound() { return this.e(); } // SportBukkit - alias for following method
     public SoundEffect e() {
         return this.q;
     }
 
+    public SoundEffect fallSound() { return this.g(); } // SportBukkit - alias for following method
     public SoundEffect g() {
         return this.s;
     }
+
+    // SportBukkit start - add missing accessors
+    public SoundEffect breakSound() { return this.o;}
+    public SoundEffect hitSound() { return this.r;}
+    // SportBukkit end
 }
