--- a/net/minecraft/server/TileEntitySkull.java
+++ b/net/minecraft/server/TileEntitySkull.java
@@ -115,7 +115,13 @@
                     Property property = (Property) Iterables.getFirst(gameprofile1.getProperties().get("textures"), (Object) null);
 
                     if (property == null) {
-                        gameprofile1 = TileEntitySkull.k.fillProfileProperties(gameprofile1, true);
+                        // CraftBukkit start - implement settings.fetch-skulls
+                        if(org.bukkit.Bukkit.getFetchSkulls()) {
+                            gameprofile1 = TileEntitySkull.k.fillProfileProperties(gameprofile1, true);
+                        } else {
+                            org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for \"" + gameprofile1.getName() + "\"");
+                        }
+                        // CraftBukkit end
                     }
 
                     return gameprofile1;
