--- a/net/minecraft/server/TileEntitySkull.java
+++ b/net/minecraft/server/TileEntitySkull.java
@@ -106,6 +106,9 @@
         if (gameprofile != null && !UtilColor.b(gameprofile.getName())) {
             if (gameprofile.isComplete() && gameprofile.getProperties().containsKey("textures")) {
                 return gameprofile;
+            } else if(!org.bukkit.Bukkit.getFetchSkulls()) {
+                org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for '" + gameprofile.getName() + "'");
+                return gameprofile;
             } else if (TileEntitySkull.j != null && TileEntitySkull.k != null) {
                 GameProfile gameprofile1 = TileEntitySkull.j.getProfile(gameprofile.getName());
 
