From 03891816f7c4f1a82360125ae01e7f33e6da3877 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 17 Nov 2012 12:42:05 -0600
Subject: [PATCH] Make AsyncPlayerPreLoginEvent more versatile


diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index 1d57188..01c41f1 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -20,17 +20,23 @@ public class AsyncPlayerPreLoginEvent extends Event {
     private final UUID uniqueId;
 
     @Deprecated
-    public AsyncPlayerPreLoginEvent(final String name, final InetAddress ipAddress) {
-        this(name, ipAddress, null);
+    public AsyncPlayerPreLoginEvent(final String name, final InetAddress ipAddress, final boolean allowed) {
+        this(name, ipAddress, null, allowed);
     }
 
-    public AsyncPlayerPreLoginEvent(final String name, final InetAddress ipAddress, final UUID uniqueId) {
+    public AsyncPlayerPreLoginEvent(final String name, final InetAddress ipAddress, final UUID uniqueId, final boolean allowed) {
         super(true);
-        this.result = Result.ALLOWED;
         this.message = "";
         this.name = name;
         this.ipAddress = ipAddress;
         this.uniqueId = uniqueId;
+
+        if(allowed) {
+            this.result = Result.ALLOWED;
+        } else {
+            this.result = Result.KICK_VERIFY;
+            this.message = "Failed to verify username!";
+        }
     }
 
     /**
@@ -189,6 +195,10 @@ public class AsyncPlayerPreLoginEvent extends Event {
          */
         KICK_WHITELIST,
         /**
+         * The player's username was not verified with Mojang
+         */
+        KICK_VERIFY,
+        /**
          * The player is not allowed to log in, for reasons undefined
          */
         KICK_OTHER;
diff --git a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
index e8553f0..aa49dfa 100644
--- a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
@@ -152,6 +152,10 @@ public class PlayerPreLoginEvent extends Event {
          */
         KICK_WHITELIST,
         /**
+         * The player's username was not verified with Mojang
+         */
+        KICK_VERIFY,
+        /**
          * The player is not allowed to log in, for reasons undefined
          */
         KICK_OTHER
-- 
1.8.5.1

