From b7c8909f9e0c9cac40d89b39f532238ec4c72a7d Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 16 Jun 2013 16:57:59 -0500
Subject: [PATCH] Add getting and setting of entity count


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 2c43d9e..433d624 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1047,6 +1047,20 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getEntityCount()
+     */
+    public static int getEntityCount() {
+        return server.getEntityCount();
+    }
+
+    /**
+     * @see Server#setEntityCount()
+     */
+    public static void setEntityCount(int count) {
+        server.setEntityCount(count);
+    }
+
+    /**
      * Gets an instance of the server's default server-icon.
      *
      * @return the default server-icon; null values may be used by the
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ab6d047..ee707ac 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -855,6 +855,20 @@ public interface Server extends PluginMessageRecipient {
     ScoreboardManager getScoreboardManager();
 
     /**
+     * Gets the current entity count
+     *
+     * @return current entity count
+     */
+    int getEntityCount();
+
+    /**
+     * Sets the current entity count
+     *
+     * @param count The new entity count
+     */
+    void setEntityCount(int count);
+
+    /**
      * Gets an instance of the server's default server-icon.
      *
      * @return the default server-icon; null values may be used by the
-- 
1.9.0

