From f28e8211c94dfa737f0e21c96c360c9869928c76 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 16 Jun 2013 16:57:59 -0500
Subject: [PATCH] Add getting and setting of entity count


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 1eaf92d..6116ff4 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -706,6 +706,20 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getEntityCount()
+     */
+    public static int getEntityCount() {
+        return server.getEntityCount();
+    }
+
+    /**
+     * @see Server#setEntityCount()
+     */
+    public static void setEntityCount(int count) {
+        server.setEntityCount(count);
+    }
+
+    /**
      * @see Server#getServerIcon()
      */
     public static CachedServerIcon getServerIcon() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e14e9f1..1fffc38 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -855,6 +855,20 @@ public interface Server extends PluginMessageRecipient {
     ScoreboardManager getScoreboardManager();
 
     /**
+     * Gets the current entity count
+     *
+     * @return current entity count
+     */
+    int getEntityCount();
+
+    /**
+     * Sets the current entity count
+     *
+     * @param count The new entity count
+     */
+    void setEntityCount(int count);
+
+    /**
      * Gets an instance of the server's default server-icon.
      *
      * @return the default server-icon; null values may be used by the
-- 
1.8.5.2 (Apple Git-48)

