From 24ce25471bff1c85540f0468b276c193118da955 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sun, 16 Jun 2013 16:57:59 -0500
Subject: [PATCH] Add getting and setting of entity count


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index dafea2e..9877c2e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -692,6 +692,20 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getEntityCount()
+     */
+    public static int getEntityCount() {
+        return server.getEntityCount();
+    }
+
+    /**
+     * @see Server#setEntityCount()
+     */
+    public static void setEntityCount(int count) {
+        server.setEntityCount(count);
+    }
+
+    /**
      * @see Server#getServerIcon()
      */
     public static CachedServerIcon getServerIcon() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 3c4d541..3408a90 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -815,6 +815,20 @@ public interface Server extends PluginMessageRecipient {
     ScoreboardManager getScoreboardManager();
 
     /**
+     * Gets the current entity count
+     *
+     * @return current entity count
+     */
+    int getEntityCount();
+
+    /**
+     * Sets the current entity count
+     *
+     * @param count The new entity count
+     */
+    void setEntityCount(int count);
+
+    /**
      * Gets an instance of the server's default server-icon.
      *
      * @return the default server-icon; null values may be used by the
-- 
1.8.5.1

