From 4ef4b82925c5b693d26d8131196021e36cfd384b Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 16 May 2013 17:37:16 -0500
Subject: [PATCH] Add Ignorant Effects API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 5c5d526..7a36dbb 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1,6 +1,7 @@
 package org.bukkit.entity;
 
 import java.net.InetSocketAddress;
+import java.util.Set;
 
 import org.bukkit.Achievement;
 import org.bukkit.ChatColor;
@@ -17,6 +18,7 @@ import org.bukkit.command.CommandSender;
 import org.bukkit.conversations.Conversable;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
+import org.bukkit.potion.PotionEffectType;
 import org.bukkit.scoreboard.Scoreboard;
 
 /**
@@ -667,4 +669,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * @param scale if the client health display is scaled
      */
     public void setScaleHealth(boolean scale);
-}
+
+    public Set<PotionEffectType> getIgnorantEffects();
+
+    public void clearIgnorantEffects();
+
+    public void addIgnorantEffect(PotionEffectType effect);
+
+    public void removeIgnorantEffect(PotionEffectType effect);
+
+    public boolean hasIgnorantEffect(PotionEffectType effect);
+ }
-- 
1.7.12.4 (Apple Git-37)

