From 33e87082a18a6ba4b135e27391c0c0c58713a87e Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 16 May 2013 17:37:16 -0500
Subject: [PATCH] Add Ignorant Effects API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3ec374b..fa37170 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1,6 +1,7 @@
 package org.bukkit.entity;
 
 import java.net.InetSocketAddress;
+import java.util.Set;
 
 import org.bukkit.Achievement;
 import org.bukkit.ChatColor;
@@ -17,6 +18,7 @@ import org.bukkit.command.CommandSender;
 import org.bukkit.conversations.Conversable;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
+import org.bukkit.potion.PotionEffectType;
 import org.bukkit.scoreboard.Scoreboard;
 
 /**
@@ -721,4 +723,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * @see Player#setHealthScaled(boolean)
      */
     public double getHealthScale();
-}
+
+    public Set<PotionEffectType> getIgnorantEffects();
+
+    public void clearIgnorantEffects();
+
+    public void addIgnorantEffect(PotionEffectType effect);
+
+    public void removeIgnorantEffect(PotionEffectType effect);
+
+    public boolean hasIgnorantEffect(PotionEffectType effect);
+ }
-- 
1.7.12.4 (Apple Git-37)

