From 03c5d51b9730f97a75ba7e58230967b18bda9823 Mon Sep 17 00:00:00 2001
From: Chad Waters <authorblues@gmail.com>
Date: Wed, 6 Feb 2013 14:42:01 -0500
Subject: [PATCH] Add safe methods from @md-5's offlineUtil


diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index 013dc0a..d344184 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -3,6 +3,7 @@ package org.bukkit;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.entity.AnimalTamer;
 import org.bukkit.entity.Player;
+import org.bukkit.inventory.PlayerInventory;
 import org.bukkit.permissions.ServerOperator;
 
 public interface OfflinePlayer extends ServerOperator, AnimalTamer, ConfigurationSerializable {
@@ -93,4 +94,18 @@ public interface OfflinePlayer extends ServerOperator, AnimalTamer, Configuratio
      */
     public Location getBedSpawnLocation();
 
+    /**
+     * Get the player's inventory. This inventory will not save upon modification, it must be saved using the saveInventory() method.
+     * This inventory will also have a null owner.
+     *
+     * @return The inventory of the player, this also contains the armor slots.
+     */
+    public PlayerInventory getInventory();
+
+     /**
+     * Gets the last Location where the player was before he logged out of the server.
+     *
+     * @return Last Location of the player.
+     */
+    public Location getLocation();
 }
-- 
1.7.9.6 (Apple Git-31.1)

