From b3b4899adf5adb0e2fb165de8ceb931683c98162 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 28 Apr 2012 01:39:04 -0400
Subject: [PATCH] Add API to control if a player can push other entities


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 95b3c57..d5f8b38 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -882,6 +882,18 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     public void setFlying(boolean value);
 
     /**
+     * Change whether this player can push other entities
+     * @param yes true to push other entities, which is the default
+     */
+    public void setCollidesWithEntities(boolean yes);
+
+    /**
+     * Whether this player can push other entities
+     * @return true if the player can push other entities
+     */
+    public boolean getCollidesWithEntities();
+
+    /**
      * Sets the speed at which a client will fly. Negative values indicate
      * reverse directions.
      *
-- 
1.9.0

