From 17cdf28c83d39bd74bd764875e70bc9aac9247bc Mon Sep 17 00:00:00 2001
From: Steve Anton <anxuiz.nx@gmail.com>
Date: Sat, 4 Aug 2012 10:33:42 -0700
Subject: [PATCH] Add the overhead name API for setting the name above a
 player's avatar in game.


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 2716738..4f9cb1e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -74,6 +74,27 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     public void setPlayerListName(String name);
 
     /**
+     * Gets the name that is shown above a player's avatar in game.
+     *
+     * @return the player's overhead name
+     */
+    public String getOverheadName();
+
+    /**
+     * Sets the name that is shown above a player's avatar in game.
+     * <p />
+     * The name cannot be longer than 16 characters, but {@link ChatColor} is supported.
+     * <p />
+     * If the value is null, the name will be identical to {@link #getName()}.
+     * <p />
+     * TODO: add more info
+     *
+     * @param name new overhead name
+     * @throws IllegalArgumentException if the length of the name is too long
+     */
+    public void setOverheadName(String name);
+
+    /**
      * Set the target of the player's compass.
      *
      * @param loc Location to point to
-- 
1.7.9.6 (Apple Git-31.1)

