From 1f9ea0e905ef928fa4d9d29c2146a2e1de04092c Mon Sep 17 00:00:00 2001
From: Daniel <ZetaetaDaniel@gmail.com>
Date: Sat, 12 May 2012 16:07:48 +0100
Subject: [PATCH] Adds ability to set the number of online players sent in
 ServerListPingEvent


diff --git a/src/main/java/org/bukkit/event/server/ServerListPingEvent.java b/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
index d8369bc..d93cb47 100644
--- a/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
@@ -11,7 +11,7 @@ public class ServerListPingEvent extends ServerEvent {
     private static final HandlerList handlers = new HandlerList();
     private final InetAddress address;
     private String motd;
-    private final int numPlayers;
+    private int numPlayers;
     private int maxPlayers;
 
     public ServerListPingEvent(final InetAddress address, final String motd, final int numPlayers, final int maxPlayers) {
@@ -58,6 +58,15 @@ public class ServerListPingEvent extends ServerEvent {
     }
 
     /**
+     * Set the number of players sent.
+     *
+     * @param numPlayers the number of players to be sent
+     */
+    public void setNumPlayers(int numPlayers) {
+        this.numPlayers = numPlayers;
+    }
+
+    /**
      * Get the maximum number of players sent.
      *
      * @return the maximum number of players
@@ -69,7 +78,7 @@ public class ServerListPingEvent extends ServerEvent {
     /**
      * Set the maximum number of players sent.
      *
-     * @param maxPlayers the maximum number of player
+     * @param maxPlayers the maximum number of players
      */
     public void setMaxPlayers(int maxPlayers) {
         this.maxPlayers = maxPlayers;
-- 
1.7.9.6 (Apple Git-31.1)

