From ec49c64b2f8603b637d328b21dd0496676b0049e Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 29 May 2013 23:52:37 -0500
Subject: [PATCH] Visibility API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 7ea894a..78a9870 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -566,6 +566,8 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public void hidePlayer(Player player);
 
+    public void hide(Entity entity);
+
     /**
      * Allows this player to see a player that was previously hidden
      *
@@ -573,12 +575,16 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public void showPlayer(Player player);
 
+    public void show(Entity entity);
+
     /**
-     * Checks to see if a player has been hidden from this player
+     * Checks to see if an entity has been hidden from this player
      *
      * @param player Player to check
-     * @return True if the provided player is not being hidden from this player
+     * @return True if the provided entity is not being hidden from this player
      */
+    public boolean canSee(Entity entity);
+
     public boolean canSee(Player player);
 
     /**
-- 
1.7.12.4 (Apple Git-37)

