From 1d434bd8f0ae478e5f2266be9935e1e67d14a639 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 10 Nov 2012 12:24:48 -0600
Subject: [PATCH] Add more whitelisting methods to Bukkit


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 6639c98..423f4f8 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -465,6 +465,27 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#isWhitelisted(OfflinePlayer player)
+     */
+    public static boolean isWhitelisted(OfflinePlayer player) {
+        return server.isWhitelisted(player);
+    }
+
+    /**
+     * @see Server#addWhitelist(OfflinePlayer player)
+     */
+    public static void addWhitelist(OfflinePlayer player) {
+        server.addWhitelist(player);
+    }
+
+    /**
+     * @see Server#removeWhitelist(OfflinePlayer player)
+     */
+    public static void removeWhitelist(OfflinePlayer player) {
+        server.removeWhitelist(player);
+    }
+
+    /**
      * @see Server#getConsoleSender()
      */
     public static ConsoleCommandSender getConsoleSender() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 97b9dfe..0b52067 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -176,6 +176,28 @@ public interface Server extends PluginMessageRecipient {
     public void reloadWhitelist();
 
     /**
+     * See if a player is on the whitelist or not
+     *
+     * @param player Player to test for whitelist status
+     * @return Wether or not the player is whitelisted
+     */
+    public boolean isWhitelisted(OfflinePlayer player);
+
+    /**
+     * Add a player to the whitelist
+     *
+     * @param player Player to add to the whitelist
+     */
+    public void addWhitelist(OfflinePlayer player);
+
+    /**
+     * Remove a player from the whitelist
+     *
+     * @param player Player to remove from the whitelist
+     */
+    public void removeWhitelist(OfflinePlayer player);
+
+    /**
      * Broadcast a message to all players.
      * <p>
      * This is the same as calling {@link #broadcast(java.lang.String, java.lang.String)} to {@link #BROADCAST_CHANNEL_USERS}
-- 
1.7.9.6 (Apple Git-31.1)

