From 6e74fe0f16850c425b77e37856b6b324ac1b75c6 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Thu, 26 Jun 2014 19:16:40 -0500
Subject: [PATCH] Default to UTF-8 encoding


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index 0fd966c..85c14a0 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -226,7 +226,7 @@ public final class PluginDescriptionFile {
     private List<Permission> permissions = null;
     private Map<?, ?> lazyPermissions = null;
     private PermissionDefault defaultPerm = PermissionDefault.OP;
-    private Set<PluginAwareness> awareness = ImmutableSet.of();
+    private Set<PluginAwareness> awareness = ImmutableSet.of((PluginAwareness) PluginAwareness.Flags.UTF8);
 
     public PluginDescriptionFile(final InputStream stream) throws InvalidDescriptionException {
         loadMap(asMap(YAML.get().load(stream)));
-- 
1.8.5.1

