From 09148c33cd1fb050f52dbc6a27318bbab8fa3561 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Feb 2015 08:28:02 -0500
Subject: [PATCH] Fix TNT physics


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index bbe28c6..403ad42 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -335,6 +335,13 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getWaterPushesTNT()
+     */
+    public static boolean getWaterPushesTNT() {
+        return server.getWaterPushesTNT();
+    }
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2bc7ccb..86fe258 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -273,6 +273,11 @@ public interface Server extends PluginMessageRecipient {
     public boolean getFetchSkulls();
 
     /**
+     * Should flowing water move primed TNT entities?
+     */
+    public boolean getWaterPushesTNT();
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
-- 
1.9.0

