From 43cca469fb57cb26dd5d3eeb57c5673cf0f442d6 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Mar 2015 03:02:27 -0400
Subject: [PATCH] Relative teleport API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8ee4444..74dde51 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1567,4 +1567,12 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * Hide any title that is currently visible to the player
      */
     public void hideTitle();
+
+    /**
+     * Seamlessly move the player by the given offset and rotation angles.
+     * Because the relative calculation is applied by the client, you cannot
+     * know the player's absolute location immediately after teleporting.
+     */
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause);
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch);
 }
-- 
1.9.0

