From 9b521f1cdbc54d7104bb6fd36ce49159225acc6f Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Mon, 1 Sep 2014 00:34:00 -0500
Subject: [PATCH] Snapshot protocol


diff --git a/src/main/java/org/bukkit/ProtocolVersion.java b/src/main/java/org/bukkit/ProtocolVersion.java
new file mode 100644
index 0000000..0df20f6
--- /dev/null
+++ b/src/main/java/org/bukkit/ProtocolVersion.java
@@ -0,0 +1,41 @@
+package org.bukkit;
+
+import java.util.*;
+
+/**
+ * The ProtocolVersion enum is an enumeration of all protocol versions supported
+ * by the server implementation.
+ */
+public enum ProtocolVersion {
+
+    MC_1_7("1.7", 4, 5),
+    MC_1_8("1.8", 41, 42, 43, 44, 45, 46, 47),
+    UNKNOWN("Unknown", -1);
+
+    private static final ProtocolVersion[] VALUES = ProtocolVersion.values();
+    private final List<Integer> protocolVersions;
+    private final String versionName;
+
+    private ProtocolVersion(String versionName, Integer... protocolVersions) {
+        this.versionName = versionName;
+        this.protocolVersions = new ArrayList<Integer>(Arrays.asList(protocolVersions));
+    }
+
+    public List<Integer> getProtocolVersions() {
+        return protocolVersions;
+    }
+
+    public String getVersionName() {
+        return versionName;
+    }
+
+    public static ProtocolVersion getVersion(int versionNumber) {
+        for (ProtocolVersion ver : VALUES) {
+            if (ver.getProtocolVersions().contains(versionNumber)) {
+                return ver;
+            }
+        }
+
+        return ProtocolVersion.UNKNOWN;
+    }
+}
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c4cd0d7..fbe28f0 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1162,4 +1162,13 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * @return The locale
      */
     public String getLocale();
- }
+
+    // XimeBukkit start - 1.8 protocol support
+    /**
+     * Gets the client's version.
+     *
+     * @return {@link ProtocolVersion} enum describing what version the client is using
+     */
+    public org.bukkit.ProtocolVersion getProtocolVersion();
+    // XimeBukkit end
+}
-- 
1.8.4.msysgit.0

