From ba8f22415ddfc8bd123b2198c9b6e5418d460757 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 6 Sep 2014 14:52:27 -0400
Subject: [PATCH] Add nameTagVisibility API


diff --git a/src/main/java/org/bukkit/scoreboard/NameTagVisibility.java b/src/main/java/org/bukkit/scoreboard/NameTagVisibility.java
new file mode 100644
index 0000000..bb5d84e
--- /dev/null
+++ b/src/main/java/org/bukkit/scoreboard/NameTagVisibility.java
@@ -0,0 +1,15 @@
+package org.bukkit.scoreboard;
+
+public enum NameTagVisibility {
+    ALWAYS("always"), HIDE_FOR_OTHER_TEAMS("hideForOtherTeams"), HIDE_FOR_OWN_TEAM("hideForOwnTeam"), NEVER("never");
+
+    private final String token;
+
+    NameTagVisibility(String token) {
+        this.token = token;
+    }
+
+    public String token() {
+        return token;
+    }
+}
diff --git a/src/main/java/org/bukkit/scoreboard/Team.java b/src/main/java/org/bukkit/scoreboard/Team.java
index aeac90a..d75dc05 100644
--- a/src/main/java/org/bukkit/scoreboard/Team.java
+++ b/src/main/java/org/bukkit/scoreboard/Team.java
@@ -110,6 +110,10 @@ public interface Team {
      */
     void setCanSeeFriendlyInvisibles(boolean enabled) throws IllegalStateException;
 
+    NameTagVisibility nameTagVisibility() throws IllegalStateException;
+
+    void setNameTagVisibility(NameTagVisibility visibility) throws IllegalStateException;
+
     /**
      * Gets the Set of players on the team
      *
-- 
1.8.5.2 (Apple Git-48)

