From ad1b5845f1c16371b1deb5aaf39ff0577816ba15 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Feb 2015 08:28:02 -0500
Subject: [PATCH] Fix TNT physics


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 96b0eff..ecf29ce 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -330,6 +330,13 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getWaterPushesTNT()
+     */
+    public static boolean getWaterPushesTNT() {
+        return server.getWaterPushesTNT();
+    }
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index cd73e9a..af2806b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -268,6 +268,11 @@ public interface Server extends PluginMessageRecipient {
     public boolean getFetchSkulls();
 
     /**
+     * Should flowing water move primed TNT entities?
+     */
+    public boolean getWaterPushesTNT();
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
-- 
1.9.0

