From 2cfb93dc00b3b924ca4dd68f03da678960426c57 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Mar 2015 03:02:27 -0400
Subject: [PATCH] Relative teleport API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index d840072..1389d3e 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1314,4 +1314,12 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * Otherwise, it may be affected by ground friction before being sent.
      */
     public void updateVelocity();
+
+    /**
+     * Seamlessly move the player by the given offset and rotation angles.
+     * Because the relative calculation is applied by the client, you cannot
+     * know the player's absolute location immediately after teleporting.
+     */
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause);
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch);
 }
-- 
1.9.0

