From 5064e5fe5c1cc88646c094a1e52e5908df5442f4 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 16 Mar 2015 22:43:13 -0400
Subject: [PATCH] Force update player velocity


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3349113..e3db0bf 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1302,4 +1302,11 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * Clear the subtitle text for this player and reset title display times to default values
      */
     public void resetTitle();
+
+    /**
+     * Immediately send a velocity update packet to the player, after firing a {@link org.bukkit.event.player.PlayerVelocityEvent}.
+     * This should be called immediately after {@link #setVelocity} to ensure the client receives the velocity unaltered.
+     * Otherwise, it may be affected by ground friction before being sent.
+     */
+    public void updateVelocity();
 }
-- 
1.9.0

