From 547c45604ef63000bbd7ba670cb6ff1a71d43c95 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 20 Nov 2014 02:19:51 -0500
Subject: [PATCH] Call Plugin.onDisable before disabling the plugin


diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index 19893f3..4c62bc4 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -309,13 +309,14 @@ public abstract class JavaPlugin extends PluginBase {
      * @param enabled true if enabled, otherwise false
      */
     protected final void setEnabled(final boolean enabled) {
-        if (isEnabled != enabled) {
-            isEnabled = enabled;
-
-            if (isEnabled) {
-                onEnable();
-            } else {
+        if(!isEnabled && enabled) {
+            isEnabled = true;
+            onEnable();
+        } else if(isEnabled && !enabled) {
+            try {
                 onDisable();
+            } finally {
+                isEnabled = false;
             }
         }
     }
-- 
1.9.0

