From 2fb2452ebafd615d12bbe07ba3f17c566fa81c67 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 May 2015 01:40:20 -0400
Subject: [PATCH] Support unbreakable items


diff --git a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
index 459e09c..80a2c3b 100644
--- a/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/ItemMeta.java
@@ -155,6 +155,10 @@ public interface ItemMeta extends Cloneable, ConfigurationSerializable {
      */
     boolean hasItemFlag(ItemFlag flag);
 
+    boolean isUnbreakable();
+
+    void setUnbreakable(boolean unbreakable);
+
     @SuppressWarnings("javadoc")
     ItemMeta clone();
 }
-- 
1.9.0

