From 810294ca61553bec2b194303cb1bc737afa55f2f Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 4 Feb 2015 03:39:54 -0500
Subject: [PATCH] Add setting to disable skull fetching


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7f6c0c0..69f05fb 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -567,6 +567,13 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getFetchSkulls()
+     */
+    public static boolean getFetchSkulls() {
+        return server.getFetchSkulls();
+    }
+
+    /**
      * @see Server#getTicksPerAnimalSpawns()
      */
     public static int getTicksPerAnimalSpawns() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 999a034..ccf6d04 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -250,6 +250,11 @@ public interface Server extends PluginMessageRecipient {
     public long getConnectionThrottle();
 
     /**
+     * Should player skulls with missing skin data be fetched from Mojang?
+     */
+    public boolean getFetchSkulls();
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
-- 
1.9.0

