From 99e9b64e73ecc307d3485b7e3ea11cbf9d505b7a Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 04:18:12 -0500
Subject: [PATCH] Multi-trade API


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 1ceb072..b3dc3d7 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -118,6 +118,14 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, Permissible, Inv
     public InventoryView openMerchant(Villager trader, boolean force);
 
     /**
+     * Makes a copy of the given villager's current trading offers and opens
+     * a trading window for this player, allowing them to interact with those
+     * offers, without affecting the villager's state, or any other players who
+     * might be trading with the villager.
+     */
+    public InventoryView openMerchantCopy(Villager trader);
+
+    /**
      * Force-closes the currently open inventory view for this player, if any.
      */
     public void closeInventory();
-- 
1.9.0

