From 8dc92bbf1f867b0bddd731e1f4afd20db0513797 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 05:57:57 -0500
Subject: [PATCH] Absorption API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 1dd7712..0348884 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -227,6 +227,16 @@ public interface LivingEntity extends Attributable, Entity, Damageable, Projecti
     public void setNoDamageTicks(int ticks);
 
     /**
+     * Get the amount of absorption this entity currently has, in half-hearts.
+     */
+    float getAbsorption();
+
+    /**
+     * Set the amount of absorption this entity currently has, in half-hearts.
+     */
+    void setAbsorption(float absorption);
+
+    /**
      * Gets the player identified as the killer of the living entity.
      * <p>
      * May be null.
-- 
1.9.0

