From 2d60e15799c1c2b5cb7abee0cdb80850380ef5b1 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 6 Mar 2016 08:47:51 -0500
Subject: [PATCH] Team collision API


diff --git a/src/main/java/org/bukkit/scoreboard/CollisionRule.java b/src/main/java/org/bukkit/scoreboard/CollisionRule.java
new file mode 100644
index 0000000..509e5c1
--- /dev/null
+++ b/src/main/java/org/bukkit/scoreboard/CollisionRule.java
@@ -0,0 +1,36 @@
+package org.bukkit.scoreboard;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public enum CollisionRule {
+
+    ALWAYS("always"),
+    NEVER("never"),
+    PUSH_OTHER_TEAMS("pushOtherTeams"),
+    PUSH_OWN_TEAM("pushOwnTeam");
+
+    private final String key;
+
+    CollisionRule(String key) {
+        this.key = key;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    @Override
+    public String toString() {
+        return getKey();
+    }
+
+    private static final Map<String, CollisionRule> byKey = new HashMap<String, CollisionRule>();
+    static {
+        for(CollisionRule rule : values()) byKey.put(rule.key, rule);
+    }
+
+    public static CollisionRule byKey(String key) {
+        return byKey.get(key);
+    }
+}
diff --git a/src/main/java/org/bukkit/scoreboard/Team.java b/src/main/java/org/bukkit/scoreboard/Team.java
index 9052b4a..3854b65 100644
--- a/src/main/java/org/bukkit/scoreboard/Team.java
+++ b/src/main/java/org/bukkit/scoreboard/Team.java
@@ -132,6 +132,16 @@ public interface Team {
     void setNameTagVisibility(NameTagVisibility visibility) throws IllegalArgumentException;
 
     /**
+     * Get the player collision rule for this team
+     */
+    CollisionRule getCollisionRule();
+
+    /**
+     * Set the player collision rule for this team
+     */
+    void setCollisionRule(CollisionRule rule);
+
+    /**
      * Gets the Set of players on the team
      *
      * @return players on the team
-- 
1.9.0

