From 7bb9ad66b62654fc20eaabf277ce6324a94003fa Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 20 Mar 2016 06:44:49 -0400
Subject: [PATCH] Access items by EquipmentSlot


diff --git a/src/main/java/org/bukkit/inventory/PlayerInventory.java b/src/main/java/org/bukkit/inventory/PlayerInventory.java
index 557cc04..6ce8509 100644
--- a/src/main/java/org/bukkit/inventory/PlayerInventory.java
+++ b/src/main/java/org/bukkit/inventory/PlayerInventory.java
@@ -179,6 +179,10 @@ public interface PlayerInventory extends Inventory {
     @Deprecated
     public void setItemInHand(ItemStack stack);
 
+    ItemStack getItem(EquipmentSlot slot);
+
+    void setItem(EquipmentSlot slot, ItemStack stack);
+
     /**
      * Get the slot number of the currently held item
      *
-- 
1.9.0

