From a9f29ef8cbe34a1f56b4663c1d47936922c100c5 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 29 Mar 2016 02:08:56 -0400
Subject: [PATCH] Attack cooldown info API


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 78e1272..a9278b2 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -230,4 +230,21 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, Permissible, Inv
      *         random spawning and despawning near them.
      */
     public boolean getAffectsSpawning();
+
+    /**
+     * Get the number of ticks since this entity last swung their weapon.
+     *
+     * Damage calculations are ultimately based on this value.
+     */
+    int getAttackCooldownTicks();
+
+    /**
+     * Get the coefficient that would be applied by the attack cooldown to a
+     * melee attack, at this moment, from this entity.
+     *
+     * This is calculated from the entity's attackSpeed attribute, and the
+     * number of ticks since their last swing. It current always falls in
+     * the range 0.2 to 1.0
+     */
+    float getAttackCooldownCoefficient();
 }
-- 
1.9.0

