From 325fbb7d62e34f5f344c5294607c2f1ec18ce484 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 21 Mar 2016 01:41:04 -0400
Subject: [PATCH] Multi-passenger API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 2bd8d64..47f19a4 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -245,6 +245,28 @@ public interface Entity extends Metadatable, CommandSender, Physical {
     public abstract boolean eject();
 
     /**
+     * @return true if the given entity is a passenger of this vehicle
+     */
+    boolean hasPassenger(Entity passenger);
+
+    /**
+     * @return all of the current passengers in this vehicle
+     */
+    List<Entity> getPassengers();
+
+    /**
+     * Make the given entities passengers of this vehicle. Any existing
+     * passengers who are not in the list are ejected.
+     * @return given entities that could NOT become passengers for whatever reason
+     */
+    List<Entity> setPassengers(List<Entity> passengers);
+
+    /**
+     * Eject all passengers from this vehicle
+     */
+    void ejectAll();
+
+    /**
      * Returns the distance this entity has fallen
      *
      * @return The distance.
@@ -324,6 +346,12 @@ public interface Entity extends Metadatable, CommandSender, Physical {
     public boolean isInsideVehicle();
 
     /**
+     * Enter the given vehicle
+     * @return true if successful
+     */
+    boolean enterVehicle(Entity vehicle);
+
+    /**
      * Leave the current vehicle. If the entity is currently in a vehicle (and
      * is removed from it), true will be returned, otherwise false will be
      * returned.
-- 
1.9.0

