From e189f6b0edc4540beccd975a420a386788e8082f Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 5 Aug 2016 00:54:33 -0400
Subject: [PATCH] Fix hashCode and equals in BlockVector


diff --git a/src/main/java/org/bukkit/util/BlockVector.java b/src/main/java/org/bukkit/util/BlockVector.java
index bdf8f6d..3a51598 100644
--- a/src/main/java/org/bukkit/util/BlockVector.java
+++ b/src/main/java/org/bukkit/util/BlockVector.java
@@ -79,12 +79,13 @@ public class BlockVector extends Vector {
      */
     @Override
     public boolean equals(Object obj) {
-        if (!(obj instanceof BlockVector)) {
-            return false;
-        }
-        BlockVector other = (BlockVector) obj;
+        if(this == obj) return true;
+        if (!(obj instanceof BlockVector)) return false;
 
-        return (int) other.getX() == (int) this.x && (int) other.getY() == (int) this.y && (int) other.getZ() == (int) this.z;
+        final BlockVector that = (BlockVector) obj;
+        return this.getBlockX() == that.getBlockX() &&
+               this.getBlockY() == that.getBlockY() &&
+               this.getBlockZ() == that.getBlockZ();
 
     }
 
@@ -95,7 +96,8 @@ public class BlockVector extends Vector {
      */
     @Override
     public int hashCode() {
-        return (Integer.valueOf((int) x).hashCode() >> 13) ^ (Integer.valueOf((int) y).hashCode() >> 7) ^ Integer.valueOf((int) z).hashCode();
+        // 31 bits gives us roughly 10 bits per axis, and 1021 is the closest prime to 2^10
+        return getBlockX() + 1021 * (getBlockY() + 1021 * getBlockZ());
     }
 
     /**
-- 
1.9.0

